<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>protflow.metrics package &mdash; ProtFlow 1.0.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5d32c60e"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="_static/documentation_options.js?v=8d563738"></script>
        <script src="_static/doctools.js?v=9a2dae69"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="protflow.tools package" href="protflow.tools.html" />
    <link rel="prev" title="protflow package" href="protflow.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            ProtFlow
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="protflow.html">protflow package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="protflow.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">protflow.metrics package</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-protflow.metrics.fpocket">protflow.metrics.fpocket module</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-protflow.metrics.protparam">protflow.metrics.protparam module</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-protflow.metrics.rmsd">protflow.metrics.rmsd module</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-protflow.metrics.tmscore">protflow.metrics.tmscore module</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-protflow.metrics">Module contents</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="protflow.tools.html">protflow.tools package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="protflow.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="protflow.html#protflow-config-module">protflow.config module</a></li>
<li class="toctree-l2"><a class="reference internal" href="protflow.html#protflow-config-template-module">protflow.config_template module</a></li>
<li class="toctree-l2"><a class="reference internal" href="protflow.html#module-protflow.jobstarters">protflow.jobstarters module</a></li>
<li class="toctree-l2"><a class="reference internal" href="protflow.html#module-protflow.poses">protflow.poses module</a></li>
<li class="toctree-l2"><a class="reference internal" href="protflow.html#module-protflow.residues">protflow.residues module</a></li>
<li class="toctree-l2"><a class="reference internal" href="protflow.html#module-protflow.runners">protflow.runners module</a></li>
<li class="toctree-l2"><a class="reference internal" href="protflow.html#module-contents">Module contents</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">ProtFlow</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="protflow.html">protflow package</a></li>
      <li class="breadcrumb-item active">protflow.metrics package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/protflow.metrics.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="protflow-metrics-package">
<h1>protflow.metrics package<a class="headerlink" href="#protflow-metrics-package" title="Link to this heading"></a></h1>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Link to this heading"></a></h2>
</section>
<section id="module-protflow.metrics.fpocket">
<span id="protflow-metrics-fpocket-module"></span><h2>protflow.metrics.fpocket module<a class="headerlink" href="#module-protflow.metrics.fpocket" title="Link to this heading"></a></h2>
<section id="fpocket-module">
<h3>FPocket Module<a class="headerlink" href="#fpocket-module" title="Link to this heading"></a></h3>
<p>This module provides the functionality to integrate FPocket within the ProtFlow framework. It offers tools to run FPocket, handle its inputs and outputs, and process the resulting data in a structured and automated manner.</p>
<section id="detailed-description">
<h4>Detailed Description<a class="headerlink" href="#detailed-description" title="Link to this heading"></a></h4>
<p>The <cite>FPocket</cite> class encapsulates the functionality necessary to execute FPocket runs. It manages the configuration of paths to essential scripts and Python executables, sets up the environment, and handles the execution of FPocket processes. It also includes methods for collecting and processing output data, ensuring that the results are organized and accessible for further analysis within the ProtFlow ecosystem.
The module is designed to streamline the integration of FPocket into larger computational workflows. It supports the automatic setup of job parameters, execution of FPocket commands, and parsing of output files into a structured DataFrame format. This facilitates subsequent data analysis and visualization steps.</p>
</section>
<section id="usage">
<h4>Usage<a class="headerlink" href="#usage" title="Link to this heading"></a></h4>
<p>To use this module, create an instance of the <cite>FPocket</cite> class and invoke its <cite>run</cite> method with appropriate parameters. The module will handle the configuration, execution, and result collection processes. Detailed control over the FPocket process is provided through various parameters, allowing for customized runs tailored to specific research needs.</p>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize and use the <cite>FPocket</cite> class within a ProtFlow pipeline:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">fpocket</span> <span class="kn">import</span> <span class="n">FPocket</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the FPocket class</span>
<span class="n">fpocket</span> <span class="o">=</span> <span class="n">FPocket</span><span class="p">()</span>

<span class="c1"># Run the FPocket process</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">fpocket</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;--some-option value&quot;</span><span class="p">,</span>
    <span class="n">pose_options</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;--specific-option value&quot;</span><span class="p">],</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="further-details">
<h4>Further Details<a class="headerlink" href="#further-details" title="Link to this heading"></a></h4>
<blockquote>
<div><ul class="simple">
<li><p>Edge Cases: The module handles various edge cases, such as empty pose lists and the need to overwrite previous results. It ensures robust error handling and logging for easier debugging and verification of the FPocket process.</p></li>
<li><p>Customizability: Users can customize the FPocket process through multiple parameters, including specific options for the FPocket script and options for handling pose-specific parameters.</p></li>
<li><p>Integration: The module seamlessly integrates with other components of the ProtFlow framework, leveraging shared configurations and data structures to provide a cohesive user experience.</p></li>
</ul>
</div></blockquote>
<p>This module is intended for researchers and developers who need to incorporate FPocket into their protein design and analysis workflows. By automating many of the setup and execution steps, it allows users to focus on interpreting results and advancing their scientific inquiries.</p>
<p class="rubric">Notes</p>
<p>This module is part of the ProtFlow package and is designed to work in tandem with other components of the package, especially those related to job management in HPC environments.</p>
</section>
<section id="author">
<h4>Author<a class="headerlink" href="#author" title="Link to this heading"></a></h4>
<p>Markus Braun, Adrian Tripp</p>
</section>
<section id="version">
<h4>Version<a class="headerlink" href="#version" title="Link to this heading"></a></h4>
<p>0.1.0</p>
</section>
</section>
<dl class="py class">
<dt class="sig sig-object py" id="protflow.metrics.fpocket.FPocket">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">protflow.metrics.fpocket.</span></span><span class="sig-name descname"><span class="pre">FPocket</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fpocket_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><span class="pre">JobStarter</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/metrics/fpocket.html#FPocket"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.fpocket.FPocket" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="protflow.html#protflow.runners.Runner" title="protflow.runners.Runner"><code class="xref py py-class docutils literal notranslate"><span class="pre">Runner</span></code></a></p>
<section id="fpocket-class">
<h3>FPocket Class<a class="headerlink" href="#fpocket-class" title="Link to this heading"></a></h3>
<p>The <cite>FPocket</cite> class is a specialized class designed to facilitate the execution of FPocket within the ProtFlow framework. It extends the <cite>Runner</cite> class and incorporates specific methods to handle the setup, execution, and data collection associated with FPocket processes.</p>
<section id="id1">
<h4>Detailed Description<a class="headerlink" href="#id1" title="Link to this heading"></a></h4>
<p>The <cite>FPocket</cite> class manages all aspects of running FPocket simulations. It handles the configuration of necessary scripts and executables, prepares the environment for pocket detection processes, and executes the FPocket commands. Additionally, it collects and processes the output data, organizing it into a structured format for further analysis.</p>
<dl class="simple">
<dt>Key functionalities include:</dt><dd><ul class="simple">
<li><p>Setting up paths to FPocket scripts and executables.</p></li>
<li><p>Configuring job starter options, either automatically or manually.</p></li>
<li><p>Handling the execution of FPocket commands with support for multiple options and pose-specific parameters.</p></li>
<li><p>Collecting and processing output data into a pandas DataFrame.</p></li>
<li><p>Ensuring robust error handling and logging for easier debugging and verification of the FPocket process.</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">rtype<span class="colon">:</span></dt>
<dd class="field-odd"><p>An instance of the <cite>FPocket</cite> class, configured to run FPocket processes and handle outputs efficiently.</p>
</dd>
<dt class="field-even">raises - FileNotFoundError<span class="colon">:</span></dt>
<dd class="field-even"><p>If required files or directories are not found during the execution process.:</p>
</dd>
<dt class="field-odd">raises - ValueError<span class="colon">:</span></dt>
<dd class="field-odd"><p>If invalid arguments are provided to the methods.:</p>
</dd>
<dt class="field-even">raises - TypeError<span class="colon">:</span></dt>
<dd class="field-even"><p>If provided options are not of the expected type.:</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize and use the <cite>FPocket</cite> class:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">fpocket</span> <span class="kn">import</span> <span class="n">FPocket</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the FPocket class</span>
<span class="n">fpocket</span> <span class="o">=</span> <span class="n">FPocket</span><span class="p">()</span>

<span class="c1"># Run the FPocket process</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">fpocket</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;--some-option value&quot;</span><span class="p">,</span>
    <span class="n">pose_options</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;--specific-option value&quot;</span><span class="p">],</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id2">
<h4>Further Details<a class="headerlink" href="#id2" title="Link to this heading"></a></h4>
<blockquote>
<div><ul class="simple">
<li><p>Edge Cases: The class includes handling for various edge cases, such as empty pose lists, the need to overwrite previous results, and the presence of existing score files.</p></li>
<li><p>Customization: The class provides extensive customization options through its parameters, allowing users to tailor the FPocket process to their specific needs.</p></li>
<li><p>Integration: Seamlessly integrates with other ProtFlow components, leveraging shared configurations and data structures for a unified workflow.</p></li>
</ul>
</div></blockquote>
<p>The FPocket class is intended for researchers and developers who need to perform FPocket simulations as part of their protein design and analysis workflows. It simplifies the process, allowing users to focus on analyzing results and advancing their research.</p>
<dl class="py method">
<dt class="sig sig-object py" id="protflow.metrics.fpocket.FPocket.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fpocket_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><span class="pre">JobStarter</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/metrics/fpocket.html#FPocket.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.fpocket.FPocket.__init__" title="Link to this definition"></a></dt>
<dd><p>Initialize the FPocket class with the specified path and jobstarter configuration.</p>
<p>This constructor sets up the FPocket instance by configuring the path to the FPocket executable and initializing the jobstarter object. It ensures that the necessary components are in place for running FPocket processes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fpocket_path</strong> (<em>str</em><em>, </em><em>optional</em>) – The path to the FPocket executable. Defaults to the path specified in the ProtFlow configuration (<cite>FPOCKET_PATH</cite>).</p></li>
<li><p><strong>jobstarter</strong> (<a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><em>JobStarter</em></a><em>, </em><em>optional</em>) – An instance of the JobStarter class, which manages job execution. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An instance of the FPocket class, ready to run FPocket processes.</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – If the fpocket_path is not provided or is invalid.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize the FPocket class:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">fpocket</span> <span class="kn">import</span> <span class="n">FPocket</span>

<span class="c1"># Initialize the FPocket class with default settings</span>
<span class="n">fpocket</span> <span class="o">=</span> <span class="n">FPocket</span><span class="p">()</span>

<span class="c1"># Initialize the FPocket class with a specific jobstarter</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>
<span class="n">fpocket</span> <span class="o">=</span> <span class="n">FPocket</span><span class="p">(</span><span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Path Configuration:</strong> Ensures the FPocket executable path is set correctly, raising an error if the path is not provided or invalid.</p></li>
<li><p><strong>Job Management:</strong> Initializes the jobstarter object to manage the execution of FPocket commands, allowing for integration with job scheduling systems.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="protflow.metrics.fpocket.FPocket.index_layers">
<span class="sig-name descname"><span class="pre">index_layers</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">0</span></em><a class="headerlink" href="#protflow.metrics.fpocket.FPocket.index_layers" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.metrics.fpocket.FPocket.prep_fpocket_options">
<span class="sig-name descname"><span class="pre">prep_fpocket_options</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pose_options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/protflow/metrics/fpocket.html#FPocket.prep_fpocket_options"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.fpocket.FPocket.prep_fpocket_options" title="Link to this definition"></a></dt>
<dd><p>Prepare options for the FPocket process based on given parameters.</p>
<p>This method processes and prepares the options and pose-specific options for the FPocket run. It filters out forbidden options, merges general options with pose-specific options, and formats them for inclusion in the FPocket commands.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>poses</strong> (<a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><em>Poses</em></a>) – The Poses object containing the protein structures.</p></li>
<li><p><strong>options</strong> (<em>str</em><em> or </em><em>list</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – General options for the FPocket script. Defaults to None.</p></li>
<li><p><strong>pose_options</strong> (<em>str</em><em> or </em><em>list</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – A list of pose-specific options for the FPocket script. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A list of formatted option strings for each pose, ready to be used in the FPocket commands.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[str]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>TypeError</strong> – If options or pose_options are not of the expected type.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>prep_fpocket_options</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">fpocket</span> <span class="kn">import</span> <span class="n">FPocket</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">fpocket</span> <span class="o">=</span> <span class="n">FPocket</span><span class="p">()</span>

<span class="c1"># Prepare FPocket options</span>
<span class="n">options</span> <span class="o">=</span> <span class="s2">&quot;--some-option value&quot;</span>
<span class="n">pose_options</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;--specific-option value&quot;</span><span class="p">]</span>
<span class="n">prepared_options</span> <span class="o">=</span> <span class="n">fpocket</span><span class="o">.</span><span class="n">prep_fpocket_options</span><span class="p">(</span><span class="n">poses</span><span class="p">,</span> <span class="n">options</span><span class="p">,</span> <span class="n">pose_options</span><span class="p">)</span>

<span class="c1"># Output the prepared options</span>
<span class="nb">print</span><span class="p">(</span><span class="n">prepared_options</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Option Processing:</strong> Merges general and pose-specific options, ensuring that forbidden options are removed and the final option strings are correctly formatted.</p></li>
<li><p><strong>Customization:</strong> Allows for extensive customization of the FPocket process through both general and pose-specific options, providing flexibility in configuring FPocket runs.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.metrics.fpocket.FPocket.run">
<span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><span class="pre">JobStarter</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pose_options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_full_scores</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span></span><a class="reference internal" href="_modules/protflow/metrics/fpocket.html#FPocket.run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.fpocket.FPocket.run" title="Link to this definition"></a></dt>
<dd><p>Execute the FPocket process with given poses and jobstarter configuration.</p>
<p>This method sets up and runs the FPocket process using the provided poses and jobstarter object. It handles the configuration, execution, and collection of output data, ensuring that the results are organized and accessible for further analysis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>poses</strong> (<a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><em>Poses</em></a>) – The Poses object containing the protein structures.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – A prefix used to name and organize the output files.</p></li>
<li><p><strong>jobstarter</strong> (<a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><em>JobStarter</em></a><em>, </em><em>optional</em>) – An instance of the JobStarter class, which manages job execution. Defaults to None.</p></li>
<li><p><strong>options</strong> (<em>str</em><em> or </em><em>list</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – Additional options for the FPocket script. Defaults to None.</p></li>
<li><p><strong>pose_options</strong> (<em>str</em><em> or </em><em>list</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – A list of pose-specific options for the FPocket script. Defaults to None.</p></li>
<li><p><strong>return_full_scores</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, include detailed scores for each pocket in the output. Defaults to False.</p></li>
<li><p><strong>overwrite</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, overwrite existing output files. Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An updated Poses object containing the processed poses and results of the FPocket process.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses">Poses</a></p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>FileNotFoundError</strong> – If required files or directories are not found during the execution process.</p></li>
<li><p><strong>ValueError</strong> – If invalid arguments are provided to the method.</p></li>
<li><p><strong>TypeError</strong> – If options or pose_options are not of the expected type.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>run</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">fpocket</span> <span class="kn">import</span> <span class="n">FPocket</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the FPocket class</span>
<span class="n">fpocket</span> <span class="o">=</span> <span class="n">FPocket</span><span class="p">()</span>

<span class="c1"># Run the FPocket process</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">fpocket</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;--some-option value&quot;</span><span class="p">,</span>
    <span class="n">pose_options</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;--specific-option value&quot;</span><span class="p">],</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Setup and Execution:</strong> The method ensures that the environment is correctly set up, directories are prepared, and necessary commands are constructed and executed. It moves the poses to the working directory and compiles the FPocket commands for execution.</p></li>
<li><p><strong>Output Management:</strong> The method handles the collection and processing of output data, ensuring that results are organized into a structured DataFrame. It includes the location of each pocket and integrates the results back into the Poses object.</p></li>
<li><p><strong>Customization:</strong> Extensive customization options are provided through parameters, allowing users to tailor the FPocket process to their specific needs, including the ability to specify additional FPocket options and pose-specific parameters.</p></li>
</ul>
</dd>
</dl>
<p>This method is designed to streamline the execution of FPocket processes within the ProtFlow framework, making it easier for researchers and developers to perform and analyze pocket detection simulations.</p>
</dd></dl>

</section>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="protflow.metrics.fpocket.collect_fpocket_output">
<span class="sig-prename descclassname"><span class="pre">protflow.metrics.fpocket.</span></span><span class="sig-name descname"><span class="pre">collect_fpocket_output</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_full_scores</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/protflow/metrics/fpocket.html#collect_fpocket_output"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.fpocket.collect_fpocket_output" title="Link to this definition"></a></dt>
<dd><p>Collect output from a single FPocket output file.</p>
<p>This function processes the output of a single FPocket output file, extracting scores and other relevant information into a pandas DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>output_file</strong> (<em>str</em>) – The path to the FPocket output file.</p></li>
<li><p><strong>return_full_scores</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, include detailed scores for each pocket in the output. Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A DataFrame containing the processed output from the FPocket file.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>collect_fpocket_output</cite> function:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">fpocket</span> <span class="kn">import</span> <span class="n">collect_fpocket_output</span>

<span class="c1"># Specify the output file</span>
<span class="n">output_file</span> <span class="o">=</span> <span class="s2">&quot;path/to/output_file&quot;</span>

<span class="c1"># Collect output</span>
<span class="n">output</span> <span class="o">=</span> <span class="n">collect_fpocket_output</span><span class="p">(</span><span class="n">output_file</span><span class="p">,</span> <span class="n">return_full_scores</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># Display the output</span>
<span class="nb">print</span><span class="p">(</span><span class="n">output</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Output Processing:</strong> The function reads the FPocket output file, extracts relevant scores and information, and formats them into a DataFrame.</p></li>
<li><p><strong>Detailed Scores:</strong> If the return_full_scores parameter is set to True, the function includes detailed scores for each pocket in the DataFrame.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="protflow.metrics.fpocket.collect_fpocket_scores">
<span class="sig-prename descclassname"><span class="pre">protflow.metrics.fpocket.</span></span><span class="sig-name descname"><span class="pre">collect_fpocket_scores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_full_scores</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/protflow/metrics/fpocket.html#collect_fpocket_scores"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.fpocket.collect_fpocket_scores" title="Link to this definition"></a></dt>
<dd><p>Collect scores from an FPocket output directory.</p>
<p>This function collects and processes the scores from FPocket output files located in the specified directory. It aggregates the scores into a pandas DataFrame for further analysis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>output_dir</strong> (<em>str</em>) – The path to the directory containing FPocket output files.</p></li>
<li><p><strong>return_full_scores</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, include detailed scores for each pocket in the output. Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A DataFrame containing the collected scores from the FPocket output files.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>collect_fpocket_scores</cite> function:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">fpocket</span> <span class="kn">import</span> <span class="n">collect_fpocket_scores</span>

<span class="c1"># Specify the output directory</span>
<span class="n">output_dir</span> <span class="o">=</span> <span class="s2">&quot;path/to/output_directory&quot;</span>

<span class="c1"># Collect scores</span>
<span class="n">scores</span> <span class="o">=</span> <span class="n">collect_fpocket_scores</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">return_full_scores</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># Display the scores</span>
<span class="nb">print</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Score Aggregation:</strong> The function looks for FPocket output directories, extracts scores from each output file, and combines them into a single DataFrame.</p></li>
<li><p><strong>Detailed Scores:</strong> If the return_full_scores parameter is set to True, the function includes detailed scores for each pocket in the DataFrame.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="protflow.metrics.fpocket.get_outfile_name">
<span class="sig-prename descclassname"><span class="pre">protflow.metrics.fpocket.</span></span><span class="sig-name descname"><span class="pre">get_outfile_name</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">outdir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">str</span></span></span><a class="reference internal" href="_modules/protflow/metrics/fpocket.html#get_outfile_name"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.fpocket.get_outfile_name" title="Link to this definition"></a></dt>
<dd><p>Get the name of the output file from the output directory.</p>
<p>This function generates the name of the FPocket output file based on the specified output directory.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>outdir</strong> (<em>str</em>) – The path to the FPocket output directory.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The name of the output file within the specified directory.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>get_outfile_name</cite> function:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">fpocket</span> <span class="kn">import</span> <span class="n">get_outfile_name</span>

<span class="c1"># Specify the output directory</span>
<span class="n">outdir</span> <span class="o">=</span> <span class="s2">&quot;path/to/output_directory&quot;</span>

<span class="c1"># Get the output file name</span>
<span class="n">output_file_name</span> <span class="o">=</span> <span class="n">get_outfile_name</span><span class="p">(</span><span class="n">outdir</span><span class="p">)</span>

<span class="c1"># Display the output file name</span>
<span class="nb">print</span><span class="p">(</span><span class="n">output_file_name</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>File Naming:</strong> The function constructs the output file name by modifying the output directory name and appending the appropriate suffix.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="protflow.metrics.fpocket.parse_fpocket_outfile">
<span class="sig-prename descclassname"><span class="pre">protflow.metrics.fpocket.</span></span><span class="sig-name descname"><span class="pre">parse_fpocket_outfile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/protflow/metrics/fpocket.html#parse_fpocket_outfile"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.fpocket.parse_fpocket_outfile" title="Link to this definition"></a></dt>
<dd><p>Parse the FPocket output file to extract scores.</p>
<p>This function reads and parses the FPocket output file, extracting scores and other relevant information into a pandas DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>output_file</strong> (<em>str</em>) – The path to the FPocket output file.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A DataFrame containing the parsed scores from the FPocket output file.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>parse_fpocket_outfile</cite> function:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">fpocket</span> <span class="kn">import</span> <span class="n">parse_fpocket_outfile</span>

<span class="c1"># Specify the output file</span>
<span class="n">output_file</span> <span class="o">=</span> <span class="s2">&quot;path/to/output_file&quot;</span>

<span class="c1"># Parse the output file</span>
<span class="n">scores</span> <span class="o">=</span> <span class="n">parse_fpocket_outfile</span><span class="p">(</span><span class="n">output_file</span><span class="p">)</span>

<span class="c1"># Display the scores</span>
<span class="nb">print</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>File Parsing:</strong> The function reads the FPocket output file, extracts relevant scores and information, and formats them into a DataFrame.</p></li>
<li><p><strong>Score Extraction:</strong> The function processes the file line by line, extracting score data and organizing it into a structured format.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-protflow.metrics.protparam">
<span id="protflow-metrics-protparam-module"></span><h2>protflow.metrics.protparam module<a class="headerlink" href="#module-protflow.metrics.protparam" title="Link to this heading"></a></h2>
<section id="protparam-module">
<h3>ProtParam Module<a class="headerlink" href="#protparam-module" title="Link to this heading"></a></h3>
<p>This module provides the functionality to integrate ProtParam calculations within the ProtFlow framework. It offers tools to compute various protein sequence features using the BioPython <cite>Bio.SeqUtils.ProtParam</cite> module, handling inputs and outputs efficiently, and processing the resulting data in a structured and automated manner.</p>
<section id="id3">
<h4>Detailed Description<a class="headerlink" href="#id3" title="Link to this heading"></a></h4>
<p>The <cite>ProtParam</cite> class encapsulates the functionality necessary to execute ProtParam calculations. It manages the configuration of paths to essential scripts and Python executables, sets up the environment, and handles the execution of parameter calculations. It also includes methods for collecting and processing output data, ensuring that the results are organized and accessible for further analysis within the ProtFlow ecosystem.
The module is designed to streamline the integration of ProtParam into larger computational workflows. It supports the automatic setup of job parameters, execution of ProtParam commands, and parsing of output files into a structured DataFrame format. This facilitates subsequent data analysis and visualization steps.</p>
</section>
<section id="id4">
<h4>Usage<a class="headerlink" href="#id4" title="Link to this heading"></a></h4>
<p>To use this module, create an instance of the <cite>ProtParam</cite> class and invoke its <cite>run</cite> method with appropriate parameters. The module will handle the configuration, execution, and result collection processes. Detailed control over the ProtParam process is provided through various parameters, allowing for customized runs tailored to specific research needs.</p>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize and use the <cite>ProtParam</cite> class within a ProtFlow pipeline:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">protparam</span> <span class="kn">import</span> <span class="n">ProtParam</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the ProtParam class</span>
<span class="n">protparam</span> <span class="o">=</span> <span class="n">ProtParam</span><span class="p">()</span>

<span class="c1"># Run the ProtParam calculation process</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">protparam</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">seq_col</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">pH</span><span class="o">=</span><span class="mi">7</span><span class="p">,</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id5">
<h4>Further Details<a class="headerlink" href="#id5" title="Link to this heading"></a></h4>
<blockquote>
<div><ul class="simple">
<li><p>Edge Cases: The module handles various edge cases, such as empty pose lists and the need to overwrite previous results. It ensures robust error handling and logging for easier debugging and verification of the ProtParam process.</p></li>
<li><p>Customizability: Users can customize the ProtParam process through multiple parameters, including the pH for determining protein total charge, specific options for the ProtParam script, and options for handling pose-specific parameters.</p></li>
<li><p>Integration: The module seamlessly integrates with other components of the ProtFlow framework, leveraging shared configurations and data structures to provide a cohesive user experience.</p></li>
</ul>
</div></blockquote>
<p>This module is intended for researchers and developers who need to incorporate ProtParam calculations into their protein design and analysis workflows. By automating many of the setup and execution steps, it allows users to focus on interpreting results and advancing their scientific inquiries.</p>
<p class="rubric">Notes</p>
<p>This module is part of the ProtFlow package and is designed to work in tandem with other components of the package, especially those related to job management in HPC environments.</p>
</section>
<section id="authors">
<h4>Authors<a class="headerlink" href="#authors" title="Link to this heading"></a></h4>
<p>Markus Braun, Adrian Tripp</p>
</section>
<section id="id6">
<h4>Version<a class="headerlink" href="#id6" title="Link to this heading"></a></h4>
<p>0.1.0</p>
</section>
</section>
<dl class="py class">
<dt class="sig sig-object py" id="protflow.metrics.protparam.ProtParam">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">protflow.metrics.protparam.</span></span><span class="sig-name descname"><span class="pre">ProtParam</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">default_python</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'python3'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/metrics/protparam.html#ProtParam"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.protparam.ProtParam" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="protflow.html#protflow.runners.Runner" title="protflow.runners.Runner"><code class="xref py py-class docutils literal notranslate"><span class="pre">Runner</span></code></a></p>
<p>Class handling the calculation of protparams from sequence using the BioPython Bio.SeqUtils.ProtParam module</p>
<dl class="py method">
<dt class="sig sig-object py" id="protflow.metrics.protparam.ProtParam.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">default_python</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'python3'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/metrics/protparam.html#ProtParam.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.protparam.ProtParam.__init__" title="Link to this definition"></a></dt>
<dd><p>Initialize the ProtParam class.</p>
<p>This constructor sets up the necessary environment for running ProtParam calculations. It initializes the job starter and sets the path to the Python executable within the ProtFlow environment.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>jobstarter</strong> (<em>str</em><em>, </em><em>optional</em>) – The job starter to be used for executing ProtParam commands. If not provided, it defaults to None.</p></li>
<li><p><strong>default_python</strong> (<em>str</em><em>, </em><em>optional</em>) – The path to the Python executable within the ProtFlow environment. The default value is constructed using the PROTFLOW_ENV environment variable.</p></li>
</ul>
</dd>
</dl>
<dl class="py attribute">
<dt class="sig sig-object py" id="protflow.metrics.protparam.ProtParam.jobstarter">
<span class="sig-name descname"><span class="pre">jobstarter</span></span><a class="headerlink" href="#protflow.metrics.protparam.ProtParam.jobstarter" title="Link to this definition"></a></dt>
<dd><p>Stores the job starter to be used for executing ProtParam commands.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="protflow.metrics.protparam.ProtParam.python">
<span class="sig-name descname"><span class="pre">python</span></span><a class="headerlink" href="#protflow.metrics.protparam.ProtParam.python" title="Link to this definition"></a></dt>
<dd><p>The path to the Python executable within the ProtFlow environment.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="field-list simple">
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>FileNotFoundError</strong> – If the default Python executable is not found in the specified path.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize the <cite>ProtParam</cite> class:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protparam</span> <span class="kn">import</span> <span class="n">ProtParam</span>

<span class="c1"># Initialize the ProtParam class with default settings</span>
<span class="n">protparam</span> <span class="o">=</span> <span class="n">ProtParam</span><span class="p">()</span>

<span class="c1"># Initialize the ProtParam class with a specific job starter</span>
<span class="n">custom_jobstarter</span> <span class="o">=</span> <span class="s2">&quot;my_custom_jobstarter&quot;</span>
<span class="n">protparam</span> <span class="o">=</span> <span class="n">ProtParam</span><span class="p">(</span><span class="n">jobstarter</span><span class="o">=</span><span class="n">custom_jobstarter</span><span class="p">)</span>
</pre></div>
</div>
<p>The <cite>__init__</cite> method ensures that the ProtParam class is ready to perform protein sequence parameter calculations within the ProtFlow framework, setting up the environment and configurations necessary for successful execution.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.metrics.protparam.ProtParam.run">
<span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pH</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><span class="pre">JobStarter</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/protflow/metrics/protparam.html#ProtParam.run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.protparam.ProtParam.run" title="Link to this definition"></a></dt>
<dd><section id="protparam-class">
<h3>ProtParam Class<a class="headerlink" href="#protparam-class" title="Link to this heading"></a></h3>
<p>The <cite>ProtParam</cite> class is a specialized class designed to facilitate the calculation of protein sequence parameters within the ProtFlow framework. It extends the <cite>Runner</cite> class and incorporates specific methods to handle the setup, execution, and data collection associated with ProtParam calculations.</p>
<section id="id7">
<h4>Detailed Description<a class="headerlink" href="#id7" title="Link to this heading"></a></h4>
<p>The <cite>ProtParam</cite> class manages all aspects of running ProtParam calculations. It handles the configuration of necessary scripts and executables, prepares the environment for sequence feature calculations, and executes the ProtParam commands. Additionally, it collects and processes the output data, organizing it into a structured format for further analysis.</p>
<dl class="simple">
<dt>Key functionalities include:</dt><dd><ul class="simple">
<li><p>Setting up paths to ProtParam scripts and Python executables.</p></li>
<li><p>Configuring job starter options, either automatically or manually.</p></li>
<li><p>Handling the execution of ProtParam commands with support for various input types.</p></li>
<li><p>Collecting and processing output data into a pandas DataFrame.</p></li>
<li><p>Customizing the sequence feature calculations based on user-defined parameters such as pH.</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">rtype<span class="colon">:</span></dt>
<dd class="field-odd"><p>An instance of the <cite>ProtParam</cite> class, configured to run ProtParam calculations and handle outputs efficiently.</p>
</dd>
<dt class="field-even">raises - FileNotFoundError<span class="colon">:</span></dt>
<dd class="field-even"><p>If required files or directories are not found during the execution process.:</p>
</dd>
<dt class="field-odd">raises - ValueError<span class="colon">:</span></dt>
<dd class="field-odd"><p>If invalid arguments are provided to the methods.:</p>
</dd>
<dt class="field-even">raises - TypeError<span class="colon">:</span></dt>
<dd class="field-even"><p>If the input poses are not of the expected type.:</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize and use the <cite>ProtParam</cite> class:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">protparam</span> <span class="kn">import</span> <span class="n">ProtParam</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the ProtParam class</span>
<span class="n">protparam</span> <span class="o">=</span> <span class="n">ProtParam</span><span class="p">()</span>

<span class="c1"># Run the ProtParam calculation process</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">protparam</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">seq_col</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">pH</span><span class="o">=</span><span class="mi">7</span><span class="p">,</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id8">
<h4>Further Details<a class="headerlink" href="#id8" title="Link to this heading"></a></h4>
<blockquote>
<div><ul class="simple">
<li><p>Edge Cases: The class includes handling for various edge cases, such as empty pose lists, the need to overwrite previous results, and the presence of existing score files.</p></li>
<li><p>Customization: The class provides extensive customization options through its parameters, allowing users to tailor the ProtParam calculations to their specific needs.</p></li>
<li><p>Integration: Seamlessly integrates with other ProtFlow components, leveraging shared configurations and data structures for a unified workflow.</p></li>
</ul>
</div></blockquote>
<p>The ProtParam class is intended for researchers and developers who need to perform ProtParam calculations as part of their protein design and analysis workflows. It simplifies the process, allowing users to focus on analyzing results and advancing their research.</p>
</section>
</section>
</dd></dl>

</dd></dl>

</section>
<section id="module-protflow.metrics.rmsd">
<span id="protflow-metrics-rmsd-module"></span><h2>protflow.metrics.rmsd module<a class="headerlink" href="#module-protflow.metrics.rmsd" title="Link to this heading"></a></h2>
<section id="rmsd-module">
<h3>RMSD Module<a class="headerlink" href="#rmsd-module" title="Link to this heading"></a></h3>
<p>This module provides the functionality to calculate Root Mean Square Deviation (RMSD) values for protein structures within the ProtFlow framework. It offers tools to run RMSD calculations, handle inputs and outputs, and process the resulting data in a structured and automated manner.</p>
<section id="id9">
<h4>Detailed Description<a class="headerlink" href="#id9" title="Link to this heading"></a></h4>
<p>The <cite>BackboneRMSD</cite> and <cite>MotifRMSD</cite> classes encapsulate the functionality necessary to execute RMSD calculations. These classes manage the configuration of paths to essential scripts and Python executables, set up the environment, and handle the execution of RMSD calculations. They also include methods for collecting and processing output data, ensuring that the results are organized and accessible for further analysis within the ProtFlow ecosystem.</p>
<p>The module is designed to streamline the integration of RMSD calculations into larger computational workflows. It supports the automatic setup of job parameters, execution of RMSD commands, and parsing of output files into a structured DataFrame format. This facilitates subsequent data analysis and visualization steps.</p>
</section>
<section id="id10">
<h4>Usage<a class="headerlink" href="#id10" title="Link to this heading"></a></h4>
<p>To use this module, create an instance of the <cite>BackboneRMSD</cite> or <cite>MotifRMSD</cite> class and invoke their <cite>run</cite> methods with appropriate parameters. The module will handle the configuration, execution, and result collection processes. Detailed control over the RMSD calculation process is provided through various parameters, allowing for customized runs tailored to specific research needs.</p>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize and use the <cite>BackboneRMSD</cite> class within a ProtFlow pipeline:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">rmsd</span> <span class="kn">import</span> <span class="n">BackboneRMSD</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the BackboneRMSD class</span>
<span class="n">backbone_rmsd</span> <span class="o">=</span> <span class="n">BackboneRMSD</span><span class="p">()</span>

<span class="c1"># Run the RMSD calculation</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">backbone_rmsd</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">ref_col</span><span class="o">=</span><span class="s2">&quot;reference&quot;</span><span class="p">,</span>
    <span class="n">chains</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">],</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id11">
<h4>Further Details<a class="headerlink" href="#id11" title="Link to this heading"></a></h4>
<blockquote>
<div><ul class="simple">
<li><p>Edge Cases: The module handles various edge cases, such as empty pose lists and the need to overwrite previous results. It ensures robust error handling and logging for easier debugging and verification of the RMSD calculation process.</p></li>
<li><p>Customizability: Users can customize the RMSD calculation process through multiple parameters, including the specific atoms and chains to be used in the calculation, as well as jobstarter configurations.</p></li>
<li><p>Integration: The module seamlessly integrates with other components of the ProtFlow framework, leveraging shared configurations and data structures to provide a cohesive user experience.</p></li>
</ul>
</div></blockquote>
<p>This module is intended for researchers and developers who need to incorporate RMSD calculations into their protein design and analysis workflows. By automating many of the setup and execution steps, it allows users to focus on interpreting results and advancing their scientific inquiries.</p>
<p class="rubric">Notes</p>
<p>This module is part of the ProtFlow package and is designed to work in tandem with other components of the package, especially those related to job management in HPC environments.</p>
</section>
<section id="id12">
<h4>Author<a class="headerlink" href="#id12" title="Link to this heading"></a></h4>
<p>Markus Braun, Adrian Tripp</p>
</section>
<section id="id13">
<h4>Version<a class="headerlink" href="#id13" title="Link to this heading"></a></h4>
<p>0.1.0</p>
</section>
</section>
<dl class="py class">
<dt class="sig sig-object py" id="protflow.metrics.rmsd.BackboneRMSD">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">protflow.metrics.rmsd.</span></span><span class="sig-name descname"><span class="pre">BackboneRMSD</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">atoms</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">['CA']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chains</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/metrics/rmsd.html#BackboneRMSD"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.rmsd.BackboneRMSD" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="protflow.html#protflow.runners.Runner" title="protflow.runners.Runner"><code class="xref py py-class docutils literal notranslate"><span class="pre">Runner</span></code></a></p>
<section id="backbonermsd-class">
<h3>BackboneRMSD Class<a class="headerlink" href="#backbonermsd-class" title="Link to this heading"></a></h3>
<p>The <cite>BackboneRMSD</cite> class is a specialized class designed to facilitate the calculation of backbone RMSD values within the ProtFlow framework. It extends the <cite>Runner</cite> class and incorporates specific methods to handle the setup, execution, and data collection associated with RMSD calculations.</p>
<section id="id14">
<h4>Detailed Description<a class="headerlink" href="#id14" title="Link to this heading"></a></h4>
<p>The <cite>BackboneRMSD</cite> class manages all aspects of calculating RMSD for protein backbones. It handles the configuration of necessary scripts and executables, prepares the environment for RMSD calculations, and executes the commands. Additionally, it collects and processes the output data, organizing it into a structured format for further analysis.</p>
<dl class="simple">
<dt>Key functionalities include:</dt><dd><ul class="simple">
<li><p>Setting up paths to RMSD calculation scripts and Python executables.</p></li>
<li><p>Configuring job starter options, either automatically or manually.</p></li>
<li><p>Handling the execution of RMSD commands with support for different atoms and chains.</p></li>
<li><p>Collecting and processing output data into a pandas DataFrame.</p></li>
<li><p>Managing overwrite options and handling existing score files.</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">rtype<span class="colon">:</span></dt>
<dd class="field-odd"><p>An instance of the <cite>BackboneRMSD</cite> class, configured to run RMSD calculations and handle outputs efficiently.</p>
</dd>
<dt class="field-even">raises - FileNotFoundError<span class="colon">:</span></dt>
<dd class="field-even"><p>If required files or directories are not found during the execution process.:</p>
</dd>
<dt class="field-odd">raises - ValueError<span class="colon">:</span></dt>
<dd class="field-odd"><p>If invalid arguments are provided to the methods.:</p>
</dd>
<dt class="field-even">raises - TypeError<span class="colon">:</span></dt>
<dd class="field-even"><p>If atoms or chains are not of the expected type.:</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize and use the <cite>BackboneRMSD</cite> class:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">rmsd</span> <span class="kn">import</span> <span class="n">BackboneRMSD</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">LocalJobStarter</span><span class="p">(</span><span class="n">max_cores</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>

<span class="c1"># Initialize the BackboneRMSD class</span>
<span class="n">backbone_rmsd</span> <span class="o">=</span> <span class="n">BackboneRMSD</span><span class="p">()</span>

<span class="c1"># Run the RMSD calculation</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">backbone_rmsd</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">ref_col</span><span class="o">=</span><span class="s2">&quot;reference_location&quot;</span><span class="p">,</span>
    <span class="n">chains</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">],</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id15">
<h4>Further Details<a class="headerlink" href="#id15" title="Link to this heading"></a></h4>
<blockquote>
<div><ul class="simple">
<li><p>Edge Cases: The class includes handling for various edge cases, such as empty pose lists, the need to overwrite previous results, and the presence of existing score files.</p></li>
<li><p>Customization: The class provides extensive customization options through its parameters, allowing users to tailor the RMSD calculation process to their specific needs.</p></li>
<li><p>Integration: Seamlessly integrates with other ProtFlow components, leveraging shared configurations and data structures for a unified workflow.</p></li>
</ul>
</div></blockquote>
<p>The BackboneRMSD class is intended for researchers and developers who need to perform backbone RMSD calculations as part of their protein design and analysis workflows. It simplifies the process, allowing users to focus on analyzing results and advancing their research.</p>
<dl class="py method">
<dt class="sig sig-object py" id="protflow.metrics.rmsd.BackboneRMSD.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">atoms</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">['CA']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chains</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/metrics/rmsd.html#BackboneRMSD.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.rmsd.BackboneRMSD.__init__" title="Link to this definition"></a></dt>
<dd><p>Initialize the BackboneRMSD class.</p>
<p>This constructor sets up the BackboneRMSD instance with default or provided parameters. It configures the reference column, atoms, chains, jobstarter, and overwrite options for RMSD calculations.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ref_col</strong> (<em>str</em><em>, </em><em>optional</em>) – The reference column for RMSD calculations. Defaults to None.</p></li>
<li><p><strong>atoms</strong> (<em>list</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – The list of atom names to calculate RMSD over. Defaults to [“CA”].</p></li>
<li><p><strong>chains</strong> (<em>list</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – The list of chain names to calculate RMSD over. Defaults to None.</p></li>
<li><p><strong>overwrite</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, overwrite existing output files. Defaults to False.</p></li>
<li><p><strong>jobstarter</strong> (<em>str</em><em>, </em><em>optional</em>) – The jobstarter configuration for running the RMSD calculations. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize the BackboneRMSD class:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">rmsd</span> <span class="kn">import</span> <span class="n">BackboneRMSD</span>

<span class="c1"># Initialize the BackboneRMSD class with default parameters</span>
<span class="n">backbone_rmsd</span> <span class="o">=</span> <span class="n">BackboneRMSD</span><span class="p">()</span>

<span class="c1"># Initialize the BackboneRMSD class with custom parameters</span>
<span class="n">backbone_rmsd</span> <span class="o">=</span> <span class="n">BackboneRMSD</span><span class="p">(</span><span class="n">ref_col</span><span class="o">=</span><span class="s2">&quot;reference&quot;</span><span class="p">,</span> <span class="n">atoms</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;CA&quot;</span><span class="p">,</span> <span class="s2">&quot;CB&quot;</span><span class="p">],</span> <span class="n">chains</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">],</span> <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">jobstarter</span><span class="o">=</span><span class="s2">&quot;custom_starter&quot;</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Default Values:</strong> If no parameters are provided, the class initializes with default values suitable for basic RMSD calculations.</p></li>
<li><p><strong>Parameter Storage:</strong> The parameters provided during initialization are stored as instance variables, which are used in subsequent method calls.</p></li>
<li><p><strong>Custom Configuration:</strong> Users can customize the RMSD calculation process by providing specific values for the reference column, atoms, chains, and jobstarter.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.metrics.rmsd.BackboneRMSD.calc_all_atom_rmsd">
<span class="sig-name descname"><span class="pre">calc_all_atom_rmsd</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/protflow/metrics/rmsd.html#BackboneRMSD.calc_all_atom_rmsd"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.rmsd.BackboneRMSD.calc_all_atom_rmsd" title="Link to this definition"></a></dt>
<dd><p>Method to calculate all-atom RMSD between poses</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.metrics.rmsd.BackboneRMSD.run">
<span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><span class="pre">JobStarter</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chains</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/protflow/metrics/rmsd.html#BackboneRMSD.run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.rmsd.BackboneRMSD.run" title="Link to this definition"></a></dt>
<dd><p>Calculate the backbone RMSD for given poses and jobstarter configuration.</p>
<p>This method sets up and runs the RMSD calculation process using the provided poses and jobstarter object. It handles the configuration, execution, and collection of output data, ensuring that the results are organized and accessible for further analysis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>poses</strong> (<a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><em>Poses</em></a>) – The Poses object containing the protein structures.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – A prefix used to name and organize the output files.</p></li>
<li><p><strong>ref_col</strong> (<em>str</em><em>, </em><em>optional</em>) – The reference column for RMSD calculations. Defaults to None.</p></li>
<li><p><strong>jobstarter</strong> (<a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><em>JobStarter</em></a><em>, </em><em>optional</em>) – An instance of the JobStarter class, which manages job execution. Defaults to None.</p></li>
<li><p><strong>chains</strong> (<em>list</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – A list of chain names to calculate RMSD over. Defaults to None.</p></li>
<li><p><strong>overwrite</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, overwrite existing output files. Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An instance of the RunnerOutput class, containing the processed poses and results of the RMSD calculation.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="protflow.html#protflow.runners.RunnerOutput" title="protflow.runners.RunnerOutput">RunnerOutput</a></p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>FileNotFoundError</strong> – If required files or directories are not found during the execution process.</p></li>
<li><p><strong>ValueError</strong> – If invalid arguments are provided to the method.</p></li>
<li><p><strong>TypeError</strong> – If chains are not of the expected type.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>run</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">rmsd</span> <span class="kn">import</span> <span class="n">BackboneRMSD</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">LocalJobStarter</span><span class="p">(</span><span class="n">max_cores</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>

<span class="c1"># Initialize the BackboneRMSD class</span>
<span class="n">backbone_rmsd</span> <span class="o">=</span> <span class="n">BackboneRMSD</span><span class="p">()</span>

<span class="c1"># Run the RMSD calculation</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">backbone_rmsd</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">ref_col</span><span class="o">=</span><span class="s2">&quot;reference&quot;</span><span class="p">,</span>
    <span class="n">chains</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">],</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Setup and Execution:</strong> The method ensures that the environment is correctly set up, directories are prepared, and necessary commands are constructed and executed. It supports splitting poses into sublists for parallel processing.</p></li>
<li><p><strong>Input Handling:</strong> The method prepares input JSON files for each sublist of poses and constructs commands for running RMSD calculations using BioPython.</p></li>
<li><p><strong>Output Management:</strong> The method handles the collection and processing of output data from multiple score files, concatenating them into a single DataFrame and saving the results.</p></li>
<li><p><strong>Customization:</strong> Extensive customization options are provided through parameters, allowing users to tailor the RMSD calculation process to their specific needs, including specifying atoms and chains for RMSD calculations.</p></li>
</ul>
</dd>
</dl>
<p>This method is designed to streamline the execution of backbone RMSD calculations within the ProtFlow framework, making it easier for researchers and developers to perform and analyze RMSD calculations.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.metrics.rmsd.BackboneRMSD.set_atoms">
<span class="sig-name descname"><span class="pre">set_atoms</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">atoms</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/protflow/metrics/rmsd.html#BackboneRMSD.set_atoms"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.rmsd.BackboneRMSD.set_atoms" title="Link to this definition"></a></dt>
<dd><p>Set the atoms for RMSD calculations.</p>
<p>This method sets the list of atom names to calculate RMSD over. If “all” is provided, all atoms will be considered.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>atoms</strong> (<em>list</em><em>[</em><em>str</em><em>]</em>) – The list of atom names to calculate RMSD over.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>TypeError</strong> – If atoms is not a list of strings.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>set_atoms</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">rmsd</span> <span class="kn">import</span> <span class="n">BackboneRMSD</span>

<span class="c1"># Initialize the BackboneRMSD class</span>
<span class="n">backbone_rmsd</span> <span class="o">=</span> <span class="n">BackboneRMSD</span><span class="p">()</span>

<span class="c1"># Set the atoms for RMSD calculation</span>
<span class="n">backbone_rmsd</span><span class="o">.</span><span class="n">set_atoms</span><span class="p">([</span><span class="s2">&quot;CA&quot;</span><span class="p">,</span> <span class="s2">&quot;CB&quot;</span><span class="p">])</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Usage:</strong> The list of atoms specifies which atoms in the protein backbone will be considered during RMSD calculations.</p></li>
<li><p><strong>Validation:</strong> The method includes validation to ensure that the atoms parameter is a list of strings, representing valid atom names.</p></li>
<li><p><strong>Flexibility:</strong> Users can specify any set of atoms or choose to include all atoms by setting the parameter to “all”.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.metrics.rmsd.BackboneRMSD.set_chains">
<span class="sig-name descname"><span class="pre">set_chains</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chains</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/protflow/metrics/rmsd.html#BackboneRMSD.set_chains"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.rmsd.BackboneRMSD.set_chains" title="Link to this definition"></a></dt>
<dd><p>Set the chains for RMSD calculations.</p>
<p>This method sets the list of chain names to calculate RMSD over. It ensures that the provided chains parameter is a list of strings or a single string representing chain names.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>chains</strong> (<em>list</em><em>[</em><em>str</em><em>] or </em><em>str</em>) – The list of chain names or a single chain name to calculate RMSD over.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>TypeError</strong> – If chains is not a list of strings or a single string.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>set_chains</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">rmsd</span> <span class="kn">import</span> <span class="n">BackboneRMSD</span>

<span class="c1"># Initialize the BackboneRMSD class</span>
<span class="n">backbone_rmsd</span> <span class="o">=</span> <span class="n">BackboneRMSD</span><span class="p">()</span>

<span class="c1"># Set the chains for RMSD calculation</span>
<span class="n">backbone_rmsd</span><span class="o">.</span><span class="n">set_chains</span><span class="p">([</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">])</span>

<span class="c1"># Alternatively, set a single chain</span>
<span class="n">backbone_rmsd</span><span class="o">.</span><span class="n">set_chains</span><span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Usage:</strong> The chains parameter specifies which chains in the protein structure will be considered during RMSD calculations.</p></li>
<li><p><strong>Validation:</strong> The method includes validation to ensure that the chains parameter is either a list of strings or a single string, representing valid chain names.</p></li>
<li><p><strong>Flexibility:</strong> Users can specify multiple chains as a list or a single chain as a string, providing flexibility in how the RMSD calculations are configured.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.metrics.rmsd.BackboneRMSD.set_jobstarter">
<span class="sig-name descname"><span class="pre">set_jobstarter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><span class="pre">JobStarter</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/protflow/metrics/rmsd.html#BackboneRMSD.set_jobstarter"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.rmsd.BackboneRMSD.set_jobstarter" title="Link to this definition"></a></dt>
<dd><p>Set the jobstarter configuration for the BackboneRMSD runner.</p>
<p>This method sets the jobstarter configuration to be used in the RMSD calculation process.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>jobstarter</strong> (<a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><em>JobStarter</em></a>) – The jobstarter configuration for running the RMSD calculations.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>TypeError</strong> – If jobstarter is not of type JobStarter.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>set_jobstarter</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">rmsd</span> <span class="kn">import</span> <span class="n">BackboneRMSD</span>

<span class="c1"># Initialize the BackboneRMSD class</span>
<span class="n">backbone_rmsd</span> <span class="o">=</span> <span class="n">BackboneRMSD</span><span class="p">()</span>

<span class="c1"># Set the jobstarter configuration</span>
<span class="n">backbone_rmsd</span><span class="o">.</span><span class="n">set_jobstarter</span><span class="p">(</span><span class="s2">&quot;custom_starter&quot;</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Usage:</strong> The jobstarter configuration specifies how the RMSD calculations will be managed and executed, particularly in HPC environments.</p></li>
<li><p><strong>Validation:</strong> The method includes validation to ensure that the jobstarter parameter is of the correct type.</p></li>
<li><p><strong>Integration:</strong> The jobstarter configuration set by this method is used by other methods in the class to manage the execution of RMSD calculations.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.metrics.rmsd.BackboneRMSD.set_ref_col">
<span class="sig-name descname"><span class="pre">set_ref_col</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/protflow/metrics/rmsd.html#BackboneRMSD.set_ref_col"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.rmsd.BackboneRMSD.set_ref_col" title="Link to this definition"></a></dt>
<dd><p>Set the reference column for RMSD calculations.</p>
<p>This method sets the default reference column to be used in the RMSD calculation process.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ref_col</strong> (<em>str</em>) – The reference column for RMSD calculations.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>TypeError</strong> – If ref_col is not of type string.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>set_ref_col</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">rmsd</span> <span class="kn">import</span> <span class="n">BackboneRMSD</span>

<span class="c1"># Initialize the BackboneRMSD class</span>
<span class="n">backbone_rmsd</span> <span class="o">=</span> <span class="n">BackboneRMSD</span><span class="p">()</span>

<span class="c1"># Set the reference column</span>
<span class="n">backbone_rmsd</span><span class="o">.</span><span class="n">set_ref_col</span><span class="p">(</span><span class="s2">&quot;reference&quot;</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Usage:</strong> The reference column is used to identify which column in the input data contains the reference structures for RMSD calculation.</p></li>
<li><p><strong>Validation:</strong> The method includes validation to ensure that the reference column is of the correct type.</p></li>
<li><p><strong>Integration:</strong> The reference column set by this method is used by other methods in the class to perform RMSD calculations.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
</section>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="protflow.metrics.rmsd.MotifRMSD">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">protflow.metrics.rmsd.</span></span><span class="sig-name descname"><span class="pre">MotifRMSD</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_motif</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_motif</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_chains</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_chains</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><span class="pre">JobStarter</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/metrics/rmsd.html#MotifRMSD"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.rmsd.MotifRMSD" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="protflow.html#protflow.runners.Runner" title="protflow.runners.Runner"><code class="xref py py-class docutils literal notranslate"><span class="pre">Runner</span></code></a></p>
<section id="motifrmsd-class">
<h3>MotifRMSD Class<a class="headerlink" href="#motifrmsd-class" title="Link to this heading"></a></h3>
<p>The <cite>MotifRMSD</cite> class is a specialized class designed to facilitate the calculation of RMSD values for specific motifs within protein structures in the ProtFlow framework. It extends the <cite>Runner</cite> class and incorporates specific methods to handle the setup, execution, and data collection associated with motif-specific RMSD calculations.</p>
<section id="id16">
<h4>Detailed Description<a class="headerlink" href="#id16" title="Link to this heading"></a></h4>
<p>The <cite>MotifRMSD</cite> class manages all aspects of calculating RMSD for specified motifs within protein structures. It handles the configuration of necessary scripts and executables, prepares the environment for RMSD calculations, and executes the commands. Additionally, it collects and processes the output data, organizing it into a structured format for further analysis.</p>
<dl class="simple">
<dt>Key functionalities include:</dt><dd><ul class="simple">
<li><p>Setting up paths to motif RMSD calculation scripts and Python executables.</p></li>
<li><p>Configuring job starter options, either automatically or manually.</p></li>
<li><p>Handling the execution of RMSD commands with support for various motifs and chains.</p></li>
<li><p>Collecting and processing output data into a pandas DataFrame.</p></li>
<li><p>Managing overwrite options and handling existing score files.</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">rtype<span class="colon">:</span></dt>
<dd class="field-odd"><p>An instance of the <cite>MotifRMSD</cite> class, configured to run motif RMSD calculations and handle outputs efficiently.</p>
</dd>
<dt class="field-even">raises - FileNotFoundError<span class="colon">:</span></dt>
<dd class="field-even"><p>If required files or directories are not found during the execution process.:</p>
</dd>
<dt class="field-odd">raises - ValueError<span class="colon">:</span></dt>
<dd class="field-odd"><p>If invalid arguments are provided to the methods.:</p>
</dd>
<dt class="field-even">raises - TypeError<span class="colon">:</span></dt>
<dd class="field-even"><p>If motifs or chains are not of the expected type.:</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize and use the <cite>MotifRMSD</cite> class:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">rmsd</span> <span class="kn">import</span> <span class="n">MotifRMSD</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the MotifRMSD class</span>
<span class="n">motif_rmsd</span> <span class="o">=</span> <span class="n">MotifRMSD</span><span class="p">()</span>

<span class="c1"># Run the motif RMSD calculation</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">motif_rmsd</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_2&quot;</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">ref_col</span><span class="o">=</span><span class="s2">&quot;reference&quot;</span><span class="p">,</span>
    <span class="n">ref_motif</span><span class="o">=</span><span class="s2">&quot;motif_A&quot;</span><span class="p">,</span>
    <span class="n">target_motif</span><span class="o">=</span><span class="s2">&quot;motif_B&quot;</span><span class="p">,</span>
    <span class="n">atoms</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;CA&quot;</span><span class="p">,</span> <span class="s2">&quot;CB&quot;</span><span class="p">],</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id17">
<h4>Further Details<a class="headerlink" href="#id17" title="Link to this heading"></a></h4>
<blockquote>
<div><ul class="simple">
<li><p>Edge Cases: The class includes handling for various edge cases, such as empty pose lists, the need to overwrite previous results, and the presence of existing score files.</p></li>
<li><p>Customization: The class provides extensive customization options through its parameters, allowing users to tailor the motif RMSD calculation process to their specific needs.</p></li>
<li><p>Integration: Seamlessly integrates with other ProtFlow components, leveraging shared configurations and data structures for a unified workflow.</p></li>
</ul>
</div></blockquote>
<p>The MotifRMSD class is intended for researchers and developers who need to perform RMSD calculations for specific motifs as part of their protein design and analysis workflows. It simplifies the process, allowing users to focus on analyzing results and advancing their research.</p>
<dl class="py method">
<dt class="sig sig-object py" id="protflow.metrics.rmsd.MotifRMSD.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_motif</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_motif</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_chains</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_chains</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><span class="pre">JobStarter</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/metrics/rmsd.html#MotifRMSD.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.rmsd.MotifRMSD.__init__" title="Link to this definition"></a></dt>
<dd><p>Initialize the MotifRMSD class.</p>
<p>This constructor sets up the MotifRMSD instance with default or provided parameters. It configures the reference column, target motif, reference motif, target chains, reference chains, jobstarter, and overwrite options for RMSD calculations.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ref_col</strong> (<em>str</em><em>, </em><em>optional</em>) – The reference column for RMSD calculations. Defaults to None.</p></li>
<li><p><strong>target_motif</strong> (<em>str</em><em>, </em><em>optional</em>) – The target motif for RMSD calculations. Defaults to None.</p></li>
<li><p><strong>ref_motif</strong> (<em>str</em><em>, </em><em>optional</em>) – The reference motif for RMSD calculations. Defaults to None.</p></li>
<li><p><strong>target_chains</strong> (<em>list</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – The list of chain names for the target motif. Defaults to None.</p></li>
<li><p><strong>ref_chains</strong> (<em>list</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – The list of chain names for the reference motif. Defaults to None.</p></li>
<li><p><strong>jobstarter</strong> (<a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><em>JobStarter</em></a><em>, </em><em>optional</em>) – The jobstarter configuration for running the RMSD calculations. Defaults to None.</p></li>
<li><p><strong>overwrite</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, overwrite existing output files. Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize the MotifRMSD class:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">rmsd</span> <span class="kn">import</span> <span class="n">MotifRMSD</span>

<span class="c1"># Initialize the MotifRMSD class with default parameters</span>
<span class="n">motif_rmsd</span> <span class="o">=</span> <span class="n">MotifRMSD</span><span class="p">()</span>

<span class="c1"># Initialize the MotifRMSD class with custom parameters</span>
<span class="n">motif_rmsd</span> <span class="o">=</span> <span class="n">MotifRMSD</span><span class="p">(</span>
    <span class="n">ref_col</span><span class="o">=</span><span class="s2">&quot;reference&quot;</span><span class="p">,</span>
    <span class="n">target_motif</span><span class="o">=</span><span class="s2">&quot;motif_A&quot;</span><span class="p">,</span>
    <span class="n">ref_motif</span><span class="o">=</span><span class="s2">&quot;motif_B&quot;</span><span class="p">,</span>
    <span class="n">target_chains</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">],</span>
    <span class="n">ref_chains</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;B&quot;</span><span class="p">],</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">JobStarter</span><span class="p">(),</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Default Values:</strong> If no parameters are provided, the class initializes with default values suitable for basic motif-specific RMSD calculations.</p></li>
<li><p><strong>Parameter Storage:</strong> The parameters provided during initialization are stored as instance variables, which are used in subsequent method calls.</p></li>
<li><p><strong>Custom Configuration:</strong> Users can customize the motif RMSD calculation process by providing specific values for the reference column, target motif, reference motif, target chains, reference chains, jobstarter, and overwrite option.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.metrics.rmsd.MotifRMSD.run">
<span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><span class="pre">JobStarter</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_motif</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Any</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_motif</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Any</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">atoms</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/metrics/rmsd.html#MotifRMSD.run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.rmsd.MotifRMSD.run" title="Link to this definition"></a></dt>
<dd><p>Calculate the motif-specific RMSD for given poses and jobstarter configuration.</p>
<p>This method sets up and runs the motif-specific RMSD calculation process using the provided poses and jobstarter object. It handles the configuration, execution, and collection of output data, ensuring that the results are organized and accessible for further analysis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>poses</strong> (<a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><em>Poses</em></a>) – The Poses object containing the protein structures.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – A prefix used to name and organize the output files.</p></li>
<li><p><strong>jobstarter</strong> (<a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><em>JobStarter</em></a><em>, </em><em>optional</em>) – An instance of the JobStarter class, which manages job execution. Defaults to None.</p></li>
<li><p><strong>ref_col</strong> (<em>str</em><em>, </em><em>optional</em>) – The reference column for RMSD calculations. Defaults to None.</p></li>
<li><p><strong>ref_motif</strong> (<em>Any</em><em>, </em><em>optional</em>) – The reference motif for RMSD calculations. Defaults to None.</p></li>
<li><p><strong>target_motif</strong> (<em>Any</em><em>, </em><em>optional</em>) – The target motif for RMSD calculations. Defaults to None.</p></li>
<li><p><strong>atoms</strong> (<em>list</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – The list of atom names to calculate RMSD over. Defaults to None.</p></li>
<li><p><strong>overwrite</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, overwrite existing output files. Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An instance of the RunnerOutput class, containing the processed poses and results of the RMSD calculation.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="protflow.html#protflow.runners.RunnerOutput" title="protflow.runners.RunnerOutput">RunnerOutput</a></p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>FileNotFoundError</strong> – If required files or directories are not found during the execution process.</p></li>
<li><p><strong>ValueError</strong> – If invalid arguments are provided to the method.</p></li>
<li><p><strong>TypeError</strong> – If motifs or atoms are not of the expected type.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>run</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">rmsd</span> <span class="kn">import</span> <span class="n">MotifRMSD</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the MotifRMSD class</span>
<span class="n">motif_rmsd</span> <span class="o">=</span> <span class="n">MotifRMSD</span><span class="p">()</span>

<span class="c1"># Run the motif RMSD calculation</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">motif_rmsd</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_2&quot;</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">ref_col</span><span class="o">=</span><span class="s2">&quot;reference&quot;</span><span class="p">,</span>
    <span class="n">ref_motif</span><span class="o">=</span><span class="s2">&quot;motif_A&quot;</span><span class="p">,</span>
    <span class="n">target_motif</span><span class="o">=</span><span class="s2">&quot;motif_B&quot;</span><span class="p">,</span>
    <span class="n">atoms</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;CA&quot;</span><span class="p">,</span> <span class="s2">&quot;CB&quot;</span><span class="p">],</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Setup and Execution:</strong> The method ensures that the environment is correctly set up, directories are prepared, and necessary commands are constructed and executed. It supports splitting poses into sublists for parallel processing.</p></li>
<li><p><strong>Input Handling:</strong> The method prepares input JSON files for each sublist of poses and constructs commands for running motif-specific RMSD calculations.</p></li>
<li><p><strong>Output Management:</strong> The method handles the collection and processing of output data from multiple score files, concatenating them into a single DataFrame and saving the results.</p></li>
<li><p><strong>Customization:</strong> Extensive customization options are provided through parameters, allowing users to tailor the motif RMSD calculation process to their specific needs, including specifying reference and target motifs, as well as atoms for RMSD calculations.</p></li>
</ul>
</dd>
</dl>
<p>This method is designed to streamline the execution of motif-specific RMSD calculations within the ProtFlow framework, making it easier for researchers and developers to perform and analyze motif-specific RMSD calculations.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.metrics.rmsd.MotifRMSD.set_jobstarter">
<span class="sig-name descname"><span class="pre">set_jobstarter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/protflow/metrics/rmsd.html#MotifRMSD.set_jobstarter"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.rmsd.MotifRMSD.set_jobstarter" title="Link to this definition"></a></dt>
<dd><p>Set the jobstarter configuration for the MotifRMSD runner.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>jobstarter</strong> (<a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><em>JobStarter</em></a>) – The jobstarter configuration.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If jobstarter is not of type JobStarter.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.metrics.rmsd.MotifRMSD.set_ref_chains">
<span class="sig-name descname"><span class="pre">set_ref_chains</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chains</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/protflow/metrics/rmsd.html#MotifRMSD.set_ref_chains"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.rmsd.MotifRMSD.set_ref_chains" title="Link to this definition"></a></dt>
<dd><p>Set the reference chains for RMSD calculations.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>chains</strong> (<em>list</em><em>[</em><em>str</em><em>]</em>) – The list of reference chain names.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.metrics.rmsd.MotifRMSD.set_ref_col">
<span class="sig-name descname"><span class="pre">set_ref_col</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/protflow/metrics/rmsd.html#MotifRMSD.set_ref_col"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.rmsd.MotifRMSD.set_ref_col" title="Link to this definition"></a></dt>
<dd><p>Set the reference column for RMSD calculations.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>col</strong> (<em>str</em>) – The reference column name.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.metrics.rmsd.MotifRMSD.set_ref_motif">
<span class="sig-name descname"><span class="pre">set_ref_motif</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/protflow/metrics/rmsd.html#MotifRMSD.set_ref_motif"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.rmsd.MotifRMSD.set_ref_motif" title="Link to this definition"></a></dt>
<dd><p>Method to set reference motif. :motif: has to be string and should be a column name in poses.df that will be passed to the .run() function</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.metrics.rmsd.MotifRMSD.set_target_chains">
<span class="sig-name descname"><span class="pre">set_target_chains</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chains</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/protflow/metrics/rmsd.html#MotifRMSD.set_target_chains"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.rmsd.MotifRMSD.set_target_chains" title="Link to this definition"></a></dt>
<dd><p>Set the target chains for RMSD calculations.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>chains</strong> (<em>list</em><em>[</em><em>str</em><em>]</em>) – The list of target chain names.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.metrics.rmsd.MotifRMSD.set_target_motif">
<span class="sig-name descname"><span class="pre">set_target_motif</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/protflow/metrics/rmsd.html#MotifRMSD.set_target_motif"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.rmsd.MotifRMSD.set_target_motif" title="Link to this definition"></a></dt>
<dd><p>Method to set target motif. :motif: has to be string and should be a column name in poses.df that will be passed to the .run() function</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.metrics.rmsd.MotifRMSD.setup_input_dict">
<span class="sig-name descname"><span class="pre">setup_input_dict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_motif</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Any</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_motif</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Any</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span></span></span><a class="reference internal" href="_modules/protflow/metrics/rmsd.html#MotifRMSD.setup_input_dict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.rmsd.MotifRMSD.setup_input_dict" title="Link to this definition"></a></dt>
<dd><p>Set up the input dictionary for motif RMSD calculations.</p>
<p>This method prepares a dictionary that can be written to a JSON file and used as input for the motif RMSD calculation script. The dictionary contains mappings of poses to reference PDB files, target motifs, and reference motifs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>poses</strong> (<a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><em>Poses</em></a>) – The Poses object containing the protein structures.</p></li>
<li><p><strong>ref_col</strong> (<em>str</em>) – The reference column for RMSD calculations.</p></li>
<li><p><strong>ref_motif</strong> (<em>Any</em><em>, </em><em>optional</em>) – The reference motif for RMSD calculations. Defaults to None.</p></li>
<li><p><strong>target_motif</strong> (<em>Any</em><em>, </em><em>optional</em>) – The target motif for RMSD calculations. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dictionary structured for input to the motif RMSD calculation script.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>TypeError</strong> – If ref_motif or target_motif is not of the expected type.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>setup_input_dict</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">rmsd</span> <span class="kn">import</span> <span class="n">MotifRMSD</span>
<span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>

<span class="c1"># Initialize the MotifRMSD class</span>
<span class="n">motif_rmsd</span> <span class="o">=</span> <span class="n">MotifRMSD</span><span class="p">()</span>

<span class="c1"># Create a Poses object</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>

<span class="c1"># Set up the input dictionary for RMSD calculations</span>
<span class="n">input_dict</span> <span class="o">=</span> <span class="n">motif_rmsd</span><span class="o">.</span><span class="n">setup_input_dict</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">ref_col</span><span class="o">=</span><span class="s2">&quot;reference&quot;</span><span class="p">,</span>
    <span class="n">ref_motif</span><span class="o">=</span><span class="s2">&quot;motif_A&quot;</span><span class="p">,</span>
    <span class="n">target_motif</span><span class="o">=</span><span class="s2">&quot;motif_B&quot;</span>
<span class="p">)</span>

<span class="c1"># Print the input dictionary</span>
<span class="nb">print</span><span class="p">(</span><span class="n">input_dict</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Dictionary Structure:</strong> The input dictionary maps each pose to its reference PDB file, target motif, and reference motif.</p></li>
<li><p><strong>Parameter Handling:</strong> The method handles different types of inputs for motifs, ensuring that they are correctly formatted for the RMSD calculation script.</p></li>
<li><p><strong>Integration:</strong> The input dictionary prepared by this method is used by the <cite>run</cite> method to execute motif RMSD calculations.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
</section>
</dd></dl>

</section>
<section id="module-protflow.metrics.tmscore">
<span id="protflow-metrics-tmscore-module"></span><h2>protflow.metrics.tmscore module<a class="headerlink" href="#module-protflow.metrics.tmscore" title="Link to this heading"></a></h2>
<section id="tmscore-module">
<h3>TMscore Module<a class="headerlink" href="#tmscore-module" title="Link to this heading"></a></h3>
<p>This module provides the functionality to integrate TMscore calculations within the ProtFlow framework. It offers tools to run TMscore and TMalign, handle their inputs and outputs, and process the resulting data in a structured and automated manner.</p>
<section id="id18">
<h4>Detailed Description<a class="headerlink" href="#id18" title="Link to this heading"></a></h4>
<p>The <cite>TMalign</cite> and <cite>TMscore</cite> classes encapsulate the functionality necessary to execute TM-align and TM-score runs, respectively. These classes manage the configuration of paths to essential scripts and Python executables, set up the environment, and handle the execution of scoring processes. They include methods for collecting and processing output data, ensuring that the results are organized and accessible for further analysis within the ProtFlow ecosystem.</p>
<p>The module is designed to streamline the integration of TM-align and TM-score into larger computational workflows. It supports the automatic setup of job parameters, execution of TM-align/TM-score commands, and parsing of output files into a structured DataFrame format. This facilitates subsequent data analysis and visualization steps.</p>
</section>
<section id="id19">
<h4>Usage<a class="headerlink" href="#id19" title="Link to this heading"></a></h4>
<p>To use this module, create an instance of the <cite>TMalign</cite> or <cite>TMscore</cite> class and invoke their <cite>run</cite> method with appropriate parameters. The module will handle the configuration, execution, and result collection processes. Detailed control over the scoring process is provided through various parameters, allowing for customized runs tailored to specific research needs.</p>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize and use the <cite>TMalign</cite> class within a ProtFlow pipeline:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">tmscore</span> <span class="kn">import</span> <span class="n">TMalign</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the TMalign class</span>
<span class="n">tmalign</span> <span class="o">=</span> <span class="n">TMalign</span><span class="p">()</span>

<span class="c1"># Run the alignment process</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">tmalign</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">ref_col</span><span class="o">=</span><span class="s2">&quot;reference_pdb&quot;</span><span class="p">,</span>
    <span class="n">sc_tm_score</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;-a&quot;</span><span class="p">,</span>
    <span class="n">pose_options</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;-b&quot;</span><span class="p">],</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
<p>Here is an example of how to initialize and use the <cite>TMscore</cite> class within a ProtFlow pipeline:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">tmscore</span> <span class="kn">import</span> <span class="n">TMscore</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the TMscore class</span>
<span class="n">tmscore</span> <span class="o">=</span> <span class="n">TMscore</span><span class="p">()</span>

<span class="c1"># Run the scoring process</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">tmscore</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_2&quot;</span><span class="p">,</span>
    <span class="n">ref_col</span><span class="o">=</span><span class="s2">&quot;reference_pdb&quot;</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;-c&quot;</span><span class="p">,</span>
    <span class="n">pose_options</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;-d&quot;</span><span class="p">],</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id20">
<h4>Further Details<a class="headerlink" href="#id20" title="Link to this heading"></a></h4>
<blockquote>
<div><ul class="simple">
<li><p>Edge Cases: The module handles various edge cases, such as empty pose lists and the need to overwrite previous results. It ensures robust error handling and logging for easier debugging and verification of the scoring process.</p></li>
<li><p>Customizability: Users can customize the scoring process through multiple parameters, including specific options for the TM-align or TM-score scripts, and options for handling pose-specific parameters.</p></li>
<li><p>Integration: The module seamlessly integrates with other components of the ProtFlow framework, leveraging shared configurations and data structures to provide a cohesive user experience.</p></li>
</ul>
</div></blockquote>
<p>This module is intended for researchers and developers who need to incorporate TM-align or TM-score into their protein structure comparison and analysis workflows. By automating many of the setup and execution steps, it allows users to focus on interpreting results and advancing their scientific inquiries.</p>
<p class="rubric">Notes</p>
<p>This module is part of the ProtFlow package and is designed to work in tandem with other components of the package, especially those related to job management in HPC environments.</p>
</section>
<section id="id21">
<h4>Author<a class="headerlink" href="#id21" title="Link to this heading"></a></h4>
<p>Markus Braun, Adrian Tripp</p>
</section>
<section id="id22">
<h4>Version<a class="headerlink" href="#id22" title="Link to this heading"></a></h4>
<p>0.1.0</p>
</section>
</section>
<dl class="py class">
<dt class="sig sig-object py" id="protflow.metrics.tmscore.TMalign">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">protflow.metrics.tmscore.</span></span><span class="sig-name descname"><span class="pre">TMalign</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><span class="pre">JobStarter</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">application</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/metrics/tmscore.html#TMalign"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.tmscore.TMalign" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="protflow.html#protflow.runners.Runner" title="protflow.runners.Runner"><code class="xref py py-class docutils literal notranslate"><span class="pre">Runner</span></code></a></p>
<section id="tmalign-class">
<h3>TMalign Class<a class="headerlink" href="#tmalign-class" title="Link to this heading"></a></h3>
<p>The <cite>TMalign</cite> class is a specialized class designed to facilitate the execution of TMalign within the ProtFlow framework. It extends the <cite>Runner</cite> class and incorporates specific methods to handle the setup, execution, and data collection associated with TMalign processes.</p>
<section id="id23">
<h4>Detailed Description<a class="headerlink" href="#id23" title="Link to this heading"></a></h4>
<p>The <cite>TMalign</cite> class manages all aspects of running TMalign simulations. It handles the configuration of necessary scripts and executables, prepares the environment for alignment processes, and executes the alignment commands. Additionally, it collects and processes the output data, organizing it into a structured format for further analysis.</p>
<dl class="simple">
<dt>Key functionalities include:</dt><dd><ul class="simple">
<li><p>Setting up paths to TMalign executables.</p></li>
<li><p>Configuring job starter options, either automatically or manually.</p></li>
<li><p>Handling the execution of TMalign commands with support for various alignment options.</p></li>
<li><p>Collecting and processing output data into a pandas DataFrame.</p></li>
<li><p>Normalizing TM scores based on the reference structure and calculating self-consistency scores.</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">rtype<span class="colon">:</span></dt>
<dd class="field-odd"><p>An instance of the <cite>TMalign</cite> class, configured to run TMalign processes and handle outputs efficiently.</p>
</dd>
<dt class="field-even">raises - FileNotFoundError<span class="colon">:</span></dt>
<dd class="field-even"><p>If the TMalign executable is not found in the specified environment.:</p>
</dd>
<dt class="field-odd">raises - ValueError<span class="colon">:</span></dt>
<dd class="field-odd"><p>If invalid arguments are provided to the methods or if required reference columns are missing.:</p>
</dd>
<dt class="field-even">raises - RuntimeError<span class="colon">:</span></dt>
<dd class="field-even"><p>If no TM scores are found in the output files.:</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize and use the <cite>TMalign</cite> class:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">tmscore</span> <span class="kn">import</span> <span class="n">TMalign</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the TMalign class</span>
<span class="n">tmalign</span> <span class="o">=</span> <span class="n">TMalign</span><span class="p">()</span>

<span class="c1"># Run the alignment process</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">tmalign</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">ref_col</span><span class="o">=</span><span class="s2">&quot;reference_pdb&quot;</span><span class="p">,</span>
    <span class="n">sc_tm_score</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;-a&quot;</span><span class="p">,</span>
    <span class="n">pose_options</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;-b&quot;</span><span class="p">],</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id24">
<h4>Further Details<a class="headerlink" href="#id24" title="Link to this heading"></a></h4>
<blockquote>
<div><ul class="simple">
<li><p>Edge Cases: The class includes handling for various edge cases, such as empty pose lists, the need to overwrite previous results, and the presence of existing score files.</p></li>
<li><p>Customization: The class provides extensive customization options through its parameters, allowing users to tailor the alignment process to their specific needs.</p></li>
<li><p>Integration: Seamlessly integrates with other ProtFlow components, leveraging shared configurations and data structures for a unified workflow.</p></li>
</ul>
</div></blockquote>
</section>
<section id="difference-between-tmscore-and-tmalign">
<h4>Difference Between TMscore and TMalign<a class="headerlink" href="#difference-between-tmscore-and-tmalign" title="Link to this heading"></a></h4>
<ul class="simple">
<li><p><strong>TMscore</strong>: This class calculates the TM-score between protein structures without superimposing them. It is suitable for comparing the overall similarity of protein structures in a sequence-length independent manner. TMscore is used when you need to score the structural similarity directly without modifying the positions of the structures.</p></li>
<li><p><strong>TMalign</strong>: This class not only calculates the TM-score but also superimposes the structures before scoring. It is used when structural alignment and superimposition are necessary to get a more accurate measure of structural similarity, considering the spatial arrangement of the protein structures.</p></li>
</ul>
<p>The TMalign class is intended for researchers and developers who need to perform TMalign alignments as part of their protein structure comparison and analysis workflows. It simplifies the process, allowing users to focus on analyzing results and advancing their research.</p>
<dl class="py method">
<dt class="sig sig-object py" id="protflow.metrics.tmscore.TMalign.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><span class="pre">JobStarter</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">application</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/metrics/tmscore.html#TMalign.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.tmscore.TMalign.__init__" title="Link to this definition"></a></dt>
<dd><p>Initialize the TMalign class with optional jobstarter and application path.</p>
<p>This method sets up the TMalign class by configuring the jobstarter and the path to the TMalign executable. It ensures that the necessary components are ready for executing TMalign processes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>jobstarter</strong> (<a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><em>JobStarter</em></a><em>, </em><em>optional</em>) – An optional jobstarter configuration. Defaults to None.</p></li>
<li><p><strong>application</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the TMalign executable. If not provided, it defaults to the TMalign executable in the ProtFlow environment.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the TMalign executable is not found in the specified environment.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize the <cite>TMalign</cite> class:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">tmscore</span> <span class="kn">import</span> <span class="n">TMalign</span>

<span class="c1"># Initialize the TMalign class</span>
<span class="n">tmalign</span> <span class="o">=</span> <span class="n">TMalign</span><span class="p">(</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">LocalJobStarter</span><span class="p">(</span><span class="n">max_cores</span><span class="o">=</span><span class="mi">4</span><span class="p">),</span>
    <span class="n">application</span><span class="o">=</span><span class="s2">&quot;/path/to/TMalign&quot;</span>
<span class="p">)</span>

<span class="c1"># Check the instance</span>
<span class="nb">print</span><span class="p">(</span><span class="n">tmalign</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Jobstarter Configuration:</strong> This parameter allows setting up the jobstarter for managing job execution.</p></li>
<li><p><strong>Application Path:</strong> This parameter sets the path to the TMalign executable, ensuring the correct executable is used for alignment processes.</p></li>
</ul>
</dd>
</dl>
<p>This method is designed to prepare the TMalign class for executing TMalign processes, ensuring that all necessary configurations are in place.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.metrics.tmscore.TMalign.collect_scores">
<span class="sig-name descname"><span class="pre">collect_scores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/protflow/metrics/tmscore.html#TMalign.collect_scores"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.tmscore.TMalign.collect_scores" title="Link to this definition"></a></dt>
<dd><p>Collect scores from TMalign output files.</p>
<p>This method collects and processes the scores from the output files generated by TMalign. It reads the scores, extracts relevant information, and organizes the data into a structured pandas DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>output_dir</strong> (<em>str</em>) – The directory where TMalign output files are located.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A DataFrame containing the collected scores.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>RuntimeError</strong> – If no TM scores are found in the output files.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>collect_scores</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">tmscore</span> <span class="kn">import</span> <span class="n">TMalign</span>

<span class="c1"># Initialize the TMalign class</span>
<span class="n">tmalign</span> <span class="o">=</span> <span class="n">TMalign</span><span class="p">()</span>

<span class="c1"># Collect scores</span>
<span class="n">scores_df</span> <span class="o">=</span> <span class="n">tmalign</span><span class="o">.</span><span class="n">collect_scores</span><span class="p">(</span>
    <span class="n">output_dir</span><span class="o">=</span><span class="s2">&quot;output/&quot;</span>
<span class="p">)</span>

<span class="c1"># Print the scores DataFrame</span>
<span class="nb">print</span><span class="p">(</span><span class="n">scores_df</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Score Extraction:</strong> The method reads the output files, extracts relevant scores, and organizes them into a pandas DataFrame.</p></li>
<li><p><strong>Validation:</strong> Ensures that the scores are correctly extracted and that no errors occurred during the process.</p></li>
</ul>
</dd>
</dl>
<p>This method is designed to streamline the collection and processing of scores from TMalign output files, ensuring that all relevant data is accurately captured and organized.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.metrics.tmscore.TMalign.prep_ref">
<span class="sig-name descname"><span class="pre">prep_ref</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ref</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/protflow/metrics/tmscore.html#TMalign.prep_ref"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.tmscore.TMalign.prep_ref" title="Link to this definition"></a></dt>
<dd><p>Prepare the reference structures for TMalign.</p>
<p>This method prepares the reference structures for alignment based on the provided reference column or specific PDB file. It ensures that the references are correctly formatted for the TMalign process.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ref</strong> (<em>str</em>) – The reference structure, either as a path to a PDB file or as a column name in the Poses DataFrame.</p></li>
<li><p><strong>poses</strong> (<a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><em>Poses</em></a>) – The Poses object containing the protein structures.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A list of reference paths for each pose.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[str]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the ref parameter is not a string or if the reference column is missing from the Poses DataFrame.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>prep_ref</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">tmscore</span> <span class="kn">import</span> <span class="n">TMalign</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">tmalign</span> <span class="o">=</span> <span class="n">TMalign</span><span class="p">()</span>

<span class="c1"># Prepare reference structures</span>
<span class="n">ref_list</span> <span class="o">=</span> <span class="n">tmalign</span><span class="o">.</span><span class="n">prep_ref</span><span class="p">(</span>
    <span class="n">ref</span><span class="o">=</span><span class="s2">&quot;reference_pdb&quot;</span><span class="p">,</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span>
<span class="p">)</span>

<span class="c1"># Print the reference list</span>
<span class="nb">print</span><span class="p">(</span><span class="n">ref_list</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Reference Handling:</strong> The method can handle both a single PDB file and a column name referring to multiple PDB files within the Poses DataFrame.</p></li>
<li><p><strong>Validation:</strong> Ensures that the provided reference is valid and exists in the Poses DataFrame if specified as a column name.</p></li>
</ul>
</dd>
</dl>
<p>This method is designed to streamline the preparation of reference structures for TMalign processes, ensuring that all references are correctly formatted and validated.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.metrics.tmscore.TMalign.run">
<span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sc_tm_score</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pose_options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><span class="pre">JobStarter</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/protflow/metrics/tmscore.html#TMalign.run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.tmscore.TMalign.run" title="Link to this definition"></a></dt>
<dd><p>Execute the TMalign process with given poses and jobstarter configuration.</p>
<p>This method sets up and runs the TMalign process using the provided poses and jobstarter object. It handles the configuration, execution, and collection of output data, ensuring that the results are organized and accessible for further analysis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>poses</strong> (<a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><em>Poses</em></a>) – The Poses object containing the protein structures.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – A prefix used to name and organize the output files.</p></li>
<li><p><strong>ref_col</strong> (<em>str</em>) – Column containing paths to PDB files used as reference for TM score calculation.</p></li>
<li><p><strong>sc_tm_score</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, calculates the self-consistency TM score for each backbone in ref_col and adds it into the column {prefix}_sc_tm. Defaults to True.</p></li>
<li><p><strong>options</strong> (<em>str</em><em>, </em><em>optional</em>) – Additional command-line options for the TMalign script. Defaults to None.</p></li>
<li><p><strong>pose_options</strong> (<em>str</em><em>, </em><em>optional</em>) – Name of poses.df column containing options for TMalign. Defaults to None.</p></li>
<li><p><strong>overwrite</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, overwrite existing output files. Defaults to False.</p></li>
<li><p><strong>jobstarter</strong> (<a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><em>JobStarter</em></a><em>, </em><em>optional</em>) – An instance of the JobStarter class, which manages job execution. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An instance of the RunnerOutput class, containing the processed poses and results of the TMalign process.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="protflow.html#protflow.runners.RunnerOutput" title="protflow.runners.RunnerOutput">RunnerOutput</a></p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>FileNotFoundError</strong> – If the TMalign executable is not found in the specified environment.</p></li>
<li><p><strong>ValueError</strong> – If invalid arguments are provided to the method or if required reference columns are missing.</p></li>
<li><p><strong>RuntimeError</strong> – If no TM scores are found in the output files.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>run</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">tmscore</span> <span class="kn">import</span> <span class="n">TMalign</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the TMalign class</span>
<span class="n">tmalign</span> <span class="o">=</span> <span class="n">TMalign</span><span class="p">()</span>

<span class="c1"># Run the alignment process</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">tmalign</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">ref_col</span><span class="o">=</span><span class="s2">&quot;reference_pdb&quot;</span><span class="p">,</span>
    <span class="n">sc_tm_score</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;-a&quot;</span><span class="p">,</span>
    <span class="n">pose_options</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;-b&quot;</span><span class="p">],</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Setup and Execution:</strong> The method ensures that the environment is correctly set up, directories are prepared, and necessary commands are constructed and executed.</p></li>
<li><p><strong>Reference Preparation:</strong> The method prepares the reference structures for alignment based on the provided reference column or specific PDB file.</p></li>
<li><p><strong>Output Management:</strong> The method handles the collection and processing of output data, including merging and normalizing TM scores, ensuring that results are organized and accessible for further analysis.</p></li>
<li><p><strong>Customization:</strong> Extensive customization options are provided through parameters, allowing users to tailor the alignment process to their specific needs.</p></li>
</ul>
</dd>
</dl>
<p>This method is designed to streamline the execution of TMalign processes within the ProtFlow framework, making it easier for researchers and developers to perform and analyze protein structure alignments.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.metrics.tmscore.TMalign.write_cmd">
<span class="sig-name descname"><span class="pre">write_cmd</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pose_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pose_options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">str</span></span></span><a class="reference internal" href="_modules/protflow/metrics/tmscore.html#TMalign.write_cmd"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.tmscore.TMalign.write_cmd" title="Link to this definition"></a></dt>
<dd><p>Write the command to run TMalign.</p>
<p>This method constructs the command to execute TMalign based on the provided parameters. It formats the options and flags correctly and sets up the command to be run in the environment.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pose_path</strong> (<em>str</em>) – The path to the pose file.</p></li>
<li><p><strong>ref_path</strong> (<em>str</em>) – The path to the reference file.</p></li>
<li><p><strong>output_dir</strong> (<em>str</em>) – The directory where output files will be saved.</p></li>
<li><p><strong>options</strong> (<em>str</em><em>, </em><em>optional</em>) – Additional command-line options for TMalign. Defaults to None.</p></li>
<li><p><strong>pose_options</strong> (<em>str</em><em>, </em><em>optional</em>) – Pose-specific options for TMalign. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The constructed command string to run TMalign.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>write_cmd</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">tmscore</span> <span class="kn">import</span> <span class="n">TMalign</span>

<span class="c1"># Initialize the TMalign class</span>
<span class="n">tmalign</span> <span class="o">=</span> <span class="n">TMalign</span><span class="p">()</span>

<span class="c1"># Write the command</span>
<span class="n">cmd</span> <span class="o">=</span> <span class="n">tmalign</span><span class="o">.</span><span class="n">write_cmd</span><span class="p">(</span>
    <span class="n">pose_path</span><span class="o">=</span><span class="s2">&quot;pose.pdb&quot;</span><span class="p">,</span>
    <span class="n">ref_path</span><span class="o">=</span><span class="s2">&quot;reference.pdb&quot;</span><span class="p">,</span>
    <span class="n">output_dir</span><span class="o">=</span><span class="s2">&quot;output/&quot;</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;-a&quot;</span><span class="p">,</span>
    <span class="n">pose_options</span><span class="o">=</span><span class="s2">&quot;-b&quot;</span>
<span class="p">)</span>

<span class="c1"># Print the command</span>
<span class="nb">print</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Command Construction:</strong> The method constructs the command string by parsing and formatting the provided options and pose-specific options.</p></li>
<li><p><strong>Output Management:</strong> Ensures that the output files are correctly named and saved in the specified directory.</p></li>
</ul>
</dd>
</dl>
<p>This method is designed to streamline the construction of commands for TMalign processes, ensuring that all necessary options are correctly formatted and included.</p>
</dd></dl>

</section>
</section>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="protflow.metrics.tmscore.TMscore">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">protflow.metrics.tmscore.</span></span><span class="sig-name descname"><span class="pre">TMscore</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">application</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/metrics/tmscore.html#TMscore"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.tmscore.TMscore" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="protflow.html#protflow.runners.Runner" title="protflow.runners.Runner"><code class="xref py py-class docutils literal notranslate"><span class="pre">Runner</span></code></a></p>
<section id="tmscore-class">
<h3>TMscore Class<a class="headerlink" href="#tmscore-class" title="Link to this heading"></a></h3>
<p>The <cite>TMscore</cite> class is a specialized class designed to facilitate the execution of TMscore within the ProtFlow framework. It extends the <cite>Runner</cite> class and incorporates specific methods to handle the setup, execution, and data collection associated with TMscore processes.</p>
<section id="id25">
<h4>Detailed Description<a class="headerlink" href="#id25" title="Link to this heading"></a></h4>
<p>The <cite>TMscore</cite> class manages all aspects of running TMscore simulations. It handles the configuration of necessary scripts and executables, prepares the environment for scoring processes, and executes the scoring commands. Additionally, it collects and processes the output data, organizing it into a structured format for further analysis.</p>
<dl class="simple">
<dt>Key functionalities include:</dt><dd><ul class="simple">
<li><p>Setting up paths to TMscore executables.</p></li>
<li><p>Configuring job starter options, either automatically or manually.</p></li>
<li><p>Handling the execution of TMscore commands with support for various scoring options.</p></li>
<li><p>Collecting and processing output data into a pandas DataFrame.</p></li>
</ul>
</dd>
</dl>
</section>
<section id="id26">
<h4>Difference Between TMscore and TMalign<a class="headerlink" href="#id26" title="Link to this heading"></a></h4>
<ul class="simple">
<li><p><strong>TMscore</strong>: This class calculates the TM-score between protein structures without superimposing them. It is suitable for comparing the overall similarity of protein structures in a sequence-length independent manner. TMscore is used when you need to score the structural similarity directly without modifying the positions of the structures.</p></li>
<li><p><strong>TMalign</strong>: This class not only calculates the TM-score but also superimposes the structures before scoring. It is used when structural alignment and superimposition are necessary to get a more accurate measure of structural similarity, considering the spatial arrangement of the protein structures.</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">rtype<span class="colon">:</span></dt>
<dd class="field-odd"><p>An instance of the <cite>TMscore</cite> class, configured to run TMscore processes and handle outputs efficiently.</p>
</dd>
<dt class="field-even">raises - FileNotFoundError<span class="colon">:</span></dt>
<dd class="field-even"><p>If the TMscore executable is not found in the specified environment.:</p>
</dd>
<dt class="field-odd">raises - ValueError<span class="colon">:</span></dt>
<dd class="field-odd"><p>If invalid arguments are provided to the methods or if required reference columns are missing.:</p>
</dd>
<dt class="field-even">raises - RuntimeError<span class="colon">:</span></dt>
<dd class="field-even"><p>If no TM scores are found in the output files.:</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize and use the <cite>TMscore</cite> class:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">tmscore</span> <span class="kn">import</span> <span class="n">TMscore</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the TMscore class</span>
<span class="n">tmscore</span> <span class="o">=</span> <span class="n">TMscore</span><span class="p">()</span>

<span class="c1"># Run the scoring process</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">tmscore</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_2&quot;</span><span class="p">,</span>
    <span class="n">ref_col</span><span class="o">=</span><span class="s2">&quot;reference_pdb&quot;</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;-c&quot;</span><span class="p">,</span>
    <span class="n">pose_options</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;-d&quot;</span><span class="p">],</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id27">
<h4>Further Details<a class="headerlink" href="#id27" title="Link to this heading"></a></h4>
<blockquote>
<div><ul class="simple">
<li><p>Edge Cases: The class includes handling for various edge cases, such as empty pose lists, the need to overwrite previous results, and the presence of existing score files.</p></li>
<li><p>Customization: The class provides extensive customization options through its parameters, allowing users to tailor the scoring process to their specific needs.</p></li>
<li><p>Integration: Seamlessly integrates with other ProtFlow components, leveraging shared configurations and data structures for a unified workflow.</p></li>
</ul>
</div></blockquote>
<p>The TMscore class is intended for researchers and developers who need to perform TMscore calculations as part of their protein structure comparison and analysis workflows. It simplifies the process, allowing users to focus on analyzing results and advancing their research.</p>
<dl class="py method">
<dt class="sig sig-object py" id="protflow.metrics.tmscore.TMscore.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">application</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/metrics/tmscore.html#TMscore.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.tmscore.TMscore.__init__" title="Link to this definition"></a></dt>
<dd><p>Initialize the TMscore class with optional jobstarter and application path.</p>
<p>This method sets up the TMscore class by configuring the jobstarter and the path to the TMscore executable. It ensures that the necessary components are ready for executing TMscore processes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>jobstarter</strong> (<em>str</em><em>, </em><em>optional</em>) – An optional jobstarter configuration. Defaults to None.</p></li>
<li><p><strong>application</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to the TMscore executable. If not provided, it defaults to the TMscore executable in the ProtFlow environment.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize the <cite>TMscore</cite> class:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">tmscore</span> <span class="kn">import</span> <span class="n">TMscore</span>

<span class="c1"># Initialize the TMscore class</span>
<span class="n">tmscore</span> <span class="o">=</span> <span class="n">TMscore</span><span class="p">(</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="s2">&quot;local&quot;</span><span class="p">,</span>
    <span class="n">application</span><span class="o">=</span><span class="s2">&quot;/path/to/TMscore&quot;</span>
<span class="p">)</span>

<span class="c1"># Check the instance</span>
<span class="nb">print</span><span class="p">(</span><span class="n">tmscore</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Jobstarter Configuration:</strong> This parameter allows setting up the jobstarter for managing job execution.</p></li>
<li><p><strong>Application Path:</strong> This parameter sets the path to the TMscore executable, ensuring the correct executable is used for scoring processes.</p></li>
</ul>
</dd>
</dl>
<p>This method is designed to prepare the TMscore class for executing TMscore processes, ensuring that all necessary configurations are in place.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.metrics.tmscore.TMscore.collect_scores">
<span class="sig-name descname"><span class="pre">collect_scores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/protflow/metrics/tmscore.html#TMscore.collect_scores"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.tmscore.TMscore.collect_scores" title="Link to this definition"></a></dt>
<dd><p>Collect scores from TMscore output files.</p>
<p>This method collects and processes the scores from the output files generated TMscore. It reads the scores, extracts relevant information, and organizes the data into a structured pandas DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>output_dir</strong> (<em>str</em>) – The directory where TMscore output files are located.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A DataFrame containing the collected scores.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>RuntimeError</strong> – If no TM scores are found in the output files.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>collect_scores</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">tmscore</span> <span class="kn">import</span> <span class="n">TMscore</span>

<span class="c1"># Initialize the TMscore class</span>
<span class="n">tmalign</span> <span class="o">=</span> <span class="n">TMscore</span><span class="p">()</span>

<span class="c1"># Collect scores</span>
<span class="n">scores_df</span> <span class="o">=</span> <span class="n">tmscore</span><span class="o">.</span><span class="n">collect_scores</span><span class="p">(</span>
    <span class="n">output_dir</span><span class="o">=</span><span class="s2">&quot;output/&quot;</span>
<span class="p">)</span>

<span class="c1"># Print the scores DataFrame</span>
<span class="nb">print</span><span class="p">(</span><span class="n">scores_df</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Score Extraction:</strong> The method reads the output files, extracts relevant scores, and organizes them into a pandas DataFrame.</p></li>
<li><p><strong>Validation:</strong> Ensures that the scores are correctly extracted and that no errors occurred during the process.</p></li>
</ul>
</dd>
</dl>
<p>This method is designed to streamline the collection and processing of scores from TMscore output files, ensuring that all relevant data is accurately captured and organized.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.metrics.tmscore.TMscore.run">
<span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pose_options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><span class="pre">JobStarter</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/protflow/metrics/tmscore.html#TMscore.run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.tmscore.TMscore.run" title="Link to this definition"></a></dt>
<dd><p>Execute the TMscore process with given poses and jobstarter configuration.</p>
<p>This method sets up and runs the TMscore process using the provided poses and jobstarter object. It handles the configuration, execution, and collection of output data, ensuring that the results are organized and accessible for further analysis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>poses</strong> (<a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><em>Poses</em></a>) – The Poses object containing the protein structures.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – A prefix used to name and organize the output files.</p></li>
<li><p><strong>ref_col</strong> (<em>str</em>) – Column containing paths to PDB files used as reference for TM score calculation.</p></li>
<li><p><strong>options</strong> (<em>str</em><em>, </em><em>optional</em>) – Additional command-line options for the TMscore script. Defaults to None.</p></li>
<li><p><strong>pose_options</strong> (<em>str</em><em>, </em><em>optional</em>) – Name of poses.df column containing options for TMscore. Defaults to None.</p></li>
<li><p><strong>overwrite</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, overwrite existing output files. Defaults to False.</p></li>
<li><p><strong>jobstarter</strong> (<a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><em>JobStarter</em></a><em>, </em><em>optional</em>) – An instance of the JobStarter class, which manages job execution. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An instance of the RunnerOutput class, containing the processed poses and results of the TMscore process.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="protflow.html#protflow.runners.RunnerOutput" title="protflow.runners.RunnerOutput">RunnerOutput</a></p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>FileNotFoundError</strong> – If the TMscore executable is not found in the specified environment.</p></li>
<li><p><strong>ValueError</strong> – If invalid arguments are provided to the method or if required reference columns are missing.</p></li>
<li><p><strong>RuntimeError</strong> – If no TM scores are found in the output files.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>run</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">tmscore</span> <span class="kn">import</span> <span class="n">TMscore</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the TMscore class</span>
<span class="n">tmscore</span> <span class="o">=</span> <span class="n">TMscore</span><span class="p">()</span>

<span class="c1"># Run the scoring process</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">tmscore</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_2&quot;</span><span class="p">,</span>
    <span class="n">ref_col</span><span class="o">=</span><span class="s2">&quot;reference_pdb&quot;</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;-c&quot;</span><span class="p">,</span>
    <span class="n">pose_options</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;-d&quot;</span><span class="p">],</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Setup and Execution:</strong> The method ensures that the environment is correctly set up, directories are prepared, and necessary commands are constructed and executed.</p></li>
<li><p><strong>Reference Handling:</strong> The method validates the reference column and prepares the reference structures for scoring.</p></li>
<li><p><strong>Output Management:</strong> The method handles the collection and processing of output data, ensuring that results are organized and accessible for further analysis.</p></li>
<li><p><strong>Customization:</strong> Extensive customization options are provided through parameters, allowing users to tailor the scoring process to their specific needs.</p></li>
</ul>
</dd>
</dl>
<p>This method is designed to streamline the execution of TMscore processes within the ProtFlow framework, making it easier for researchers and developers to perform and analyze protein structure comparisons.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.metrics.tmscore.TMscore.write_cmd">
<span class="sig-name descname"><span class="pre">write_cmd</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pose_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pose_options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">str</span></span></span><a class="reference internal" href="_modules/protflow/metrics/tmscore.html#TMscore.write_cmd"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.metrics.tmscore.TMscore.write_cmd" title="Link to this definition"></a></dt>
<dd><p>Write the command to run TMscore.</p>
<p>This method constructs the command to execute TMscore based on the provided parameters. It formats the options and flags correctly and sets up the command to be run in the environment.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pose_path</strong> (<em>str</em>) – The path to the pose file.</p></li>
<li><p><strong>ref_path</strong> (<em>str</em>) – The path to the reference file.</p></li>
<li><p><strong>output_dir</strong> (<em>str</em>) – The directory where output files will be saved.</p></li>
<li><p><strong>options</strong> (<em>str</em><em>, </em><em>optional</em>) – Additional command-line options for TMscore. Defaults to None.</p></li>
<li><p><strong>pose_options</strong> (<em>str</em><em>, </em><em>optional</em>) – Pose-specific options for TMscore. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The constructed command string to run TMscore.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>write_cmd</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">tmscore</span> <span class="kn">import</span> <span class="n">TMscore</span>

<span class="c1"># Initialize the TMscore class</span>
<span class="n">tmscore</span> <span class="o">=</span> <span class="n">TMscore</span><span class="p">()</span>

<span class="c1"># Write the command</span>
<span class="n">cmd</span> <span class="o">=</span> <span class="n">tmscore</span><span class="o">.</span><span class="n">write_cmd</span><span class="p">(</span>
    <span class="n">pose_path</span><span class="o">=</span><span class="s2">&quot;pose.pdb&quot;</span><span class="p">,</span>
    <span class="n">ref_path</span><span class="o">=</span><span class="s2">&quot;reference.pdb&quot;</span><span class="p">,</span>
    <span class="n">output_dir</span><span class="o">=</span><span class="s2">&quot;output/&quot;</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;-a&quot;</span><span class="p">,</span>
    <span class="n">pose_options</span><span class="o">=</span><span class="s2">&quot;-b&quot;</span>
<span class="p">)</span>

<span class="c1"># Print the command</span>
<span class="nb">print</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Command Construction:</strong> The method constructs the command string by parsing and formatting the provided options and pose-specific options.</p></li>
<li><p><strong>Output Management:</strong> Ensures that the output files are correctly named and saved in the specified directory.</p></li>
</ul>
</dd>
</dl>
<p>This method is designed to streamline the construction of commands for TMscore processes, ensuring that all necessary options are correctly formatted and included.</p>
</dd></dl>

</section>
</section>
</dd></dl>

</section>
<section id="module-protflow.metrics">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-protflow.metrics" title="Link to this heading"></a></h2>
<p>protflow.metrics subpackage init</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="protflow.html" class="btn btn-neutral float-left" title="protflow package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="protflow.tools.html" class="btn btn-neutral float-right" title="protflow.tools package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Your Name.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>