<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>protflow.tools package &mdash; ProtFlow 1.0.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5d32c60e"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="_static/documentation_options.js?v=8d563738"></script>
        <script src="_static/doctools.js?v=9a2dae69"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="protflow.metrics package" href="protflow.metrics.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            ProtFlow
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="protflow.html">protflow package</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="protflow.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="protflow.metrics.html">protflow.metrics package</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">protflow.tools package</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-protflow.tools.attnpacker">protflow.tools.attnpacker module</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-protflow.tools.colabfold">protflow.tools.colabfold module</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-protflow.tools.esmfold">protflow.tools.esmfold module</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-protflow.tools.ligandmpnn">protflow.tools.ligandmpnn module</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-protflow.tools.protein_edits">protflow.tools.protein_edits module</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-protflow.tools.protein_generator">protflow.tools.protein_generator module</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-protflow.tools.residue_selectors">protflow.tools.residue_selectors module</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-protflow.tools.rfdiffusion">protflow.tools.rfdiffusion module</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-protflow.tools.rosetta">protflow.tools.rosetta module</a></li>
<li class="toctree-l4"><a class="reference internal" href="#protflow-tools-rosettascripts-module">protflow.tools.rosettascripts module</a></li>
<li class="toctree-l4"><a class="reference internal" href="#module-protflow.tools">Module contents</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="protflow.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="protflow.html#protflow-config-module">protflow.config module</a></li>
<li class="toctree-l2"><a class="reference internal" href="protflow.html#protflow-config-template-module">protflow.config_template module</a></li>
<li class="toctree-l2"><a class="reference internal" href="protflow.html#module-protflow.jobstarters">protflow.jobstarters module</a></li>
<li class="toctree-l2"><a class="reference internal" href="protflow.html#module-protflow.poses">protflow.poses module</a></li>
<li class="toctree-l2"><a class="reference internal" href="protflow.html#module-protflow.residues">protflow.residues module</a></li>
<li class="toctree-l2"><a class="reference internal" href="protflow.html#module-protflow.runners">protflow.runners module</a></li>
<li class="toctree-l2"><a class="reference internal" href="protflow.html#module-contents">Module contents</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">ProtFlow</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="protflow.html">protflow package</a></li>
      <li class="breadcrumb-item active">protflow.tools package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/protflow.tools.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="protflow-tools-package">
<h1>protflow.tools package<a class="headerlink" href="#protflow-tools-package" title="Link to this heading"></a></h1>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Link to this heading"></a></h2>
</section>
<section id="module-protflow.tools.attnpacker">
<span id="protflow-tools-attnpacker-module"></span><h2>protflow.tools.attnpacker module<a class="headerlink" href="#module-protflow.tools.attnpacker" title="Link to this heading"></a></h2>
<section id="attnpacker-module">
<h3>AttnPacker Module<a class="headerlink" href="#attnpacker-module" title="Link to this heading"></a></h3>
<p>This module provides the functionality to integrate AttnPacker within the ProtFlow framework. It offers tools to run AttnPacker, handle its inputs and outputs, and process the resulting data in a structured and automated manner.</p>
<section id="detailed-description">
<h4>Detailed Description<a class="headerlink" href="#detailed-description" title="Link to this heading"></a></h4>
<p>The <cite>AttnPacker</cite> class encapsulates the functionality necessary to execute AttnPacker runs. It manages the configuration of paths to essential scripts and Python executables, sets up the environment, and handles the execution of packing processes. It also includes methods for collecting and processing output data, ensuring that the results are organized and accessible for further analysis within the ProtFlow ecosystem.
The module is designed to streamline the integration of AttnPacker into larger computational workflows. It supports the automatic setup of job parameters, execution of AttnPacker commands, and parsing of output files into a structured DataFrame format. This facilitates subsequent data analysis and visualization steps.</p>
</section>
<section id="usage">
<h4>Usage<a class="headerlink" href="#usage" title="Link to this heading"></a></h4>
<p>To use this module, create an instance of the <cite>AttnPacker</cite> class and invoke its <cite>run</cite> method with appropriate parameters. The module will handle the configuration, execution, and result collection processes. Detailed control over the packing process is provided through various parameters, allowing for customized runs tailored to specific research needs.</p>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize and use the <cite>AttnPacker</cite> class within a ProtFlow pipeline:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">attnpacker</span> <span class="kn">import</span> <span class="n">AttnPacker</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the AttnPacker class</span>
<span class="n">attnpacker</span> <span class="o">=</span> <span class="n">AttnPacker</span><span class="p">()</span>

<span class="c1"># Run the packing process</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">attnpacker</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;packing.num_designs=10&quot;</span><span class="p">,</span>
    <span class="n">pose_options</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;packing.input_pdb=&#39;input.pdb&#39;&quot;</span><span class="p">],</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="further-details">
<h4>Further Details<a class="headerlink" href="#further-details" title="Link to this heading"></a></h4>
<blockquote>
<div><ul class="simple">
<li><p>Edge Cases: The module handles various edge cases, such as empty pose lists and the need to overwrite previous results. It ensures robust error handling and logging for easier debugging and verification of the packing process.</p></li>
<li><p>Customizability: Users can customize the packing process through multiple parameters, including the number of packings, specific options for the AttnPacker script, and options for handling pose-specific parameters.</p></li>
<li><p>Integration: The module seamlessly integrates with other components of the ProtFlow framework, leveraging shared configurations and data structures to provide a cohesive user experience.</p></li>
</ul>
</div></blockquote>
<p>This module is intended for researchers and developers who need to incorporate AttnPacker into their protein design and analysis workflows. By automating many of the setup and execution steps, it allows users to focus on interpreting results and advancing their scientific inquiries.</p>
<p class="rubric">Notes</p>
<p>This module is part of the ProtFlow package and is designed to work in tandem with other components of the package, especially those related to job management in HPC environments.</p>
</section>
<section id="authors">
<h4>Authors<a class="headerlink" href="#authors" title="Link to this heading"></a></h4>
<p>Markus Braun, Adrian Tripp</p>
</section>
<section id="version">
<h4>Version<a class="headerlink" href="#version" title="Link to this heading"></a></h4>
<p>0.1.0</p>
</section>
</section>
<dl class="py class">
<dt class="sig sig-object py" id="protflow.tools.attnpacker.AttnPacker">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">protflow.tools.attnpacker.</span></span><span class="sig-name descname"><span class="pre">AttnPacker</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">script_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">python_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/tools/attnpacker.html#AttnPacker"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.attnpacker.AttnPacker" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="protflow.html#protflow.runners.Runner" title="protflow.runners.Runner"><code class="xref py py-class docutils literal notranslate"><span class="pre">Runner</span></code></a></p>
<section id="attnpacker-class">
<h3>AttnPacker Class<a class="headerlink" href="#attnpacker-class" title="Link to this heading"></a></h3>
<p>The <cite>AttnPacker</cite> class is a specialized class designed to facilitate the execution of AttnPacker within the ProtFlow framework. It extends the <cite>Runner</cite> class and incorporates specific methods to handle the setup, execution, and data collection associated with packing processes.</p>
<section id="id1">
<h4>Detailed Description<a class="headerlink" href="#id1" title="Link to this heading"></a></h4>
<p>The <cite>AttnPacker</cite> class manages all aspects of running AttnPacker simulations. It handles the configuration of necessary scripts and executables, prepares the environment for packing processes, and executes the packing commands. Additionally, it collects and processes the output data, organizing it into a structured format for further analysis.</p>
<dl class="simple">
<dt>Key functionalities include:</dt><dd><ul class="simple">
<li><p>Setting up paths to AttnPacker scripts and Python executables.</p></li>
<li><p>Configuring job starter options, either automatically or manually.</p></li>
<li><p>Handling the execution of AttnPacker commands with support for multiple packings.</p></li>
<li><p>Collecting and processing output data into a pandas DataFrame.</p></li>
<li><p>Ensuring robust error handling and logging for easier debugging and verification.</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">rtype<span class="colon">:</span></dt>
<dd class="field-odd"><p>An instance of the <cite>AttnPacker</cite> class, configured to run AttnPacker processes and handle outputs efficiently.</p>
</dd>
<dt class="field-even">raises - FileNotFoundError<span class="colon">:</span></dt>
<dd class="field-even"><p>If required files or directories are not found during the execution process.:</p>
</dd>
<dt class="field-odd">raises - ValueError<span class="colon">:</span></dt>
<dd class="field-odd"><p>If invalid arguments are provided to the methods.:</p>
</dd>
<dt class="field-even">raises - KeyError<span class="colon">:</span></dt>
<dd class="field-even"><p>If forbidden options are included in the command parameters.:</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize and use the <cite>AttnPacker</cite> class:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">attnpacker</span> <span class="kn">import</span> <span class="n">AttnPacker</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the AttnPacker class</span>
<span class="n">attnpacker</span> <span class="o">=</span> <span class="n">AttnPacker</span><span class="p">()</span>

<span class="c1"># Run the packing process</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">attnpacker</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;packing.num_designs=10&quot;</span><span class="p">,</span>
    <span class="n">pose_options</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;packing.input_pdb=&#39;input.pdb&#39;&quot;</span><span class="p">],</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id2">
<h4>Further Details<a class="headerlink" href="#id2" title="Link to this heading"></a></h4>
<blockquote>
<div><ul class="simple">
<li><p>Edge Cases: The class includes handling for various edge cases, such as empty pose lists, the need to overwrite previous results, and the presence of existing score files.</p></li>
<li><p>Customization: The class provides extensive customization options through its parameters, allowing users to tailor the packing process to their specific needs.</p></li>
<li><p>Integration: Seamlessly integrates with other ProtFlow components, leveraging shared configurations and data structures for a unified workflow.</p></li>
</ul>
</div></blockquote>
<p>The AttnPacker class is intended for researchers and developers who need to perform packing simulations as part of their protein design and analysis workflows. It simplifies the process, allowing users to focus on analyzing results and advancing their research.</p>
<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.attnpacker.AttnPacker.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">script_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">python_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/protflow/tools/attnpacker.html#AttnPacker.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.attnpacker.AttnPacker.__init__" title="Link to this definition"></a></dt>
<dd><p>sbatch_options are set automatically, but can also be manually set. Manual setting is not recommended.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.attnpacker.AttnPacker.run">
<span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><span class="pre">JobStarter</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pose_options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span></span><a class="reference internal" href="_modules/protflow/tools/attnpacker.html#AttnPacker.run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.attnpacker.AttnPacker.run" title="Link to this definition"></a></dt>
<dd><p>Execute the AttnPacker process with given poses and jobstarter configuration.</p>
<p>This method sets up and runs the AttnPacker process using the provided poses and jobstarter object. It handles the configuration, execution, and collection of output data, ensuring that the results are organized and accessible for further analysis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>poses</strong> (<a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><em>Poses</em></a>) – The Poses object containing the protein structures.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – A prefix used to name and organize the output files.</p></li>
<li><p><strong>jobstarter</strong> (<a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><em>JobStarter</em></a><em>, </em><em>optional</em>) – An instance of the JobStarter class, which manages job execution. Defaults to None.</p></li>
<li><p><strong>options</strong> (<em>str</em><em>, </em><em>optional</em>) – Additional options for the AttnPacker script. Defaults to None.</p></li>
<li><p><strong>pose_options</strong> (<em>list</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – A list of pose-specific options for the AttnPacker script. Defaults to None.</p></li>
<li><p><strong>overwrite</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, overwrite existing output files. Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An updated Poses object containing the processed poses and results of the AttnPacker process.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses">Poses</a></p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>FileNotFoundError</strong> – If required files or directories are not found during the execution process.</p></li>
<li><p><strong>ValueError</strong> – If invalid arguments are provided to the method.</p></li>
<li><p><strong>KeyError</strong> – If forbidden options are included in the command parameters.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>run</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">attnpacker</span> <span class="kn">import</span> <span class="n">AttnPacker</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the AttnPacker class</span>
<span class="n">attnpacker</span> <span class="o">=</span> <span class="n">AttnPacker</span><span class="p">()</span>

<span class="c1"># Run the packing process</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">attnpacker</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;packing.num_designs=10&quot;</span><span class="p">,</span>
    <span class="n">pose_options</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;packing.input_pdb=&#39;input.pdb&#39;&quot;</span><span class="p">],</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Setup and Execution:</strong> The method ensures that the environment is correctly set up, directories are prepared, and necessary commands are constructed and executed. It sets up specific directories for output PDBs and checks for existing score files.</p></li>
<li><p><strong>Output Management:</strong> The method handles the collection and processing of output data, reading scores from a CSV file and organizing results into a structured DataFrame. It ensures that results are accessible for further analysis.</p></li>
<li><p><strong>Customization:</strong> Extensive customization options are provided through parameters, allowing users to tailor the packing process to their specific needs. Users can specify additional options and pose-specific parameters for the AttnPacker script.</p></li>
</ul>
</dd>
</dl>
<p>This method is designed to streamline the execution of AttnPacker processes within the ProtFlow framework, making it easier for researchers and developers to perform and analyze packing simulations.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.attnpacker.AttnPacker.write_cmd">
<span class="sig-name descname"><span class="pre">write_cmd</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pose_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pose_options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/tools/attnpacker.html#AttnPacker.write_cmd"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.attnpacker.AttnPacker.write_cmd" title="Link to this definition"></a></dt>
<dd><p>Write the command to run the AttnPacker script for a given pose.</p>
<p>This method constructs the command line string necessary to execute the AttnPacker script for a given pose. It incorporates the specified options and pose-specific parameters, ensuring that the command is correctly formatted and includes all required arguments. It also checks for forbidden options to prevent conflicts.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pose_path</strong> (<em>str</em>) – The path to the input PDB file for the pose.</p></li>
<li><p><strong>output_dir</strong> (<em>str</em>) – The directory where output files will be stored.</p></li>
<li><p><strong>options</strong> (<em>str</em><em>, </em><em>optional</em>) – Additional options for the AttnPacker script. Defaults to None.</p></li>
<li><p><strong>pose_options</strong> (<em>str</em><em>, </em><em>optional</em>) – A list of pose-specific options for the AttnPacker script. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The command line string to execute the AttnPacker script with the specified parameters.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>KeyError</strong> – If forbidden options are included in the command parameters.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>write_cmd</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">attnpacker</span> <span class="kn">import</span> <span class="n">AttnPacker</span>

<span class="c1"># Initialize the AttnPacker class</span>
<span class="n">attnpacker</span> <span class="o">=</span> <span class="n">AttnPacker</span><span class="p">()</span>

<span class="c1"># Define the input pose path and output directory</span>
<span class="n">pose_path</span> <span class="o">=</span> <span class="s2">&quot;input.pdb&quot;</span>
<span class="n">output_dir</span> <span class="o">=</span> <span class="s2">&quot;/path/to/output&quot;</span>

<span class="c1"># Write the command with additional options and pose-specific parameters</span>
<span class="n">cmd</span> <span class="o">=</span> <span class="n">attnpacker</span><span class="o">.</span><span class="n">write_cmd</span><span class="p">(</span>
    <span class="n">pose_path</span><span class="o">=</span><span class="n">pose_path</span><span class="p">,</span>
    <span class="n">output_dir</span><span class="o">=</span><span class="n">output_dir</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;packing.num_designs=10&quot;</span><span class="p">,</span>
    <span class="n">pose_options</span><span class="o">=</span><span class="s2">&quot;packing.input_pdb=&#39;input.pdb&#39;&quot;</span>
<span class="p">)</span>

<span class="c1"># Print the command</span>
<span class="nb">print</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Command Construction:</strong> This method ensures that the command string is correctly constructed with all necessary arguments. It includes paths to the script directory, output directory, input PDB file, and score file, as well as any additional options.</p></li>
<li><p><strong>Validation:</strong> The method checks for forbidden options that could conflict with the required arguments, raising a KeyError if any are found. This helps ensure that the command is valid and will run correctly.</p></li>
</ul>
</dd>
</dl>
<p>This method is designed to facilitate the execution of AttnPacker processes within the ProtConductor framework, providing a flexible and robust way to construct and run commands for packing simulations.</p>
</dd></dl>

</section>
</section>
</dd></dl>

</section>
<section id="module-protflow.tools.colabfold">
<span id="protflow-tools-colabfold-module"></span><h2>protflow.tools.colabfold module<a class="headerlink" href="#module-protflow.tools.colabfold" title="Link to this heading"></a></h2>
<section id="colabfold-module">
<h3>ColabFold Module<a class="headerlink" href="#colabfold-module" title="Link to this heading"></a></h3>
<p>This module provides functionality to integrate ColabFold within the ProtFlow framework, enabling the execution of AlphaFold2 runs on ColabFold. It includes tools to handle inputs, execute runs, and process outputs in a structured and automated manner.</p>
<section id="id3">
<h4>Detailed Description<a class="headerlink" href="#id3" title="Link to this heading"></a></h4>
<p>The <cite>ColabFold</cite> class encapsulates all necessary functionalities to run AlphaFold2 through ColabFold. It manages the configuration of essential scripts and paths, sets up the environment, and handles the execution of prediction processes. The class also includes methods for collecting and processing output data, ensuring the results are organized and accessible for further analysis within the ProtFlow ecosystem.</p>
<p>This module streamlines the integration of ColabFold into larger computational workflows by supporting the automatic setup of job parameters, execution of ColabFold commands, and parsing of output files into a structured DataFrame format. This facilitates subsequent data analysis and visualization steps.</p>
</section>
<section id="id4">
<h4>Usage<a class="headerlink" href="#id4" title="Link to this heading"></a></h4>
<p>To use this module, create an instance of the <cite>ColabFold</cite> class and invoke its <cite>run</cite> method with appropriate parameters. The module will handle the configuration, execution, and result collection processes. Detailed control over the prediction process is provided through various parameters, allowing for customized runs tailored to specific research needs.</p>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize and use the <cite>ColabFold</cite> class within a ProtFlow pipeline:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">colabfold</span> <span class="kn">import</span> <span class="n">ColabFold</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">LocalJobStarter</span><span class="p">(</span><span class="n">max_cores</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>

<span class="c1"># Initialize the ColabFold class</span>
<span class="n">colabfold</span> <span class="o">=</span> <span class="n">ColabFold</span><span class="p">()</span>

<span class="c1"># Run the prediction process</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">colabfold</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;--msa-mode single-sequence&quot;</span><span class="p">,</span>
    <span class="n">pose_options</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id5">
<h4>Further Details<a class="headerlink" href="#id5" title="Link to this heading"></a></h4>
<ul class="simple">
<li><p><strong>Edge Cases:</strong> The module handles various edge cases, such as empty pose lists and the need to overwrite previous results. It ensures robust error handling and logging for easier debugging and verification of the prediction process.</p></li>
<li><p><strong>Customizability:</strong> Users can customize the prediction process through multiple parameters, including the number of diffusions, specific options for the ColabFold script, and options for handling pose-specific parameters.</p></li>
<li><p><strong>Integration:</strong> The module seamlessly integrates with other components of the ProtFlow framework, leveraging shared configurations and data structures to provide a cohesive user experience.</p></li>
</ul>
<p>This module is intended for researchers and developers who need to incorporate ColabFold into their protein design and analysis workflows. By automating many of the setup and execution steps, it allows users to focus on interpreting results and advancing their scientific inquiries.</p>
<p class="rubric">Notes</p>
<p>This module is part of the ProtFlow package and is designed to work in tandem with other components of the package, especially those related to job management in HPC environments.</p>
</section>
<section id="id6">
<h4>Authors<a class="headerlink" href="#id6" title="Link to this heading"></a></h4>
<p>Markus Braun, Adrian Tripp</p>
</section>
<section id="id7">
<h4>Version<a class="headerlink" href="#id7" title="Link to this heading"></a></h4>
<p>0.1.0</p>
</section>
</section>
<dl class="py class">
<dt class="sig sig-object py" id="protflow.tools.colabfold.Colabfold">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">protflow.tools.colabfold.</span></span><span class="sig-name descname"><span class="pre">Colabfold</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">script_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/tools/colabfold.html#Colabfold"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.colabfold.Colabfold" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="protflow.html#protflow.runners.Runner" title="protflow.runners.Runner"><code class="xref py py-class docutils literal notranslate"><span class="pre">Runner</span></code></a></p>
<section id="colabfold-class">
<h3>ColabFold Class<a class="headerlink" href="#colabfold-class" title="Link to this heading"></a></h3>
<p>The <cite>ColabFold</cite> class is a specialized class designed to facilitate the execution of AlphaFold2 within the ColabFold environment as part of the ProtFlow framework. It extends the <cite>Runner</cite> class and incorporates specific methods to handle the setup, execution, and data collection associated with AlphaFold2 prediction processes.</p>
<section id="id8">
<h4>Detailed Description<a class="headerlink" href="#id8" title="Link to this heading"></a></h4>
<p>The <cite>ColabFold</cite> class manages all aspects of running AlphaFold2 predictions through ColabFold. It handles the configuration of necessary scripts and executables, prepares the environment for the prediction processes, and executes the prediction commands. Additionally, it collects and processes the output data, organizing it into a structured format for further analysis.</p>
<dl class="simple">
<dt>Key functionalities include:</dt><dd><ul class="simple">
<li><p>Setting up paths to ColabFold scripts and necessary directories.</p></li>
<li><p>Configuring job starter options, either automatically or manually.</p></li>
<li><p>Handling the execution of AlphaFold2 prediction commands with support for batch processing.</p></li>
<li><p>Collecting and processing output data into a pandas DataFrame.</p></li>
<li><p>Managing input FASTA files and preparing them for prediction.</p></li>
<li><p>Overwriting previous results if specified.</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">rtype<span class="colon">:</span></dt>
<dd class="field-odd"><p>An instance of the <cite>ColabFold</cite> class, configured to run AlphaFold2 prediction processes and handle outputs efficiently.</p>
</dd>
<dt class="field-even">raises - FileNotFoundError<span class="colon">:</span></dt>
<dd class="field-even"><p>If required files or directories are not found during the execution process.:</p>
</dd>
<dt class="field-odd">raises - ValueError<span class="colon">:</span></dt>
<dd class="field-odd"><p>If invalid arguments are provided to the methods.:</p>
</dd>
<dt class="field-even">raises - TypeError<span class="colon">:</span></dt>
<dd class="field-even"><p>If pose options are not of the expected type.:</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize and use the <cite>ColabFold</cite> class:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">colabfold</span> <span class="kn">import</span> <span class="n">ColabFold</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the ColabFold class</span>
<span class="n">colabfold</span> <span class="o">=</span> <span class="n">ColabFold</span><span class="p">()</span>

<span class="c1"># Run the prediction process</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">colabfold</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;inference.num_designs=10&quot;</span><span class="p">,</span>
    <span class="n">pose_options</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;inference.input_pdb=&#39;input.pdb&#39;&quot;</span><span class="p">],</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id9">
<h4>Further Details<a class="headerlink" href="#id9" title="Link to this heading"></a></h4>
<blockquote>
<div><ul class="simple">
<li><p>Edge Cases: The class includes handling for various edge cases, such as empty pose lists, the need to overwrite previous results, and the presence of existing score files.</p></li>
<li><p>Customization: The class provides extensive customization options through its parameters, allowing users to tailor the prediction process to their specific needs.</p></li>
<li><p>Integration: Seamlessly integrates with other ProtFlow components, leveraging shared configurations and data structures for a unified workflow.</p></li>
</ul>
</div></blockquote>
<p>The ColabFold class is intended for researchers and developers who need to perform AlphaFold2 predictions as part of their protein design and analysis workflows. It simplifies the process, allowing users to focus on analyzing results and advancing their research.</p>
<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.colabfold.Colabfold.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">script_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/protflow/tools/colabfold.html#Colabfold.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.colabfold.Colabfold.__init__" title="Link to this definition"></a></dt>
<dd><section id="init-method">
<h5>__init__ Method<a class="headerlink" href="#init-method" title="Link to this heading"></a></h5>
<p>The <cite>__init__</cite> method initializes an instance of the <cite>ColabFold</cite> class, setting up necessary configurations for running AlphaFold2 predictions through ColabFold within the ProtFlow framework.</p>
<section id="id10">
<h6>Detailed Description<a class="headerlink" href="#id10" title="Link to this heading"></a></h6>
<p>This method sets up the paths to the ColabFold script and initializes default values for various attributes required for running predictions. It also allows for the optional configuration of a job starter.</p>
<dl class="field-list simple">
<dt class="field-odd">param script_path<span class="colon">:</span></dt>
<dd class="field-odd"><p>The path to the ColabFold script. Defaults to <cite>protflow.config.COLABFOLD_SCRIPT_PATH</cite>.</p>
</dd>
<dt class="field-even">type script_path<span class="colon">:</span></dt>
<dd class="field-even"><p>str, optional</p>
</dd>
<dt class="field-odd">param jobstarter<span class="colon">:</span></dt>
<dd class="field-odd"><p>An instance of the <cite>JobStarter</cite> class for managing job execution. Defaults to None.</p>
</dd>
<dt class="field-even">type jobstarter<span class="colon">:</span></dt>
<dd class="field-even"><p>JobStarter, optional</p>
</dd>
<dt class="field-odd">returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
<dt class="field-even">raises ValueError<span class="colon">:</span></dt>
<dd class="field-even"><p>If the <cite>script_path</cite> is not provided.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize the <cite>ColabFold</cite> class:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">colabfold</span> <span class="kn">import</span> <span class="n">ColabFold</span>

<span class="c1"># Initialize the ColabFold class</span>
<span class="n">colabfold</span> <span class="o">=</span> <span class="n">ColabFold</span><span class="p">(</span><span class="n">script_path</span><span class="o">=</span><span class="s1">&#39;/path/to/colabfold.py&#39;</span><span class="p">,</span> <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.colabfold.Colabfold.collect_scores">
<span class="sig-name descname"><span class="pre">collect_scores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">work_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_return_poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/protflow/tools/colabfold.html#Colabfold.collect_scores"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.colabfold.Colabfold.collect_scores" title="Link to this definition"></a></dt>
<dd><section id="collect-scores-method">
<h5>collect_scores Method<a class="headerlink" href="#collect-scores-method" title="Link to this heading"></a></h5>
<p>The <cite>collect_scores</cite> method collects and processes the prediction scores from the ColabFold output, organizing them into a pandas DataFrame for further analysis.</p>
<section id="id11">
<h6>Detailed Description<a class="headerlink" href="#id11" title="Link to this heading"></a></h6>
<p>This method gathers the prediction scores from the output files generated by ColabFold. It processes these scores and organizes them into a structured DataFrame, which includes various statistical measures.</p>
<dl class="field-list simple">
<dt class="field-odd">param work_dir<span class="colon">:</span></dt>
<dd class="field-odd"><p>The working directory where the ColabFold outputs are stored.</p>
</dd>
<dt class="field-even">type work_dir<span class="colon">:</span></dt>
<dd class="field-even"><p>str</p>
</dd>
<dt class="field-odd">param num_return_poses<span class="colon">:</span></dt>
<dd class="field-odd"><p>The number of top poses to return based on the prediction scores. Defaults to 1.</p>
</dd>
<dt class="field-even">type num_return_poses<span class="colon">:</span></dt>
<dd class="field-even"><p>int, optional</p>
</dd>
<dt class="field-odd">returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A DataFrame containing the collected and processed scores.</p>
</dd>
<dt class="field-even">rtype<span class="colon">:</span></dt>
<dd class="field-even"><p>pd.DataFrame</p>
</dd>
<dt class="field-odd">raises FileNotFoundError<span class="colon">:</span></dt>
<dd class="field-odd"><p>If no output files are found in the specified directory.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>collect_scores</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Collect and process the prediction scores</span>
<span class="n">scores</span> <span class="o">=</span> <span class="n">colabfold</span><span class="o">.</span><span class="n">collect_scores</span><span class="p">(</span><span class="n">work_dir</span><span class="o">=</span><span class="s1">&#39;/path/to/work_dir&#39;</span><span class="p">,</span> <span class="n">num_return_poses</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id12">
<h6>Further Details<a class="headerlink" href="#id12" title="Link to this heading"></a></h6>
<blockquote>
<div><ul class="simple">
<li><p><strong>JSON and PDB File Parsing:</strong> The method identifies and parses JSON and PDB files generated by ColabFold, extracting relevant score information from these files.</p></li>
<li><p><strong>Statistical Measures:</strong> For each set of predictions, the method calculates various statistics, including mean pLDDT, max PAE, and PTM scores, organizing these measures into the DataFrame.</p></li>
<li><p><strong>Rank and Description:</strong> The scores are ranked and annotated with descriptions to help identify the top poses based on prediction quality.</p></li>
<li><p><strong>File Handling:</strong> The method includes robust file handling to ensure that only the relevant files are processed, and any existing files are correctly identified or overwritten as needed.</p></li>
<li><p><strong>Pose Location:</strong> The final DataFrame includes the file paths to the predicted PDB files, facilitating easy access for further analysis or visualization.</p></li>
</ul>
</div></blockquote>
</section>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.colabfold.Colabfold.prep_fastas_for_prediction">
<span class="sig-name descname"><span class="pre">prep_fastas_for_prediction</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fasta_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_filenum</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/protflow/tools/colabfold.html#Colabfold.prep_fastas_for_prediction"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.colabfold.Colabfold.prep_fastas_for_prediction" title="Link to this definition"></a></dt>
<dd><section id="prep-fastas-for-prediction-method">
<h5>prep_fastas_for_prediction Method<a class="headerlink" href="#prep-fastas-for-prediction-method" title="Link to this heading"></a></h5>
<p>The <cite>prep_fastas_for_prediction</cite> method prepares input FASTA files for AlphaFold2 predictions by splitting the input sequences into batches and writing them to files.</p>
<section id="id13">
<h6>Detailed Description<a class="headerlink" href="#id13" title="Link to this heading"></a></h6>
<p>This method divides the input protein sequences into batches, which can help in managing computational resources effectively. It writes the batches into FASTA files stored in the specified directory.</p>
<dl class="field-list simple">
<dt class="field-odd">param poses<span class="colon">:</span></dt>
<dd class="field-odd"><p>List of paths to input FASTA files.</p>
</dd>
<dt class="field-even">type poses<span class="colon">:</span></dt>
<dd class="field-even"><p>list[str]</p>
</dd>
<dt class="field-odd">param fasta_dir<span class="colon">:</span></dt>
<dd class="field-odd"><p>Directory where the new FASTA files will be written.</p>
</dd>
<dt class="field-even">type fasta_dir<span class="colon">:</span></dt>
<dd class="field-even"><p>str</p>
</dd>
<dt class="field-odd">param max_filenum<span class="colon">:</span></dt>
<dd class="field-odd"><p>Maximum number of FASTA files to write.</p>
</dd>
<dt class="field-even">type max_filenum<span class="colon">:</span></dt>
<dd class="field-even"><p>int</p>
</dd>
<dt class="field-odd">returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>List of paths to the prepared FASTA files.</p>
</dd>
<dt class="field-even">rtype<span class="colon">:</span></dt>
<dd class="field-even"><p>list[str]</p>
</dd>
<dt class="field-odd">raises None<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>prep_fastas_for_prediction</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Prepare input FASTA files for prediction</span>
<span class="n">fastas</span> <span class="o">=</span> <span class="n">colabfold</span><span class="o">.</span><span class="n">prep_fastas_for_prediction</span><span class="p">(</span><span class="n">poses</span><span class="o">=</span><span class="n">poses_list</span><span class="p">,</span> <span class="n">fasta_dir</span><span class="o">=</span><span class="s1">&#39;/path/to/fasta_dir&#39;</span><span class="p">,</span> <span class="n">max_filenum</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.colabfold.Colabfold.run">
<span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><span class="pre">JobStarter</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pose_options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_top_n_poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="protflow.html#protflow.runners.RunnerOutput" title="protflow.runners.RunnerOutput"><span class="pre">RunnerOutput</span></a></span></span><a class="reference internal" href="_modules/protflow/tools/colabfold.html#Colabfold.run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.colabfold.Colabfold.run" title="Link to this definition"></a></dt>
<dd><section id="run-method">
<h5>run Method<a class="headerlink" href="#run-method" title="Link to this heading"></a></h5>
<p>The <cite>run</cite> method of the <cite>ColabFold</cite> class executes AlphaFold2 predictions using ColabFold within the ProtFlow framework. It manages the setup, execution, and result collection processes, providing a streamlined way to integrate AlphaFold2 predictions into larger computational workflows.</p>
<section id="id14">
<h6>Detailed Description<a class="headerlink" href="#id14" title="Link to this heading"></a></h6>
<p>This method orchestrates the entire prediction process, from preparing input data and configuring the environment to running the prediction commands and collecting the results. The method supports batch processing of input FASTA files and handles various edge cases, such as overwriting existing results and managing job starter options.</p>
<dl class="field-list simple">
<dt class="field-odd">param poses<span class="colon">:</span></dt>
<dd class="field-odd"><p>The Poses object containing the protein structures.</p>
</dd>
<dt class="field-even">type poses<span class="colon">:</span></dt>
<dd class="field-even"><p>Poses</p>
</dd>
<dt class="field-odd">param prefix<span class="colon">:</span></dt>
<dd class="field-odd"><p>A prefix used to name and organize the output files.</p>
</dd>
<dt class="field-even">type prefix<span class="colon">:</span></dt>
<dd class="field-even"><p>str</p>
</dd>
<dt class="field-odd">param jobstarter<span class="colon">:</span></dt>
<dd class="field-odd"><p>An instance of the JobStarter class, which manages job execution. Defaults to None.</p>
</dd>
<dt class="field-even">type jobstarter<span class="colon">:</span></dt>
<dd class="field-even"><p>JobStarter, optional</p>
</dd>
<dt class="field-odd">param options<span class="colon">:</span></dt>
<dd class="field-odd"><p>Additional options for the AlphaFold2 prediction commands. Defaults to None.</p>
</dd>
<dt class="field-even">type options<span class="colon">:</span></dt>
<dd class="field-even"><p>str, optional</p>
</dd>
<dt class="field-odd">param pose_options<span class="colon">:</span></dt>
<dd class="field-odd"><p>Specific options for handling pose-related parameters during prediction. Defaults to None.</p>
</dd>
<dt class="field-even">type pose_options<span class="colon">:</span></dt>
<dd class="field-even"><p>str, optional</p>
</dd>
<dt class="field-odd">param overwrite<span class="colon">:</span></dt>
<dd class="field-odd"><p>If True, existing results will be overwritten. Defaults to False.</p>
</dd>
<dt class="field-even">type overwrite<span class="colon">:</span></dt>
<dd class="field-even"><p>bool, optional</p>
</dd>
<dt class="field-odd">param return_top_n_poses<span class="colon">:</span></dt>
<dd class="field-odd"><p>The number of top poses to return based on the prediction scores. Defaults to 1.</p>
</dd>
<dt class="field-even">type return_top_n_poses<span class="colon">:</span></dt>
<dd class="field-even"><p>int, optional</p>
</dd>
<dt class="field-odd">returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An object containing the results of the AlphaFold2 predictions, organized in a pandas DataFrame.</p>
</dd>
<dt class="field-even">rtype<span class="colon">:</span></dt>
<dd class="field-even"><p>RunnerOutput</p>
</dd>
<dt class="field-odd">raises FileNotFoundError<span class="colon">:</span></dt>
<dd class="field-odd"><p>If required files or directories are not found during the execution process.</p>
</dd>
<dt class="field-even">raises ValueError<span class="colon">:</span></dt>
<dd class="field-even"><p>If invalid arguments are provided to the methods.</p>
</dd>
<dt class="field-odd">raises TypeError<span class="colon">:</span></dt>
<dd class="field-odd"><p>If pose options are not of the expected type.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>run</cite> method of the <cite>ColabFold</cite> class:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">colabfold</span> <span class="kn">import</span> <span class="n">ColabFold</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the ColabFold class</span>
<span class="n">colabfold</span> <span class="o">=</span> <span class="n">ColabFold</span><span class="p">()</span>

<span class="c1"># Run the prediction process</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">colabfold</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;inference.num_designs=10&quot;</span><span class="p">,</span>
    <span class="n">pose_options</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;inference.input_pdb=&#39;input.pdb&#39;&quot;</span><span class="p">],</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id15">
<h6>Further Details<a class="headerlink" href="#id15" title="Link to this heading"></a></h6>
<blockquote>
<div><ul class="simple">
<li><p><strong>Batch Processing:</strong> The method can handle large sets of input sequences by batching them into smaller groups, which helps in managing computational resources effectively.</p></li>
<li><p><strong>Overwrite Handling:</strong> If <cite>overwrite</cite> is set to True, the method will clean up previous results, ensuring that the new predictions do not get mixed up with old data.</p></li>
<li><p><strong>Job Starter Configuration:</strong> The method allows for flexible job management by accepting a <cite>JobStarter</cite> instance. If not provided, it uses the default job starter associated with the poses.</p></li>
<li><p><strong>Score Collection:</strong> The method gathers the prediction scores and relevant data into a pandas DataFrame, facilitating easy analysis and integration with other ProtFlow components.</p></li>
<li><p><strong>Error Handling:</strong> Robust error handling is incorporated to manage issues such as missing files or incorrect configurations, ensuring that the process can be debugged and verified efficiently.</p></li>
</ul>
</div></blockquote>
</section>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.colabfold.Colabfold.write_cmd">
<span class="sig-name descname"><span class="pre">write_cmd</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pose_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pose_options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/tools/colabfold.html#Colabfold.write_cmd"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.colabfold.Colabfold.write_cmd" title="Link to this definition"></a></dt>
<dd><section id="write-cmd-method">
<h5>write_cmd Method<a class="headerlink" href="#write-cmd-method" title="Link to this heading"></a></h5>
<p>The <cite>write_cmd</cite> method constructs the command string necessary to run the ColabFold script with the specified options and input files.</p>
<section id="id16">
<h6>Detailed Description<a class="headerlink" href="#id16" title="Link to this heading"></a></h6>
<p>This method generates the command string used to execute the ColabFold script. It incorporates various options and pose-specific parameters provided by the user.</p>
<dl class="field-list simple">
<dt class="field-odd">param pose_path<span class="colon">:</span></dt>
<dd class="field-odd"><p>Path to the input FASTA file.</p>
</dd>
<dt class="field-even">type pose_path<span class="colon">:</span></dt>
<dd class="field-even"><p>str</p>
</dd>
<dt class="field-odd">param output_dir<span class="colon">:</span></dt>
<dd class="field-odd"><p>Directory where the prediction outputs will be stored.</p>
</dd>
<dt class="field-even">type output_dir<span class="colon">:</span></dt>
<dd class="field-even"><p>str</p>
</dd>
<dt class="field-odd">param options<span class="colon">:</span></dt>
<dd class="field-odd"><p>Additional options for the ColabFold script. Defaults to None.</p>
</dd>
<dt class="field-even">type options<span class="colon">:</span></dt>
<dd class="field-even"><p>str, optional</p>
</dd>
<dt class="field-odd">param pose_options<span class="colon">:</span></dt>
<dd class="field-odd"><p>Specific options for handling pose-related parameters. Defaults to None.</p>
</dd>
<dt class="field-even">type pose_options<span class="colon">:</span></dt>
<dd class="field-even"><p>str, optional</p>
</dd>
<dt class="field-odd">returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The constructed command string.</p>
</dd>
<dt class="field-even">rtype<span class="colon">:</span></dt>
<dd class="field-even"><p>str</p>
</dd>
<dt class="field-odd">raises None<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>write_cmd</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Write the command to run ColabFold</span>
<span class="n">cmd</span> <span class="o">=</span> <span class="n">colabfold</span><span class="o">.</span><span class="n">write_cmd</span><span class="p">(</span><span class="n">pose_path</span><span class="o">=</span><span class="s1">&#39;/path/to/pose.fa&#39;</span><span class="p">,</span> <span class="n">output_dir</span><span class="o">=</span><span class="s1">&#39;/path/to/output_dir&#39;</span><span class="p">,</span> <span class="n">options</span><span class="o">=</span><span class="s1">&#39;--num_designs=10&#39;</span><span class="p">,</span> <span class="n">pose_options</span><span class="o">=</span><span class="s1">&#39;--input_pdb=input.pdb&#39;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
</dd></dl>

</section>
</section>
</dd></dl>

</section>
<section id="module-protflow.tools.esmfold">
<span id="protflow-tools-esmfold-module"></span><h2>protflow.tools.esmfold module<a class="headerlink" href="#module-protflow.tools.esmfold" title="Link to this heading"></a></h2>
<section id="esmfold-module">
<h3>ESMFold Module<a class="headerlink" href="#esmfold-module" title="Link to this heading"></a></h3>
<p>This module provides the functionality to integrate ESMFold within the ProtFlow framework. It offers tools to run ESMFold, handle its inputs and outputs, and process the resulting data in a structured and automated manner.</p>
<section id="id17">
<h4>Detailed Description<a class="headerlink" href="#id17" title="Link to this heading"></a></h4>
<p>The <cite>ESMFold</cite> class encapsulates the functionality necessary to execute ESMFold runs. It manages the configuration of paths to essential scripts and Python executables, sets up the environment, and handles the execution of folding processes. It also includes methods for collecting and processing output data, ensuring that the results are organized and accessible for further analysis within the ProtFlow ecosystem.</p>
<p>The module is designed to streamline the integration of ESMFold into larger computational workflows. It supports the automatic setup of job parameters, execution of ESMFold commands, and parsing of output files into a structured DataFrame format. This facilitates subsequent data analysis and visualization steps.</p>
</section>
<section id="id18">
<h4>Usage<a class="headerlink" href="#id18" title="Link to this heading"></a></h4>
<p>To use this module, create an instance of the <cite>ESMFold</cite> class and invoke its <cite>run</cite> method with appropriate parameters. The module will handle the configuration, execution, and result collection processes. Detailed control over the folding process is provided through various parameters, allowing for customized runs tailored to specific research needs.</p>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize and use the <cite>ESMFold</cite> class within a ProtFlow pipeline on a SLURM based queueing system:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">esmfold</span> <span class="kn">import</span> <span class="n">ESMFold</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">SbatchArrayJobStarter</span><span class="p">(</span><span class="n">max_cores</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">gpus</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="c1"># 1 gpu per node, 10 nodes at once.</span>

<span class="c1"># Initialize the ESMFold class</span>
<span class="n">esmfold</span> <span class="o">=</span> <span class="n">ESMFold</span><span class="p">()</span>

<span class="c1"># Run the folding process</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">esmfold</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">num_batches</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;--additional_option=value&quot;</span><span class="p">,</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id19">
<h4>Further Details<a class="headerlink" href="#id19" title="Link to this heading"></a></h4>
<blockquote>
<div><ul class="simple">
<li><p>Edge Cases: The module handles various edge cases, such as empty pose lists and the need to overwrite previous results. It ensures robust error handling and logging for easier debugging and verification of the folding process.</p></li>
<li><p>Customizability: Users can customize the folding process through multiple parameters, including the number of batches, specific options for the ESMFold script, and options for handling pose-specific parameters.</p></li>
<li><p>Integration: The module seamlessly integrates with other components of the ProtFlow framework, leveraging shared configurations and data structures to provide a cohesive user experience.</p></li>
</ul>
</div></blockquote>
<p>This module is intended for researchers and developers who need to incorporate ESMFold into their protein design and analysis workflows. By automating many of the setup and execution steps, it allows users to focus on interpreting results and advancing their scientific inquiries.</p>
<p class="rubric">Notes</p>
<p>This module is part of the ProtFlow package and is designed to work in tandem with other components of the package, especially those related to job management in HPC environments.</p>
</section>
<section id="id20">
<h4>Authors<a class="headerlink" href="#id20" title="Link to this heading"></a></h4>
<p>Markus Braun, Adrian Tripp</p>
</section>
<section id="id21">
<h4>Version<a class="headerlink" href="#id21" title="Link to this heading"></a></h4>
<p>0.1.0</p>
</section>
</section>
<dl class="py class">
<dt class="sig sig-object py" id="protflow.tools.esmfold.ESMFold">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">protflow.tools.esmfold.</span></span><span class="sig-name descname"><span class="pre">ESMFold</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">python_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><span class="pre">JobStarter</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/tools/esmfold.html#ESMFold"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.esmfold.ESMFold" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="protflow.html#protflow.runners.Runner" title="protflow.runners.Runner"><code class="xref py py-class docutils literal notranslate"><span class="pre">Runner</span></code></a></p>
<section id="esmfold-class">
<h3>ESMFold Class<a class="headerlink" href="#esmfold-class" title="Link to this heading"></a></h3>
<p>The <cite>ESMFold</cite> class is a specialized class designed to facilitate the execution of ESMFold within the ProtFlow framework. It extends the <cite>Runner</cite> class and incorporates specific methods to handle the setup, execution, and data collection associated with ESMFold processes.</p>
<section id="id22">
<h4>Detailed Description<a class="headerlink" href="#id22" title="Link to this heading"></a></h4>
<p>The <cite>ESMFold</cite> class manages all aspects of running ESMFold simulations. It handles the configuration of necessary scripts and executables, prepares the environment for folding processes, and executes the folding commands. Additionally, it collects and processes the output data, organizing it into a structured format for further analysis.</p>
<dl class="simple">
<dt>Key functionalities include:</dt><dd><ul class="simple">
<li><p>Setting up paths to ESMFold scripts and Python executables.</p></li>
<li><p>Configuring job starter options, either automatically or manually.</p></li>
<li><p>Handling the execution of ESMFold commands with support for multiple batches.</p></li>
<li><p>Collecting and processing output data into a pandas DataFrame.</p></li>
<li><p>Preparing input FASTA files for ESMFold predictions and managing output directories.</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">rtype<span class="colon">:</span></dt>
<dd class="field-odd"><p>An instance of the <cite>ESMFold</cite> class, configured to run ESMFold processes and handle outputs efficiently.</p>
</dd>
<dt class="field-even">raises - FileNotFoundError<span class="colon">:</span></dt>
<dd class="field-even"><p>If required files or directories are not found during the execution process.:</p>
</dd>
<dt class="field-odd">raises - ValueError<span class="colon">:</span></dt>
<dd class="field-odd"><p>If invalid arguments are provided to the methods.:</p>
</dd>
<dt class="field-even">raises - KeyError<span class="colon">:</span></dt>
<dd class="field-even"><p>If forbidden options are included in the command options.:</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize and use the <cite>ESMFold</cite> class:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">esmfold</span> <span class="kn">import</span> <span class="n">ESMFold</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the ESMFold class</span>
<span class="n">esmfold</span> <span class="o">=</span> <span class="n">ESMFold</span><span class="p">()</span>

<span class="c1"># Run the folding process</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">esmfold</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">num_batches</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;--additional_option=value&quot;</span><span class="p">,</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id23">
<h4>Further Details<a class="headerlink" href="#id23" title="Link to this heading"></a></h4>
<blockquote>
<div><ul class="simple">
<li><p>Edge Cases: The class includes handling for various edge cases, such as empty pose lists, the need to overwrite previous results, and the presence of existing score files.</p></li>
<li><p>Customization: The class provides extensive customization options through its parameters, allowing users to tailor the folding process to their specific needs.</p></li>
<li><p>Integration: Seamlessly integrates with other ProtFlow components, leveraging shared configurations and data structures for a unified workflow.</p></li>
</ul>
</div></blockquote>
<p>The ESMFold class is intended for researchers and developers who need to perform ESMFold simulations as part of their protein design and analysis workflows. It simplifies the process, allowing users to focus on analyzing results and advancing their research.</p>
<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.esmfold.ESMFold.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">python_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><span class="pre">JobStarter</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/protflow/tools/esmfold.html#ESMFold.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.esmfold.ESMFold.__init__" title="Link to this definition"></a></dt>
<dd><p>Initialize the ESMFold class with necessary configurations.</p>
<p>This method sets up the ESMFold class, configuring paths to essential scripts and Python executables, and setting up the environment for executing ESMFold processes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>python_path</strong> (<em>str</em><em>, </em><em>optional</em>) – The path to the Python executable used for running ESMFold. Defaults to the value specified in <cite>protflow.config.ESMFOLD_PYTHON_PATH</cite>.</p></li>
<li><p><strong>jobstarter</strong> (<a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><em>JobStarter</em></a><em>, </em><em>optional</em>) – An instance of the JobStarter class, which manages job execution. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If no path is set for the ESMFold scripts or Python executable.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize the <cite>ESMFold</cite> class:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">esmfold</span> <span class="kn">import</span> <span class="n">ESMFold</span>

<span class="c1"># Initialize the ESMFold class with default configurations</span>
<span class="n">esmfold</span> <span class="o">=</span> <span class="n">ESMFold</span><span class="p">()</span>

<span class="c1"># Initialize the ESMFold class with a custom Python path and jobstarter</span>
<span class="n">custom_python_path</span> <span class="o">=</span> <span class="s2">&quot;/path/to/custom/python&quot;</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>
<span class="n">esmfold</span> <span class="o">=</span> <span class="n">ESMFold</span><span class="p">(</span><span class="n">python_path</span><span class="o">=</span><span class="n">custom_python_path</span><span class="p">,</span> <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Configuration:</strong> This method sets the script path for ESMFold inference and the Python path. If these are not correctly set in the configuration, a ValueError is raised.</p></li>
<li><p><strong>Initialization:</strong> The method initializes necessary attributes, including the script path, Python path, jobstarter, and other configurations needed for running ESMFold processes.</p></li>
</ul>
</dd>
</dl>
<p>This method prepares the ESMFold class for running folding simulations, ensuring that all necessary configurations and paths are correctly set up.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.esmfold.ESMFold.collect_scores">
<span class="sig-name descname"><span class="pre">collect_scores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">work_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scorefile</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/protflow/tools/esmfold.html#ESMFold.collect_scores"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.esmfold.ESMFold.collect_scores" title="Link to this definition"></a></dt>
<dd><p>Collect and process the scores from ESMFold output.</p>
<p>This method collects the JSON and PDB output files from ESMFold predictions, processes the data, and organizes it into a pandas DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>work_dir</strong> (<em>str</em>) – The working directory where ESMFold output files are stored.</p></li>
<li><p><strong>scorefile</strong> (<em>str</em>) – The path to the JSON file where the collected scores will be saved.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A DataFrame containing the collected scores and corresponding file locations.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>collect_scores</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">esmfold</span> <span class="kn">import</span> <span class="n">ESMFold</span>

<span class="c1"># Initialize the ESMFold class</span>
<span class="n">esmfold</span> <span class="o">=</span> <span class="n">ESMFold</span><span class="p">()</span>

<span class="c1"># Collect scores from ESMFold output</span>
<span class="n">scores_df</span> <span class="o">=</span> <span class="n">esmfold</span><span class="o">.</span><span class="n">collect_scores</span><span class="p">(</span>
    <span class="n">work_dir</span><span class="o">=</span><span class="s2">&quot;/path/to/work_dir&quot;</span><span class="p">,</span>
    <span class="n">scorefile</span><span class="o">=</span><span class="s2">&quot;/path/to/scorefile.json&quot;</span>
<span class="p">)</span>

<span class="c1"># Access the collected scores</span>
<span class="nb">print</span><span class="p">(</span><span class="n">scores_df</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Output Collection:</strong> The method scans the working directory for JSON and PDB output files, reads the JSON files into a DataFrame, and merges it with the locations of the PDB files.</p></li>
<li><p><strong>Data Organization:</strong> The method organizes the collected data into a structured DataFrame, making it accessible for further analysis and ensuring that the scores are saved to the specified JSON file.</p></li>
<li><p><strong>Output Management:</strong> The method also ensures that the temporary directories and files used during the prediction process are cleaned up, maintaining a tidy working environment.</p></li>
</ul>
</dd>
</dl>
<p>This method is designed to streamline the collection and processing of ESMFold output data, ensuring that the results are organized and accessible for further analysis.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.esmfold.ESMFold.prep_fastas_for_prediction">
<span class="sig-name descname"><span class="pre">prep_fastas_for_prediction</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fasta_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_filenum</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/protflow/tools/esmfold.html#ESMFold.prep_fastas_for_prediction"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.esmfold.ESMFold.prep_fastas_for_prediction" title="Link to this definition"></a></dt>
<dd><p>Prepare input FASTA files for ESMFold predictions.</p>
<p>This method splits the input poses into the specified number of batches, prepares the FASTA files, and writes them to the specified directory for ESMFold predictions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>poses</strong> (<em>list</em><em>[</em><em>str</em><em>]</em>) – List of paths to FASTA files.</p></li>
<li><p><strong>fasta_dir</strong> (<em>str</em>) – Directory to which the new FASTA files should be written.</p></li>
<li><p><strong>max_filenum</strong> (<em>int</em>) – Maximum number of FASTA files to be written.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of paths to the prepared FASTA files.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[str]</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>parse_fastas_for_prediction</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">esmfold</span> <span class="kn">import</span> <span class="n">ESMFold</span>

<span class="c1"># Initialize the ESMFold class</span>
<span class="n">esmfold</span> <span class="o">=</span> <span class="n">ESMFold</span><span class="p">()</span>

<span class="c1"># Prepare FASTA files for prediction</span>
<span class="n">fasta_paths</span> <span class="o">=</span> <span class="n">esmfold</span><span class="o">.</span><span class="n">parse_fastas_for_prediction</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;pose1.fa&quot;</span><span class="p">,</span> <span class="s2">&quot;pose2.fa&quot;</span><span class="p">,</span> <span class="s2">&quot;pose3.fa&quot;</span><span class="p">],</span>
    <span class="n">fasta_dir</span><span class="o">=</span><span class="s2">&quot;/path/to/fasta_dir&quot;</span><span class="p">,</span>
    <span class="n">max_filenum</span><span class="o">=</span><span class="mi">2</span>
<span class="p">)</span>

<span class="c1"># Access the prepared FASTA files</span>
<span class="nb">print</span><span class="p">(</span><span class="n">fasta_paths</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Input Preparation:</strong> The method merges and splits the input FASTA files into the specified number of batches. It ensures that the FASTA files are correctly formatted and written to the specified directory.</p></li>
<li><p><strong>Customization:</strong> Users can specify the maximum number of FASTA files to be created, allowing for flexibility in managing input data for parallel processing.</p></li>
<li><p><strong>Output Management:</strong> The method returns a list of paths to the newly created FASTA files, which are ready for ESMFold predictions.</p></li>
</ul>
</dd>
</dl>
<p>This method is designed to facilitate the preparation of input data for ESMFold, ensuring that the input FASTA files are organized and ready for processing.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.esmfold.ESMFold.run">
<span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><span class="pre">JobStarter</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_batches</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span></span><a class="reference internal" href="_modules/protflow/tools/esmfold.html#ESMFold.run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.esmfold.ESMFold.run" title="Link to this definition"></a></dt>
<dd><p>Execute the ESMFold process with given poses and jobstarter configuration.</p>
<p>This method sets up and runs the ESMFold process using the provided poses and jobstarter object. It handles the configuration, execution, and collection of output data, ensuring that the results are organized and accessible for further analysis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>poses</strong> (<a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><em>Poses</em></a><em>, </em><em>optional</em>) – The Poses object containing the protein structures. Defaults to None.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – A prefix used to name and organize the output files.</p></li>
<li><p><strong>jobstarter</strong> (<a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><em>JobStarter</em></a><em>, </em><em>optional</em>) – An instance of the JobStarter class, which manages job execution. Defaults to None.</p></li>
<li><p><strong>options</strong> (<em>str</em><em>, </em><em>optional</em>) – Additional options for the ESMFold script. Defaults to None.</p></li>
<li><p><strong>overwrite</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, overwrite existing output files. Defaults to False.</p></li>
<li><p><strong>num_batches</strong> (<em>int</em><em>, </em><em>optional</em>) – The number of batches to split the input poses into for parallel processing. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An instance of the RunnerOutput class, containing the processed poses and results of the ESMFold process.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="protflow.html#protflow.runners.RunnerOutput" title="protflow.runners.RunnerOutput">RunnerOutput</a></p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>FileNotFoundError</strong> – If required files or directories are not found during the execution process.</p></li>
<li><p><strong>ValueError</strong> – If invalid arguments are provided to the method.</p></li>
<li><p><strong>KeyError</strong> – If forbidden options are included in the command options.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>run</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">esmfold</span> <span class="kn">import</span> <span class="n">ESMFold</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the ESMFold class</span>
<span class="n">esmfold</span> <span class="o">=</span> <span class="n">ESMFold</span><span class="p">()</span>

<span class="c1"># Run the folding process</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">esmfold</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">num_batches</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;--additional_option=value&quot;</span><span class="p">,</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Setup and Execution:</strong> The method ensures that the environment is correctly set up, directories are prepared, and necessary commands are constructed and executed.</p></li>
<li><p><strong>Input Preparation:</strong> The method prepares input FASTA files by splitting the input poses into the specified number of batches, optimizing the use of parallel computing resources.</p></li>
<li><p><strong>Output Management:</strong> The method handles the collection and processing of output data, ensuring that results are organized into a pandas DataFrame and accessible for further analysis.</p></li>
<li><p><strong>Customization:</strong> Extensive customization options are provided through parameters, allowing users to tailor the folding process to their specific needs.</p></li>
</ul>
</dd>
</dl>
<p>This method is designed to streamline the execution of ESMFold processes within the ProtFlow framework, making it easier for researchers and developers to perform and analyze folding simulations.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.esmfold.ESMFold.write_cmd">
<span class="sig-name descname"><span class="pre">write_cmd</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pose_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/tools/esmfold.html#ESMFold.write_cmd"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.esmfold.ESMFold.write_cmd" title="Link to this definition"></a></dt>
<dd><p>Write the command to run ESMFold with the given parameters.</p>
<p>This method constructs the command line instruction needed to execute the ESMFold script with the specified pose path, output directory, and additional options.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pose_path</strong> (<em>str</em>) – The path to the input FASTA file.</p></li>
<li><p><strong>output_dir</strong> (<em>str</em>) – The directory where the ESMFold outputs will be stored.</p></li>
<li><p><strong>options</strong> (<em>str</em><em>, </em><em>optional</em>) – Additional command-line options for the ESMFold script. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The constructed command line instruction for running ESMFold.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>write_cmd</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">esmfold</span> <span class="kn">import</span> <span class="n">ESMFold</span>

<span class="c1"># Initialize the ESMFold class</span>
<span class="n">esmfold</span> <span class="o">=</span> <span class="n">ESMFold</span><span class="p">()</span>

<span class="c1"># Write the ESMFold command</span>
<span class="n">cmd</span> <span class="o">=</span> <span class="n">esmfold</span><span class="o">.</span><span class="n">write_cmd</span><span class="p">(</span>
    <span class="n">pose_path</span><span class="o">=</span><span class="s2">&quot;/path/to/pose.fa&quot;</span><span class="p">,</span>
    <span class="n">output_dir</span><span class="o">=</span><span class="s2">&quot;/path/to/output_dir&quot;</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;--additional_option=value&quot;</span>
<span class="p">)</span>

<span class="c1"># Access the command</span>
<span class="nb">print</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Command Construction:</strong> The method parses additional options and constructs the full command line instruction needed to run the ESMFold script. It ensures that the mandatory arguments, such as the FASTA file and output directory, are correctly included.</p></li>
<li><p><strong>Customization:</strong> Users can specify additional command-line options to customize the execution of the ESMFold script, allowing for flexibility in configuring the prediction process.</p></li>
</ul>
</dd>
</dl>
<p>This method is designed to facilitate the execution of ESMFold by constructing the necessary command line instructions, ensuring that all required parameters and options are included.</p>
</dd></dl>

</section>
</section>
</dd></dl>

</section>
<section id="module-protflow.tools.ligandmpnn">
<span id="protflow-tools-ligandmpnn-module"></span><h2>protflow.tools.ligandmpnn module<a class="headerlink" href="#module-protflow.tools.ligandmpnn" title="Link to this heading"></a></h2>
<section id="ligandmpnn-module">
<h3>LigandMPNN Module<a class="headerlink" href="#ligandmpnn-module" title="Link to this heading"></a></h3>
<p>This module provides the functionality to integrate LigandMPNN within the ProtFlow framework. It offers tools to run LigandMPNN, handle its inputs and outputs, and process the resulting data in a structured and automated manner.</p>
<section id="id24">
<h4>Detailed Description<a class="headerlink" href="#id24" title="Link to this heading"></a></h4>
<p>The <cite>LigandMPNN</cite> class encapsulates the functionality necessary to execute LigandMPNN runs. It manages the configuration of paths to essential scripts and Python executables, sets up the environment, and handles the execution of the diffusion processes. It also includes methods for collecting and processing output data, ensuring that the results are organized and accessible for further analysis within the ProtFlow ecosystem.</p>
<p>The module is designed to streamline the integration of LigandMPNN into larger computational workflows. It supports the automatic setup of job parameters, execution of LigandMPNN commands, and parsing of output files into a structured DataFrame format. This facilitates subsequent data analysis and visualization steps.</p>
</section>
<section id="id25">
<h4>Usage<a class="headerlink" href="#id25" title="Link to this heading"></a></h4>
<p>To use this module, create an instance of the <cite>LigandMPNN</cite> class and invoke its <cite>run</cite> method with appropriate parameters. The module will handle the configuration, execution, and result collection processes. Detailed control over the process is provided through various parameters, allowing for customized runs tailored to specific research needs.</p>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize and use the <cite>LigandMPNN</cite> class within a ProtFlow pipeline:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">ligandmpnn</span> <span class="kn">import</span> <span class="n">LigandMPNN</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the LigandMPNN class</span>
<span class="n">ligandmpnn</span> <span class="o">=</span> <span class="n">LigandMPNN</span><span class="p">()</span>

<span class="c1"># Run the diffusion process</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">ligandmpnn</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">nseq</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
    <span class="n">model_type</span><span class="o">=</span><span class="s2">&quot;ligand_mpnn&quot;</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;some_option=some_value&quot;</span><span class="p">,</span>
    <span class="n">pose_options</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;pose_option=pose_value&quot;</span><span class="p">],</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id26">
<h4>Further Details<a class="headerlink" href="#id26" title="Link to this heading"></a></h4>
<ul class="simple">
<li><p>Edge Cases: The module handles various edge cases, such as empty pose lists and the need to overwrite previous results. It ensures robust error handling and logging for easier debugging and verification of the process.</p></li>
<li><p>Customizability: Users can customize the process through multiple parameters, including the number of sequences, specific options for the LigandMPNN script, and options for handling pose-specific parameters.</p></li>
<li><p>Integration: The module seamlessly integrates with other components of the ProtFlow framework, leveraging shared configurations and data structures to provide a cohesive user experience.</p></li>
</ul>
<p>This module is intended for researchers and developers who need to incorporate LigandMPNN into their protein design and analysis workflows. By automating many of the setup and execution steps, it allows users to focus on interpreting results and advancing their scientific inquiries.</p>
<p class="rubric">Notes</p>
<p>This module is part of the ProtFlow package and is designed to work in tandem with other components of the package, especially those related to job management in HPC environments.</p>
</section>
<section id="author">
<h4>Author<a class="headerlink" href="#author" title="Link to this heading"></a></h4>
<p>Markus Braun, Adrian Tripp</p>
</section>
<section id="id27">
<h4>Version<a class="headerlink" href="#id27" title="Link to this heading"></a></h4>
<p>0.1.0</p>
</section>
</section>
<dl class="py class">
<dt class="sig sig-object py" id="protflow.tools.ligandmpnn.LigandMPNN">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">protflow.tools.ligandmpnn.</span></span><span class="sig-name descname"><span class="pre">LigandMPNN</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">script_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">python_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><span class="pre">JobStarter</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/tools/ligandmpnn.html#LigandMPNN"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.ligandmpnn.LigandMPNN" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="protflow.html#protflow.runners.Runner" title="protflow.runners.Runner"><code class="xref py py-class docutils literal notranslate"><span class="pre">Runner</span></code></a></p>
<section id="ligandmpnn-class">
<h3>LigandMPNN Class<a class="headerlink" href="#ligandmpnn-class" title="Link to this heading"></a></h3>
<p>The <cite>LigandMPNN</cite> class provides the necessary methods to execute LigandMPNN runs within the ProtFlow framework. This class is responsible for managing the configuration, execution, and output processing of LigandMPNN tasks.</p>
<section id="id28">
<h4>Detailed Description<a class="headerlink" href="#id28" title="Link to this heading"></a></h4>
<p>The <cite>LigandMPNN</cite> class integrates LigandMPNN into the ProtFlow pipeline by setting up the environment, running the diffusion process, and collecting the results. It ensures that the inputs and outputs are handled efficiently, making the data readily available for further analysis.</p>
<p>Key Features:
- Manages paths to essential scripts and executables.
- Configures and executes LigandMPNN processes.
- Collects and processes output data into a structured DataFrame format.
- Handles various edge cases and supports custom configurations through multiple parameters.</p>
</section>
<section id="id29">
<h4>Usage<a class="headerlink" href="#id29" title="Link to this heading"></a></h4>
<p>To use this class, initialize it with the appropriate script and Python paths, along with an optional job starter. The main functionality is provided through the <cite>run</cite> method, which requires parameters such as poses, prefix, and additional options for customization.</p>
<p class="rubric">Example</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">ligandmpnn</span> <span class="kn">import</span> <span class="n">LigandMPNN</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the LigandMPNN class</span>
<span class="n">ligandmpnn</span> <span class="o">=</span> <span class="n">LigandMPNN</span><span class="p">()</span>

<span class="c1"># Run the diffusion process</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">ligandmpnn</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">nseq</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
    <span class="n">model_type</span><span class="o">=</span><span class="s2">&quot;ligand_mpnn&quot;</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;some_option=some_value&quot;</span><span class="p">,</span>
    <span class="n">pose_options</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;pose_option=pose_value&quot;</span><span class="p">],</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
<p class="rubric">Notes</p>
<p>This class is designed to work within the ProtFlow framework and assumes that the necessary configurations and dependencies are properly set up. It leverages shared data structures and configurations from ProtFlow to provide a seamless integration experience.</p>
</section>
<section id="id30">
<h4>Author<a class="headerlink" href="#id30" title="Link to this heading"></a></h4>
<p>Markus Braun, Adrian Tripp</p>
</section>
<section id="id31">
<h4>Version<a class="headerlink" href="#id31" title="Link to this heading"></a></h4>
<p>0.1.0</p>
<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.ligandmpnn.LigandMPNN.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">script_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">python_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><span class="pre">JobStarter</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/protflow/tools/ligandmpnn.html#LigandMPNN.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.ligandmpnn.LigandMPNN.__init__" title="Link to this definition"></a></dt>
<dd><p>Initializes the LigandMPNN class.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>script_path</strong> (<em>str</em><em>, </em><em>optional</em>) – The path to the LigandMPNN script. Defaults to the configured script path in ProtFlow.</p></li>
<li><p><strong>python_path</strong> (<em>str</em><em>, </em><em>optional</em>) – The path to the Python executable to run the LigandMPNN script. Defaults to the configured Python path in ProtFlow.</p></li>
<li><p><strong>jobstarter</strong> (<a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><em>JobStarter</em></a><em>, </em><em>optional</em>) – An instance of the JobStarter class to manage job submissions. If not provided, it will use the default job starter configuration.</p></li>
</ul>
</dd>
</dl>
<section id="id32">
<h5>Detailed Description<a class="headerlink" href="#id32" title="Link to this heading"></a></h5>
<p>The <cite>__init__</cite> method sets up the necessary paths and configurations for running LigandMPNN. It searches for the provided script and Python
paths to ensure they are correct and sets them as instance attributes. Additionally, it initializes the job starter, which manages the execution
of jobs in high-performance computing (HPC) environments. This method ensures that all configurations are correctly set up before running any
LigandMPNN tasks.</p>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.ligandmpnn.LigandMPNN.check_for_batch_run">
<span class="sig-name descname"><span class="pre">check_for_batch_run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pose_options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pose_opt_cols</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/tools/ligandmpnn.html#LigandMPNN.check_for_batch_run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.ligandmpnn.LigandMPNN.check_for_batch_run" title="Link to this definition"></a></dt>
<dd><p>Checks if LigandMPNN can be run in batch mode.</p>
<p>This method determines whether the LigandMPNN process can be executed in batch mode. It does this by checking if pose-specific options are not provided and if only multi-residue columns are specified in the pose options.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pose_options</strong> (<em>str</em>) – Pose-specific options for the LigandMPNN script.</p></li>
<li><p><strong>pose_opt_cols</strong> (<em>dict</em>) – Dictionary of pose-specific options for the LigandMPNN script.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if LigandMPNN can be run in batch mode, False otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>check_for_batch_run</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Initialize the LigandMPNN class</span>
<span class="n">ligandmpnn</span> <span class="o">=</span> <span class="n">LigandMPNN</span><span class="p">()</span>

<span class="c1"># Check for batch run</span>
<span class="n">can_batch_run</span> <span class="o">=</span> <span class="n">ligandmpnn</span><span class="o">.</span><span class="n">check_for_batch_run</span><span class="p">(</span>
    <span class="n">pose_options</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">pose_opt_cols</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;fixed_residues&quot;</span><span class="p">:</span> <span class="s2">&quot;fixed_res_col&quot;</span><span class="p">}</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">can_batch_run</span><span class="p">)</span>  <span class="c1"># Outputs: True or False</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Batch Mode Check:</strong> The method checks if the <cite>pose_options</cite> is None and if the <cite>pose_opt_cols</cite> contains only multi-residue columns, which are necessary for batch processing.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.ligandmpnn.LigandMPNN.collect_scores">
<span class="sig-name descname"><span class="pre">collect_scores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">work_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_seq_threaded_pdbs_as_pose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">preserve_original_output</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/protflow/tools/ligandmpnn.html#LigandMPNN.collect_scores"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.ligandmpnn.LigandMPNN.collect_scores" title="Link to this definition"></a></dt>
<dd><p>Collects scores from the LigandMPNN output.</p>
<p>This method processes the output files generated by LigandMPNN, including multi-sequence FASTA files and PDB files. It reads, renames, and organizes these files into a structured DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>work_dir</strong> (<em>str</em>) – The directory where LigandMPNN output files are located.</p></li>
<li><p><strong>return_seq_threaded_pdbs_as_pose</strong> (<em>bool</em>) – If True, replaces FASTA files with sequence-threaded PDB files as poses.</p></li>
<li><p><strong>preserve_original_output</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, preserves the original output files. Defaults to True.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A DataFrame containing the collected scores and relevant data from the LigandMPNN output.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>FileNotFoundError</strong> – If required output files are not found in the specified directory.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>collect_scores</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Initialize the LigandMPNN class</span>
<span class="n">ligandmpnn</span> <span class="o">=</span> <span class="n">LigandMPNN</span><span class="p">()</span>

<span class="c1"># Collect scores from the output directory</span>
<span class="n">scores</span> <span class="o">=</span> <span class="n">ligandmpnn</span><span class="o">.</span><span class="n">collect_scores</span><span class="p">(</span>
    <span class="n">work_dir</span><span class="o">=</span><span class="s2">&quot;/path/to/output&quot;</span><span class="p">,</span>
    <span class="n">return_seq_threaded_pdbs_as_pose</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">preserve_original_output</span><span class="o">=</span><span class="kc">False</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span>  <span class="c1"># Outputs the collected scores DataFrame</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Output Processing:</strong> The method reads and parses multi-sequence FASTA files, converts sequences into a structured dictionary, and writes new FASTA files if necessary.</p></li>
<li><p><strong>File Management:</strong> Original output files are copied to dedicated directories, and new files are generated and organized for easy access. Optionally, original files can be preserved or deleted based on the <cite>preserve_original_output</cite> parameter.</p></li>
<li><p><strong>Error Handling:</strong> The method includes checks to ensure that required output files are present, raising errors if files are missing or paths are incorrect.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.ligandmpnn.LigandMPNN.multi_cols_only">
<span class="sig-name descname"><span class="pre">multi_cols_only</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pose_opt_cols</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="reference internal" href="_modules/protflow/tools/ligandmpnn.html#LigandMPNN.multi_cols_only"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.ligandmpnn.LigandMPNN.multi_cols_only" title="Link to this definition"></a></dt>
<dd><p>checks if only multi_res cols are in pose_opt_cols dict. Only _multi arguments can be used for ligandmpnn_batch runs.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.ligandmpnn.LigandMPNN.parse_pose_opt_cols">
<span class="sig-name descname"><span class="pre">parse_pose_opt_cols</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pose_opt_cols</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">dict</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/protflow/tools/ligandmpnn.html#LigandMPNN.parse_pose_opt_cols"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.ligandmpnn.LigandMPNN.parse_pose_opt_cols" title="Link to this definition"></a></dt>
<dd><p>Parses pose-specific options columns into pose options formatted strings.</p>
<p>This method processes the <cite>pose_opt_cols</cite> dictionary and converts its contents into a format that can be used as part of the LigandMPNN pose options. It ensures that the options are properly structured and, if necessary, writes specific arguments into JSON files.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>poses</strong> (<a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><em>Poses</em></a>) – The Poses object containing the protein structures.</p></li>
<li><p><strong>output_dir</strong> (<em>str</em>) – The directory where JSON files for multi-residue options will be saved.</p></li>
<li><p><strong>pose_opt_cols</strong> (<em>dict</em><em>, </em><em>optional</em>) – Dictionary of pose-specific options for the LigandMPNN script. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A list of dictionaries containing the parsed pose options formatted as strings.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[dict]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If both fixed_residues and redesigned_residues are defined in pose_opt_cols, or if specified columns do not exist in poses.df.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>parse_pose_opt_cols</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Initialize the LigandMPNN class</span>
<span class="n">ligandmpnn</span> <span class="o">=</span> <span class="n">LigandMPNN</span><span class="p">()</span>

<span class="c1"># Example Poses object and pose_opt_cols</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">pose_opt_cols</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;bias_AA_per_residue&quot;</span><span class="p">:</span> <span class="s2">&quot;bias_col&quot;</span><span class="p">,</span>
    <span class="s2">&quot;fixed_residues&quot;</span><span class="p">:</span> <span class="s2">&quot;fixed_res_col&quot;</span>
<span class="p">}</span>

<span class="c1"># Parse pose options</span>
<span class="n">parsed_opts</span> <span class="o">=</span> <span class="n">ligandmpnn</span><span class="o">.</span><span class="n">parse_pose_opt_cols</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">output_dir</span><span class="o">=</span><span class="s2">&quot;/path/to/output&quot;</span><span class="p">,</span>
    <span class="n">pose_opt_cols</span><span class="o">=</span><span class="n">pose_opt_cols</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">parsed_opts</span><span class="p">)</span>  <span class="c1"># Outputs the parsed pose options</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Option Parsing:</strong> The method converts the <cite>pose_opt_cols</cite> dictionary into a list of strings formatted as pose options. It handles various types of options, including those that need to be written into JSON files and those that can be parsed directly from residue selections.</p></li>
<li><p><strong>JSON Directory Setup:</strong> If necessary, the method sets up a directory for storing JSON files that contain mappings for multi-residue options.</p></li>
<li><p><strong>Error Handling:</strong> The method includes checks to ensure that incompatible options are not specified simultaneously and that all specified columns exist in the poses DataFrame.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.ligandmpnn.LigandMPNN.run">
<span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><span class="pre">JobStarter</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nseq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model_type</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pose_options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">object</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fixed_res_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">design_res_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pose_opt_cols</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_seq_threaded_pdbs_as_pose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">preserve_original_output</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span></span><a class="reference internal" href="_modules/protflow/tools/ligandmpnn.html#LigandMPNN.run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.ligandmpnn.LigandMPNN.run" title="Link to this definition"></a></dt>
<dd><p>Execute the LigandMPNN process with given poses and jobstarter configuration.</p>
<p>This method sets up and runs the LigandMPNN process using the provided poses and jobstarter object. It handles the configuration, execution, and collection of output data, ensuring that the results are organized and accessible for further analysis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>poses</strong> (<a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><em>Poses</em></a>) – The Poses object containing the protein structures.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – A prefix used to name and organize the output files.</p></li>
<li><p><strong>jobstarter</strong> (<a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><em>JobStarter</em></a><em>, </em><em>optional</em>) – An instance of the JobStarter class, which manages job execution. Defaults to None.</p></li>
<li><p><strong>nseq</strong> (<em>int</em><em>, </em><em>optional</em>) – The number of sequences to generate for each input pose. Defaults to None.</p></li>
<li><p><strong>model_type</strong> (<em>str</em><em>, </em><em>optional</em>) – The type of model to use. Defaults to ‘ligand_mpnn’.</p></li>
<li><p><strong>options</strong> (<em>str</em><em>, </em><em>optional</em>) – Additional options for the LigandMPNN script. Defaults to None.</p></li>
<li><p><strong>pose_options</strong> (<em>object</em><em>, </em><em>optional</em>) – Pose-specific options for the LigandMPNN script. Defaults to None.</p></li>
<li><p><strong>fixed_res_col</strong> (<em>str</em><em>, </em><em>optional</em>) – Column name in the poses DataFrame specifying fixed residues. Defaults to None.</p></li>
<li><p><strong>design_res_col</strong> (<em>str</em><em>, </em><em>optional</em>) – Column name in the poses DataFrame specifying residues to be redesigned. Defaults to None.</p></li>
<li><p><strong>pose_opt_cols</strong> (<em>dict</em><em>, </em><em>optional</em>) – Dictionary of pose-specific options for the LigandMPNN script. Defaults to None.</p></li>
<li><p><strong>return_seq_threaded_pdbs_as_pose</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, return sequence-threaded PDBs as poses. Defaults to False.</p></li>
<li><p><strong>preserve_original_output</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, preserve the original output files. Defaults to True.</p></li>
<li><p><strong>overwrite</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, overwrite existing output files. Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The updated Poses object containing the results of the LigandMPNN process.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses">Poses</a></p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>FileNotFoundError</strong> – If required files or directories are not found during the execution process.</p></li>
<li><p><strong>ValueError</strong> – If invalid arguments are provided to the method.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>run</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">ligandmpnn</span> <span class="kn">import</span> <span class="n">LigandMPNN</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the LigandMPNN class</span>
<span class="n">ligandmpnn</span> <span class="o">=</span> <span class="n">LigandMPNN</span><span class="p">()</span>

<span class="c1"># Run the diffusion process</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">ligandmpnn</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">nseq</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
    <span class="n">model_type</span><span class="o">=</span><span class="s2">&quot;ligand_mpnn&quot;</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;some_option=some_value&quot;</span><span class="p">,</span>
    <span class="n">pose_options</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;pose_option=pose_value&quot;</span><span class="p">],</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Setup and Execution:</strong> The method ensures that the environment is correctly set up, directories are prepared, and necessary commands are constructed and executed.</p></li>
<li><p><strong>Output Management:</strong> The method handles the collection and processing of output data, ensuring that results are organized and accessible for further analysis.</p></li>
<li><p><strong>Customization:</strong> Extensive customization options are provided through parameters, allowing users to tailor the process to their specific needs.</p></li>
</ul>
</dd>
</dl>
<p>This method is designed to streamline the execution of LigandMPNN processes within the ProtFlow framework, making it easier for researchers and developers to perform and analyze protein design simulations.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.ligandmpnn.LigandMPNN.setup_batch_run">
<span class="sig-name descname"><span class="pre">setup_batch_run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmds</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_batches</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/protflow/tools/ligandmpnn.html#LigandMPNN.setup_batch_run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.ligandmpnn.LigandMPNN.setup_batch_run" title="Link to this definition"></a></dt>
<dd><p>Concatenates commands for MPNN into batches so that MPNN does not have to be loaded individually for each PDB file.</p>
<p>This method prepares the LigandMPNN commands for batch execution. It concatenates the commands into batches to optimize the running process by reducing the overhead of loading the MPNN model multiple times.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cmds</strong> (<em>list</em><em>[</em><em>str</em><em>]</em>) – A list of commands to run LigandMPNN.</p></li>
<li><p><strong>num_batches</strong> (<em>int</em>) – The number of batches to split the commands into.</p></li>
<li><p><strong>output_dir</strong> (<em>str</em>) – The directory where the batch input JSON files will be saved.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A list of concatenated batch commands.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[str]</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>setup_batch_run</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Initialize the LigandMPNN class</span>
<span class="n">ligandmpnn</span> <span class="o">=</span> <span class="n">LigandMPNN</span><span class="p">()</span>

<span class="c1"># Example commands</span>
<span class="n">cmds</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;/path/to/python /path/to/run.py --option1=value1 --pdb_path=path1.pdb&quot;</span><span class="p">,</span>
    <span class="s2">&quot;/path/to/python /path/to/run.py --option2=value2 --pdb_path=path2.pdb&quot;</span><span class="p">,</span>
    <span class="c1"># More commands...</span>
<span class="p">]</span>

<span class="c1"># Setup batch run</span>
<span class="n">batch_cmds</span> <span class="o">=</span> <span class="n">ligandmpnn</span><span class="o">.</span><span class="n">setup_batch_run</span><span class="p">(</span>
    <span class="n">cmds</span><span class="o">=</span><span class="n">cmds</span><span class="p">,</span>
    <span class="n">num_batches</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
    <span class="n">output_dir</span><span class="o">=</span><span class="s2">&quot;/path/to/output&quot;</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">batch_cmds</span><span class="p">)</span>  <span class="c1"># Outputs the batch commands</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Batch Command Setup:</strong> The method splits the provided commands into sublists based on the number of batches. It then processes each sublist to handle multi-residue options and generate corresponding JSON files.</p></li>
<li><p><strong>JSON Directory:</strong> The method sets up a directory for storing JSON files that contain mappings for multi-residue options.</p></li>
<li><p><strong>Command Concatenation:</strong> Each command sublist is processed to extract and convert multi-residue options into JSON files, which are then referenced in the batch commands.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.ligandmpnn.LigandMPNN.write_cmd">
<span class="sig-name descname"><span class="pre">write_cmd</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pose_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nseq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pose_options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/tools/ligandmpnn.html#LigandMPNN.write_cmd"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.ligandmpnn.LigandMPNN.write_cmd" title="Link to this definition"></a></dt>
<dd><p>Writes the command to run ligandmpnn.py.</p>
<p>This method constructs the command necessary to run the LigandMPNN script, incorporating various options and parameters. It ensures that the command is correctly formatted and includes all required arguments.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pose_path</strong> (<em>str</em>) – The path to the input PDB file for the pose.</p></li>
<li><p><strong>output_dir</strong> (<em>str</em>) – The directory where the output files will be saved.</p></li>
<li><p><strong>model</strong> (<em>str</em>) – The type of model to use (e.g., “ligand_mpnn”).</p></li>
<li><p><strong>nseq</strong> (<em>int</em>) – The number of sequences to generate for each input pose. Defaults to 1.</p></li>
<li><p><strong>options</strong> (<em>str</em>) – Additional options for the LigandMPNN script.</p></li>
<li><p><strong>pose_options</strong> (<em>str</em>) – Pose-specific options for the LigandMPNN script.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The constructed command string to run LigandMPNN.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the specified model is not one of the available models.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>write_cmd</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Initialize the LigandMPNN class</span>
<span class="n">ligandmpnn</span> <span class="o">=</span> <span class="n">LigandMPNN</span><span class="p">()</span>

<span class="c1"># Write the command</span>
<span class="n">cmd</span> <span class="o">=</span> <span class="n">ligandmpnn</span><span class="o">.</span><span class="n">write_cmd</span><span class="p">(</span>
    <span class="n">pose_path</span><span class="o">=</span><span class="s2">&quot;path/to/input.pdb&quot;</span><span class="p">,</span>
    <span class="n">output_dir</span><span class="o">=</span><span class="s2">&quot;path/to/output&quot;</span><span class="p">,</span>
    <span class="n">model</span><span class="o">=</span><span class="s2">&quot;ligand_mpnn&quot;</span><span class="p">,</span>
    <span class="n">nseq</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;some_option=some_value&quot;</span><span class="p">,</span>
    <span class="n">pose_options</span><span class="o">=</span><span class="s2">&quot;pose_option=pose_value&quot;</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>  <span class="c1"># Outputs the constructed command string</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Model Validation:</strong> The method checks if the specified model is among the available models and raises an error if it is not.</p></li>
<li><p><strong>Option Parsing:</strong> The method parses generic options and pose-specific options, ensuring that necessary safety checks and defaults are applied.</p></li>
<li><p><strong>Command Construction:</strong> The method assembles the final command string, including paths, model checkpoints, options, and other necessary parameters.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="protflow.tools.ligandmpnn.parse_residues">
<span class="sig-prename descclassname"><span class="pre">protflow.tools.ligandmpnn.</span></span><span class="sig-name descname"><span class="pre">parse_residues</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">residues</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">object</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">str</span></span></span><a class="reference internal" href="_modules/protflow/tools/ligandmpnn.html#parse_residues"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.ligandmpnn.parse_residues" title="Link to this definition"></a></dt>
<dd><p>Parses residues from either ResidueSelection object, list, or MPNN-formatted string into MPNN-formatted string.</p>
<p>This function converts the input residues into a format compatible with MPNN. It supports conversion from ResidueSelection objects, comma-separated strings, and lists of residues.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>residues</strong> (<em>object</em>) – The input residues to be parsed. This can be a ResidueSelection object, a comma-separated string, or a list of residues.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The residues formatted as a string compatible with MPNN.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the input type is not supported (i.e., not a str or ResidueSelection).</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>parse_residues</cite> function:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.residues</span> <span class="kn">import</span> <span class="n">ResidueSelection</span>

<span class="c1"># Example ResidueSelection object</span>
<span class="n">residues</span> <span class="o">=</span> <span class="n">ResidueSelection</span><span class="p">([</span><span class="s2">&quot;A:10&quot;</span><span class="p">,</span> <span class="s2">&quot;A:20&quot;</span><span class="p">])</span>

<span class="c1"># Parse residues</span>
<span class="n">parsed_residues</span> <span class="o">=</span> <span class="n">parse_residues</span><span class="p">(</span><span class="n">residues</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">parsed_residues</span><span class="p">)</span>  <span class="c1"># Outputs: &quot;A:10 A:20&quot;</span>

<span class="c1"># Example string input</span>
<span class="n">residues_str</span> <span class="o">=</span> <span class="s2">&quot;A:10,A:20&quot;</span>
<span class="n">parsed_residues</span> <span class="o">=</span> <span class="n">parse_residues</span><span class="p">(</span><span class="n">residues_str</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">parsed_residues</span><span class="p">)</span>  <span class="c1"># Outputs: &quot;A:10 A:20&quot;</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>ResidueSelection Object:</strong> The function calls the <cite>to_string</cite> method of the ResidueSelection object to get the MPNN-formatted string.</p></li>
<li><p><strong>String Input:</strong> For comma-separated string inputs, the function splits the string by commas and joins the parts with spaces.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="protflow.tools.ligandmpnn.write_to_json">
<span class="sig-prename descclassname"><span class="pre">protflow.tools.ligandmpnn.</span></span><span class="sig-name descname"><span class="pre">write_to_json</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_dict</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">str</span></span></span><a class="reference internal" href="_modules/protflow/tools/ligandmpnn.html#write_to_json"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.ligandmpnn.write_to_json" title="Link to this definition"></a></dt>
<dd><p>Writes json serializable :input_dict: into file and returns path to file. Returns path to json file :output_path:</p>
</dd></dl>

</section>
<section id="module-protflow.tools.protein_edits">
<span id="protflow-tools-protein-edits-module"></span><h2>protflow.tools.protein_edits module<a class="headerlink" href="#module-protflow.tools.protein_edits" title="Link to this heading"></a></h2>
<section id="proteinedits-module">
<h3>ProteinEdits Module<a class="headerlink" href="#proteinedits-module" title="Link to this heading"></a></h3>
<p>This module provides the functionality to handle various protein editing tasks within the ProtFlow framework. It offers tools to add and remove protein chains, add sequences to proteins, and multimerize sequences in a structured and automated manner.</p>
<section id="id33">
<h4>Detailed Description<a class="headerlink" href="#id33" title="Link to this heading"></a></h4>
<p>The <cite>protein_edits</cite> module contains classes and methods designed to perform common protein editing operations. The <cite>ChainAdder</cite> class provides methods for adding chains to protein structures, including functionality for superimposing chains based on motifs or existing chains. The <cite>ChainRemover</cite> class allows for the removal of specified chains from protein structures. Additionally, methods for adding sequences to proteins and creating multimers from sequences are included, streamlining the process of preparing protein structures for further analysis.</p>
<p>The module integrates seamlessly with the ProtFlow ecosystem, leveraging shared configurations, job management capabilities, and data structures to provide a cohesive user experience. It supports automatic setup and execution of jobs, handling of input and output files, and robust error handling and logging.</p>
</section>
<section id="id34">
<h4>Usage<a class="headerlink" href="#id34" title="Link to this heading"></a></h4>
<p>To use this module, create instances of the <cite>ChainAdder</cite> or <cite>ChainRemover</cite> classes and invoke their respective methods with appropriate parameters. The module handles the configuration, execution, and result collection processes, allowing users to focus on interpreting the results.</p>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize and use the <cite>ChainAdder</cite> and <cite>ChainRemover</cite> classes within a ProtFlow pipeline:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">protein_edits</span> <span class="kn">import</span> <span class="n">ChainAdder</span><span class="p">,</span> <span class="n">ChainRemover</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the ChainAdder class</span>
<span class="n">chain_adder</span> <span class="o">=</span> <span class="n">ChainAdder</span><span class="p">(</span><span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">)</span>

<span class="c1"># Add a chain to the poses</span>
<span class="n">added_chains</span> <span class="o">=</span> <span class="n">chain_adder</span><span class="o">.</span><span class="n">add_chain</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">ref_col</span><span class="o">=</span><span class="s2">&quot;reference_column&quot;</span><span class="p">,</span>
    <span class="n">copy_chain</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Initialize the ChainRemover class</span>
<span class="n">chain_remover</span> <span class="o">=</span> <span class="n">ChainRemover</span><span class="p">(</span><span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">)</span>

<span class="c1"># Remove a chain from the poses</span>
<span class="n">removed_chains</span> <span class="o">=</span> <span class="n">chain_remover</span><span class="o">.</span><span class="n">remove_chains</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_2&quot;</span><span class="p">,</span>
    <span class="n">chains</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">],</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">added_chains</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">removed_chains</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id35">
<h4>Further Details<a class="headerlink" href="#id35" title="Link to this heading"></a></h4>
<ul class="simple">
<li><p>Edge Cases: The module handles various edge cases, such as missing chain specifications and the need to overwrite previous results. It ensures robust error handling and logging for easier debugging and verification of the process.</p></li>
<li><p>Customizability: Users can customize the processes through multiple parameters, including the chain to add or remove, sequence details for adding sequences, and the number of protomers for multimerization.</p></li>
<li><p>Integration: The module integrates with other components of the ProtFlow framework, leveraging shared configurations and data structures to provide a cohesive user experience.</p></li>
</ul>
<p>This module is intended for researchers and developers who need to incorporate protein editing tasks into their computational workflows. By automating many of the setup and execution steps, it allows users to focus on interpreting results and advancing their scientific inquiries.</p>
<p class="rubric">Notes</p>
<p>This module is part of the ProtFlow package and is designed to work in tandem with other components of the package, especially those related to job management in HPC environments.</p>
</section>
<section id="id36">
<h4>Author<a class="headerlink" href="#id36" title="Link to this heading"></a></h4>
<p>Markus Braun, Adrian Tripp</p>
</section>
<section id="id37">
<h4>Version<a class="headerlink" href="#id37" title="Link to this heading"></a></h4>
<p>0.1.0</p>
</section>
</section>
<dl class="py class">
<dt class="sig sig-object py" id="protflow.tools.protein_edits.ChainAdder">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">protflow.tools.protein_edits.</span></span><span class="sig-name descname"><span class="pre">ChainAdder</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">default_python</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'python3'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><span class="pre">JobStarter</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/tools/protein_edits.html#ChainAdder"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.protein_edits.ChainAdder" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="protflow.html#protflow.runners.Runner" title="protflow.runners.Runner"><code class="xref py py-class docutils literal notranslate"><span class="pre">Runner</span></code></a></p>
<section id="chainadder-class">
<h3>ChainAdder Class<a class="headerlink" href="#chainadder-class" title="Link to this heading"></a></h3>
<p>The <cite>ChainAdder</cite> class is a specialized class designed to facilitate the addition of chains to protein structures within the ProtFlow framework. It extends the <cite>Runner</cite> class and incorporates specific methods to handle the setup, execution, and data collection associated with chain addition processes.</p>
<section id="id38">
<h4>Detailed Description<a class="headerlink" href="#id38" title="Link to this heading"></a></h4>
<p>The <cite>ChainAdder</cite> class manages all aspects of adding chains to protein structures. It configures necessary scripts and executables, prepares the environment for the addition processes, and executes the required commands. Additionally, it collects and processes the output data, organizing it into a structured format for further analysis.</p>
<dl class="simple">
<dt>Key functionalities include:</dt><dd><ul class="simple">
<li><p>Setting up paths to chain addition scripts and Python executables.</p></li>
<li><p>Configuring job starter options, either automatically or manually.</p></li>
<li><p>Handling the execution of chain addition commands with support for superimposition on motifs or existing chains.</p></li>
<li><p>Collecting and processing output data into a structured format.</p></li>
<li><p>Providing methods for adding sequences to proteins and creating multimers from sequences.</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">rtype<span class="colon">:</span></dt>
<dd class="field-odd"><p>An instance of the <cite>ChainAdder</cite> class, configured to add chains to protein structures and handle outputs efficiently.</p>
</dd>
<dt class="field-even">raises - FileNotFoundError<span class="colon">:</span></dt>
<dd class="field-even"><p>If required files or directories are not found during the execution process.:</p>
</dd>
<dt class="field-odd">raises - ValueError<span class="colon">:</span></dt>
<dd class="field-odd"><p>If invalid arguments are provided to the methods.:</p>
</dd>
<dt class="field-even">raises - TypeError<span class="colon">:</span></dt>
<dd class="field-even"><p>If motifs or chains are not of the expected type.:</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize and use the <cite>ChainAdder</cite> class:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">protein_edits</span> <span class="kn">import</span> <span class="n">ChainAdder</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the ChainAdder class</span>
<span class="n">chain_adder</span> <span class="o">=</span> <span class="n">ChainAdder</span><span class="p">(</span><span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">)</span>

<span class="c1"># Add a chain to the poses</span>
<span class="n">added_chains</span> <span class="o">=</span> <span class="n">chain_adder</span><span class="o">.</span><span class="n">add_chain</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">ref_col</span><span class="o">=</span><span class="s2">&quot;reference_column&quot;</span><span class="p">,</span>
    <span class="n">copy_chain</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">added_chains</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id39">
<h4>Further Details<a class="headerlink" href="#id39" title="Link to this heading"></a></h4>
<blockquote>
<div><ul class="simple">
<li><p>Edge Cases: The class handles various edge cases, such as missing chain specifications and the need to overwrite previous results.</p></li>
<li><p>Customization: The class provides extensive customization options through its parameters, allowing users to tailor the chain addition process to their specific needs.</p></li>
<li><p>Integration: Seamlessly integrates with other ProtFlow components, leveraging shared configurations and data structures for a unified workflow.</p></li>
</ul>
</div></blockquote>
<p>The ChainAdder class is intended for researchers and developers who need to add chains to protein structures as part of their protein design and analysis workflows. It simplifies the process, allowing users to focus on analyzing results and advancing their research.</p>
<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.protein_edits.ChainAdder.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">default_python</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'python3'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><span class="pre">JobStarter</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/tools/protein_edits.html#ChainAdder.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.protein_edits.ChainAdder.__init__" title="Link to this definition"></a></dt>
<dd><p>Initialize the ChainAdder class.</p>
<p>This method sets up the ChainAdder class by configuring the path to the default Python executable and
initializing the job starter. The ChainAdder class is used to add chains to protein structures within
the ProtFlow framework.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>default_python</strong> (<em>str</em><em>, </em><em>optional</em>) – The path to the default Python executable, by default <cite>os.path.join(PROTFLOW_ENV, “python3”)</cite>.</p></li>
<li><p><strong>jobstarter</strong> (<a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><em>JobStarter</em></a><em>, </em><em>optional</em>) – An instance of the JobStarter class to manage job execution, by default None.</p></li>
</ul>
</dd>
</dl>
<dl class="py attribute">
<dt class="sig sig-object py" id="protflow.tools.protein_edits.ChainAdder.python">
<span class="sig-name descname"><span class="pre">python</span></span><a class="headerlink" href="#protflow.tools.protein_edits.ChainAdder.python" title="Link to this definition"></a></dt>
<dd><p>Path to the Python executable used for running scripts.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="protflow.tools.protein_edits.ChainAdder.jobstarter">
<span class="sig-name descname"><span class="pre">jobstarter</span></span><a class="headerlink" href="#protflow.tools.protein_edits.ChainAdder.jobstarter" title="Link to this definition"></a></dt>
<dd><p>An instance of the JobStarter class to manage job execution.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter">JobStarter</a></p>
</dd>
</dl>
</dd></dl>

<p class="rubric">Examples</p>
<p>Here is an example of how to initialize the ChainAdder class:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">protein_edits</span> <span class="kn">import</span> <span class="n">ChainAdder</span>

<span class="c1"># Initialize the ChainAdder class</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>
<span class="n">chain_adder</span> <span class="o">=</span> <span class="n">ChainAdder</span><span class="p">(</span><span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">)</span>
</pre></div>
</div>
<p class="rubric">Notes</p>
<p>The ChainAdder class depends on the ProtFlow environment being properly configured. Ensure that the
<cite>PROTFLOW_ENV</cite> and necessary scripts are correctly set up before using this class.</p>
<dl class="field-list simple">
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>FileNotFoundError</strong> – If the specified Python executable is not found.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.protein_edits.ChainAdder.add_chain">
<span class="sig-name descname"><span class="pre">add_chain</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy_chain</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><span class="pre">JobStarter</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span></span><a class="reference internal" href="_modules/protflow/tools/protein_edits.html#ChainAdder.add_chain"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.protein_edits.ChainAdder.add_chain" title="Link to this definition"></a></dt>
<dd><p>Add a chain to the poses.</p>
<p>This method adds a specified chain to the protein structures in <cite>poses</cite> by using the <cite>superimpose_add_chain</cite> method without any superimposition, effectively copying the chain as-is.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>poses</strong> (<a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><em>Poses</em></a>) – The Poses object containing the protein structures.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – A prefix used to name and organize the output files.</p></li>
<li><p><strong>ref_col</strong> (<em>str</em>) – The column in the poses DataFrame that references the structures to be used.</p></li>
<li><p><strong>copy_chain</strong> (<em>str</em>) – The chain identifier to copy.</p></li>
<li><p><strong>jobstarter</strong> (<a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><em>JobStarter</em></a><em>, </em><em>optional</em>) – An instance of the JobStarter class to manage job execution. Defaults to None.</p></li>
<li><p><strong>overwrite</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, overwrite existing outputs. Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An updated Poses object with the new chain added.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses">Poses</a></p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>FileNotFoundError</strong> – If required files or directories are not found during the execution process.</p></li>
<li><p><strong>ValueError</strong> – If invalid arguments are provided to the methods.</p></li>
<li><p><strong>TypeError</strong> – If invalid argument types are provided to the methods.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize and use the <cite>add_chain</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">protein_edits</span> <span class="kn">import</span> <span class="n">ChainAdder</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the ChainAdder class</span>
<span class="n">chain_adder</span> <span class="o">=</span> <span class="n">ChainAdder</span><span class="p">(</span><span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">)</span>

<span class="c1"># Add a chain to the poses</span>
<span class="n">added_chains</span> <span class="o">=</span> <span class="n">chain_adder</span><span class="o">.</span><span class="n">add_chain</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">ref_col</span><span class="o">=</span><span class="s2">&quot;reference_column&quot;</span><span class="p">,</span>
    <span class="n">copy_chain</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">added_chains</span><span class="p">)</span>
</pre></div>
</div>
<section id="id40">
<h5>Further Details<a class="headerlink" href="#id40" title="Link to this heading"></a></h5>
<ul class="simple">
<li><p><strong>Method Simplicity:</strong> This method uses <cite>superimpose_add_chain</cite> without specifying any superimposition parameters, making it a straightforward way to add chains without the complexity of superimposition.</p></li>
<li><p><strong>Path Configuration:</strong> Ensure the paths to the scripts and executables are correctly configured as per ProtFlow setup. Using default paths is recommended unless customization is necessary.</p></li>
<li><p><strong>JobStarter Integration:</strong> The JobStarter object is used to manage job execution, ensuring processes are handled efficiently. If a JobStarter is not provided, the method will not operate without it.</p></li>
</ul>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.protein_edits.ChainAdder.add_sequence">
<span class="sig-name descname"><span class="pre">add_sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sep</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">':'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/protflow/tools/protein_edits.html#ChainAdder.add_sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.protein_edits.ChainAdder.add_sequence" title="Link to this definition"></a></dt>
<dd><p>Add a sequence to the poses in .fa format.</p>
<p>This method appends a specified sequence to the protein sequences in the <cite>poses</cite> object. The sequence can be
provided directly or specified through a column in the <cite>poses</cite> DataFrame. The updated sequences are saved
in .fa format in a specified directory.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>prefix</strong> (<em>str</em>) – A prefix used to name and organize the output files.</p></li>
<li><p><strong>poses</strong> (<a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><em>Poses</em></a>) – The Poses object containing the protein structures.</p></li>
<li><p><strong>seq</strong> (<em>str</em><em>, </em><em>optional</em>) – The sequence to be added. If specified, <cite>seq_col</cite> must be None. Defaults to None.</p></li>
<li><p><strong>seq_col</strong> (<em>str</em><em>, </em><em>optional</em>) – The column in the poses DataFrame that contains the sequences to be added.
If specified, <cite>seq</cite> must be None. Defaults to None.</p></li>
<li><p><strong>sep</strong> (<em>str</em><em>, </em><em>optional</em>) – The separator to be used between the original and new sequences. Defaults to “:”.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If poses are not in .fa or .fasta format, if both <cite>seq</cite> and <cite>seq_col</cite> are specified,
    or if neither <cite>seq</cite> nor <cite>seq_col</cite> is specified.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>add_sequence</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protein_edits</span> <span class="kn">import</span> <span class="n">ChainAdder</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>

<span class="c1"># Initialize the ChainAdder class</span>
<span class="n">chain_adder</span> <span class="o">=</span> <span class="n">ChainAdder</span><span class="p">()</span>

<span class="c1"># Add a sequence to the poses</span>
<span class="n">chain_adder</span><span class="o">.</span><span class="n">add_sequence</span><span class="p">(</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">seq</span><span class="o">=</span><span class="s2">&quot;ATCGATCGATCG&quot;</span><span class="p">,</span>
    <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;:&quot;</span>
<span class="p">)</span>
</pre></div>
</div>
<section id="id41">
<h5>Further Details<a class="headerlink" href="#id41" title="Link to this heading"></a></h5>
<ul class="simple">
<li><p><strong>File Format:</strong> The method checks that all poses are in .fa or .fasta format and raises an error if not.</p></li>
<li><dl class="simple">
<dt><strong>Sequence Input:</strong> Either <cite>seq</cite> or <cite>seq_col</cite> must be specified to provide the sequence to be added.</dt><dd><p>The method ensures that both are not specified simultaneously.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>Output Directory:</strong> The method creates an output directory if it does not exist and saves the updated</dt><dd><p>sequences in this directory.</p>
</dd>
</dl>
</li>
<li><p><strong>DataFrame Update:</strong> The <cite>poses</cite> DataFrame is updated to reflect the new locations of the modified sequences.</p></li>
</ul>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.protein_edits.ChainAdder.multimerize">
<span class="sig-name descname"><span class="pre">multimerize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_protomers</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sep</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">':'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/protflow/tools/protein_edits.html#ChainAdder.multimerize"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.protein_edits.ChainAdder.multimerize" title="Link to this definition"></a></dt>
<dd><p>Create multimers from the sequences in .fa files.</p>
<p>This method takes .fa files from the <cite>poses</cite> object and creates multimers by repeating the sequence a specified number of times.
The updated sequences are saved in .fa format in a specified directory.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>prefix</strong> (<em>str</em>) – A prefix used to name and organize the output files.</p></li>
<li><p><strong>poses</strong> (<a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><em>Poses</em></a>) – The Poses object containing the protein structures.</p></li>
<li><p><strong>n_protomers</strong> (<em>int</em>) – The number of protomers in the final .fa file.</p></li>
<li><p><strong>sep</strong> (<em>str</em><em>, </em><em>optional</em>) – The separator to be used between the original and new sequences. Defaults to “:”.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If poses are not in .fa or .fasta format.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>multimerize</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protein_edits</span> <span class="kn">import</span> <span class="n">ChainAdder</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>

<span class="c1"># Initialize the ChainAdder class</span>
<span class="n">chain_adder</span> <span class="o">=</span> <span class="n">ChainAdder</span><span class="p">()</span>

<span class="c1"># Multimerize the sequences in the poses</span>
<span class="n">chain_adder</span><span class="o">.</span><span class="n">multimerize</span><span class="p">(</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">n_protomers</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
    <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;:&quot;</span>
<span class="p">)</span>
</pre></div>
</div>
<section id="id42">
<h5>Further Details<a class="headerlink" href="#id42" title="Link to this heading"></a></h5>
<ul class="simple">
<li><p><strong>File Format:</strong> The method checks that all poses are in .fa or .fasta format and raises an error if not.</p></li>
<li><p><strong>Protomers Specification:</strong> The <cite>n_protomers</cite> parameter specifies the number of times the sequence should be repeated to form a multimer.</p></li>
<li><p><strong>Output Directory:</strong> The method creates an output directory if it does not exist and saves the updated sequences in this directory.</p></li>
<li><p><strong>DataFrame Update:</strong> The <cite>poses</cite> DataFrame is updated to reflect the new locations of the modified sequences.</p></li>
</ul>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.protein_edits.ChainAdder.parse_motif">
<span class="sig-name descname"><span class="pre">parse_motif</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.residues.ResidueSelection" title="protflow.residues.ResidueSelection"><span class="pre">ResidueSelection</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Series</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">str</span></span></span><a class="reference internal" href="_modules/protflow/tools/protein_edits.html#ChainAdder.parse_motif"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.protein_edits.ChainAdder.parse_motif" title="Link to this definition"></a></dt>
<dd><p>Set up motif from target_motif input.</p>
<p>This method converts a given motif, either a <cite>ResidueSelection</cite> object or a string, into a string format suitable for further processing.
If the motif is a string, it checks if it is a column in the <cite>pose</cite> DataFrame and assumes it points to a <cite>ResidueSelection</cite> object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>motif</strong> (<a class="reference internal" href="protflow.html#protflow.residues.ResidueSelection" title="protflow.residues.ResidueSelection"><em>ResidueSelection</em></a><em> | </em><em>str</em>) – The motif to be parsed. It can be either a <cite>ResidueSelection</cite> object or a string.</p></li>
<li><p><strong>pose</strong> (<em>pd.Series</em>) – A row from the poses DataFrame that contains information about the protein structure.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The motif in string format.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError</strong> – If the motif is a string but not a column in the <cite>poses.df</cite> DataFrame.</p></li>
<li><p><strong>TypeError</strong> – If the motif is neither a <cite>ResidueSelection</cite> object nor a string.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>parse_motif</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.residues</span> <span class="kn">import</span> <span class="n">ResidueSelection</span>
<span class="kn">from</span> <span class="nn">protein_edits</span> <span class="kn">import</span> <span class="n">ChainAdder</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="c1"># Initialize the ChainAdder class</span>
<span class="n">chain_adder</span> <span class="o">=</span> <span class="n">ChainAdder</span><span class="p">()</span>

<span class="c1"># Example pose DataFrame row</span>
<span class="n">pose</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">({</span><span class="s1">&#39;motif_column&#39;</span><span class="p">:</span> <span class="n">ResidueSelection</span><span class="p">(</span><span class="o">...</span><span class="p">)})</span>

<span class="c1"># Parse a ResidueSelection object</span>
<span class="n">motif</span> <span class="o">=</span> <span class="n">ResidueSelection</span><span class="p">(</span><span class="o">...</span><span class="p">)</span>
<span class="n">motif_str</span> <span class="o">=</span> <span class="n">chain_adder</span><span class="o">.</span><span class="n">parse_motif</span><span class="p">(</span><span class="n">motif</span><span class="p">,</span> <span class="n">pose</span><span class="p">)</span>

<span class="c1"># Parse a string that is a column in the pose DataFrame</span>
<span class="n">motif_str</span> <span class="o">=</span> <span class="n">chain_adder</span><span class="o">.</span><span class="n">parse_motif</span><span class="p">(</span><span class="s1">&#39;motif_column&#39;</span><span class="p">,</span> <span class="n">pose</span><span class="p">)</span>

<span class="c1"># Access the result</span>
<span class="nb">print</span><span class="p">(</span><span class="n">motif_str</span><span class="p">)</span>
</pre></div>
</div>
<section id="id43">
<h5>Further Details<a class="headerlink" href="#id43" title="Link to this heading"></a></h5>
<ul class="simple">
<li><p><strong>ResidueSelection Handling:</strong> The method directly converts a <cite>ResidueSelection</cite> object to its string representation using its <cite>to_string</cite> method.</p></li>
<li><p><strong>String Handling:</strong> If a string is provided, the method checks if it is a column in the <cite>pose</cite> DataFrame that points to a <cite>ResidueSelection</cite> object, converting it to a string.</p></li>
<li><p><strong>Error Handling:</strong> The method raises appropriate errors if the input is not of the expected type or if the string does not correspond to a valid column in the DataFrame.</p></li>
</ul>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.protein_edits.ChainAdder.run">
<span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">poses</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/tools/protein_edits.html#ChainAdder.run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.protein_edits.ChainAdder.run" title="Link to this definition"></a></dt>
<dd><p>.run() not implemented for ChainAdder class. Use methods like: .add_chain() or .superimpose_add_chain() instead!!!</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.protein_edits.ChainAdder.superimpose_add_chain">
<span class="sig-name descname"><span class="pre">superimpose_add_chain</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy_chain</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><span class="pre">JobStarter</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_motif</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.residues.ResidueSelection" title="protflow.residues.ResidueSelection"><span class="pre">ResidueSelection</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_motif</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.residues.ResidueSelection" title="protflow.residues.ResidueSelection"><span class="pre">ResidueSelection</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_chains</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_chains</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span></span><a class="reference internal" href="_modules/protflow/tools/protein_edits.html#ChainAdder.superimpose_add_chain"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.protein_edits.ChainAdder.superimpose_add_chain" title="Link to this definition"></a></dt>
<dd><p>Add a protein chain after superimposition on a motif or chain.</p>
<p>This method adds a chain to the protein structures in <cite>poses</cite> by superimposing it on a specified motif or chain.
It sets up and executes the necessary scripts, handles the environment configuration, and processes the output.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>poses</strong> (<a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><em>Poses</em></a>) – The Poses object containing the protein structures.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – A prefix used to name and organize the output files.</p></li>
<li><p><strong>ref_col</strong> (<em>str</em>) – The column in the poses DataFrame that references the structures to be used.</p></li>
<li><p><strong>copy_chain</strong> (<em>str</em>) – The chain identifier to copy.</p></li>
<li><p><strong>jobstarter</strong> (<a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><em>JobStarter</em></a><em>, </em><em>optional</em>) – An instance of the JobStarter class to manage job execution. Defaults to None.</p></li>
<li><p><strong>target_motif</strong> (<a class="reference internal" href="protflow.html#protflow.residues.ResidueSelection" title="protflow.residues.ResidueSelection"><em>ResidueSelection</em></a><em>, </em><em>optional</em>) – The target motif for superimposition. Defaults to None.</p></li>
<li><p><strong>reference_motif</strong> (<a class="reference internal" href="protflow.html#protflow.residues.ResidueSelection" title="protflow.residues.ResidueSelection"><em>ResidueSelection</em></a><em>, </em><em>optional</em>) – The reference motif for superimposition. Defaults to None.</p></li>
<li><p><strong>target_chains</strong> (<em>list</em><em>, </em><em>optional</em>) – A list of target chains for superimposition. Defaults to None.</p></li>
<li><p><strong>reference_chains</strong> (<em>list</em><em>, </em><em>optional</em>) – A list of reference chains for superimposition. Defaults to None.</p></li>
<li><p><strong>overwrite</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, overwrite existing outputs. Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An updated Poses object with the new chain added.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses">Poses</a></p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError</strong> – If both motifs and chains are specified for superimposition.</p></li>
<li><p><strong>FileNotFoundError</strong> – If required files or directories are not found during the execution process.</p></li>
<li><p><strong>TypeError</strong> – If invalid argument types are provided to the methods.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize and use the <cite>superimpose_add_chain</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">protein_edits</span> <span class="kn">import</span> <span class="n">ChainAdder</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the ChainAdder class</span>
<span class="n">chain_adder</span> <span class="o">=</span> <span class="n">ChainAdder</span><span class="p">(</span><span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">)</span>

<span class="c1"># Add a chain to the poses</span>
<span class="n">added_chains</span> <span class="o">=</span> <span class="n">chain_adder</span><span class="o">.</span><span class="n">superimpose_add_chain</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">ref_col</span><span class="o">=</span><span class="s2">&quot;reference_column&quot;</span><span class="p">,</span>
    <span class="n">copy_chain</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">added_chains</span><span class="p">)</span>
</pre></div>
</div>
<section id="id44">
<h5>Further Details<a class="headerlink" href="#id44" title="Link to this heading"></a></h5>
<ul class="simple">
<li><p><strong>Path Configuration:</strong> Ensure the paths to the scripts and executables are correctly configured as per ProtFlow setup. Using default paths is recommended unless customization is necessary.</p></li>
<li><p><strong>JobStarter Integration:</strong> The JobStarter object is used to manage job execution, ensuring processes are handled efficiently. If a JobStarter is not provided, the method will not operate without it.</p></li>
</ul>
<p class="rubric">Notes</p>
<p>This method ensures robust error handling and logging for easier debugging and verification of the process.</p>
</section>
</dd></dl>

</section>
</section>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="protflow.tools.protein_edits.ChainRemover">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">protflow.tools.protein_edits.</span></span><span class="sig-name descname"><span class="pre">ChainRemover</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">default_python</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'python3'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><span class="pre">JobStarter</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/tools/protein_edits.html#ChainRemover"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.protein_edits.ChainRemover" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="protflow.html#protflow.runners.Runner" title="protflow.runners.Runner"><code class="xref py py-class docutils literal notranslate"><span class="pre">Runner</span></code></a></p>
<section id="chainremover-class">
<h3>ChainRemover Class<a class="headerlink" href="#chainremover-class" title="Link to this heading"></a></h3>
<p>The <cite>ChainRemover</cite> class is a specialized class designed to facilitate the removal of chains from protein structures within the ProtFlow framework. It extends the <cite>Runner</cite> class and incorporates specific methods to handle the setup, execution, and data collection associated with chain removal processes.</p>
<section id="id45">
<h4>Detailed Description<a class="headerlink" href="#id45" title="Link to this heading"></a></h4>
<p>The <cite>ChainRemover</cite> class manages all aspects of removing chains from protein structures. It configures necessary scripts and executables, prepares the environment for the removal processes, and executes the required commands. Additionally, it collects and processes the output data, organizing it into a structured format for further analysis.</p>
<dl class="simple">
<dt>Key functionalities include:</dt><dd><ul class="simple">
<li><p>Setting up paths to chain removal scripts and Python executables.</p></li>
<li><p>Configuring job starter options, either automatically or manually.</p></li>
<li><p>Handling the execution of chain removal commands with support for batch processing.</p></li>
<li><p>Collecting and processing output data into a structured format.</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">rtype<span class="colon">:</span></dt>
<dd class="field-odd"><p>An instance of the <cite>ChainRemover</cite> class, configured to remove chains from protein structures and handle outputs efficiently.</p>
</dd>
<dt class="field-even">raises - FileNotFoundError<span class="colon">:</span></dt>
<dd class="field-even"><p>If required files or directories are not found during the execution process.:</p>
</dd>
<dt class="field-odd">raises - ValueError<span class="colon">:</span></dt>
<dd class="field-odd"><p>If invalid arguments are provided to the methods.:</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize and use the <cite>ChainRemover</cite> class:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">protein_edits</span> <span class="kn">import</span> <span class="n">ChainRemover</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the ChainRemover class</span>
<span class="n">chain_remover</span> <span class="o">=</span> <span class="n">ChainRemover</span><span class="p">(</span><span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">)</span>

<span class="c1"># Remove a chain from the poses</span>
<span class="n">removed_chains</span> <span class="o">=</span> <span class="n">chain_remover</span><span class="o">.</span><span class="n">remove_chains</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_2&quot;</span><span class="p">,</span>
    <span class="n">chains</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">],</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">removed_chains</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id46">
<h4>Further Details<a class="headerlink" href="#id46" title="Link to this heading"></a></h4>
<blockquote>
<div><ul class="simple">
<li><p>Edge Cases: The class handles various edge cases, such as missing chain specifications and the need to overwrite previous results.</p></li>
<li><p>Customization: The class provides extensive customization options through its parameters, allowing users to tailor the chain removal process to their specific needs.</p></li>
<li><p>Integration: Seamlessly integrates with other ProtFlow components, leveraging shared configurations and data structures for a unified workflow.</p></li>
</ul>
</div></blockquote>
<p>The ChainRemover class is intended for researchers and developers who need to remove chains from protein structures as part of their protein design and analysis workflows. It simplifies the process, allowing users to focus on analyzing results and advancing their research.</p>
<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.protein_edits.ChainRemover.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">default_python</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'python3'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><span class="pre">JobStarter</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/tools/protein_edits.html#ChainRemover.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.protein_edits.ChainRemover.__init__" title="Link to this definition"></a></dt>
<dd><p>Initialize the ChainRemover class.</p>
<p>This method sets up the ChainRemover class by configuring the path to the default Python executable and
initializing the job starter. The ChainRemover class is used to remove chains from protein structures within
the ProtFlow framework.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>default_python</strong> (<em>str</em><em>, </em><em>optional</em>) – The path to the default Python executable. Defaults to <cite>os.path.join(PROTFLOW_ENV, “python3”)</cite>.</p></li>
<li><p><strong>jobstarter</strong> (<a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><em>JobStarter</em></a><em>, </em><em>optional</em>) – An instance of the JobStarter class to manage job execution. Defaults to None.</p></li>
</ul>
</dd>
</dl>
<dl class="py attribute">
<dt class="sig sig-object py" id="protflow.tools.protein_edits.ChainRemover.python">
<span class="sig-name descname"><span class="pre">python</span></span><a class="headerlink" href="#protflow.tools.protein_edits.ChainRemover.python" title="Link to this definition"></a></dt>
<dd><p>Path to the Python executable used for running scripts.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="protflow.tools.protein_edits.ChainRemover.jobstarter">
<span class="sig-name descname"><span class="pre">jobstarter</span></span><a class="headerlink" href="#protflow.tools.protein_edits.ChainRemover.jobstarter" title="Link to this definition"></a></dt>
<dd><p>An instance of the JobStarter class to manage job execution.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter">JobStarter</a></p>
</dd>
</dl>
</dd></dl>

<p class="rubric">Examples</p>
<p>Here is an example of how to initialize the ChainRemover class:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">protein_edits</span> <span class="kn">import</span> <span class="n">ChainRemover</span>

<span class="c1"># Initialize the ChainRemover class</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>
<span class="n">chain_remover</span> <span class="o">=</span> <span class="n">ChainRemover</span><span class="p">(</span><span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">)</span>
</pre></div>
</div>
<p class="rubric">Notes</p>
<p>The ChainRemover class depends on the ProtFlow environment being properly configured. Ensure that the
<cite>PROTFLOW_ENV</cite> and necessary scripts are correctly set up before using this class.</p>
<dl class="field-list simple">
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>FileNotFoundError</strong> – If the specified Python executable is not found.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.protein_edits.ChainRemover.remove_chains">
<span class="sig-name descname"><span class="pre">remove_chains</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chains</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><span class="pre">JobStarter</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/tools/protein_edits.html#ChainRemover.remove_chains"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.protein_edits.ChainRemover.remove_chains" title="Link to this definition"></a></dt>
<dd><p>Remove chains from the poses.</p>
<p>This method removes specified chains from the protein structures in the <cite>poses</cite> object. It sets up and executes the necessary scripts,
handles the environment configuration, and processes the output.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>poses</strong> (<a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><em>Poses</em></a>) – The Poses object containing the protein structures.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – A prefix used to name and organize the output files.</p></li>
<li><p><strong>chains</strong> (<em>list</em><em>, </em><em>optional</em>) – A list of chains to be removed. If specified, each chain in the list will be removed from the poses. Defaults to None.</p></li>
<li><p><strong>jobstarter</strong> (<a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><em>JobStarter</em></a><em>, </em><em>optional</em>) – An instance of the JobStarter class to manage job execution. Defaults to None.</p></li>
<li><p><strong>overwrite</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, overwrite existing outputs. Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An updated Poses object with the specified chains removed.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses">Poses</a></p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>FileNotFoundError</strong> – If required files or directories are not found during the execution process.</p></li>
<li><p><strong>ValueError</strong> – If invalid arguments are provided to the methods.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize and use the <cite>remove_chains</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">protein_edits</span> <span class="kn">import</span> <span class="n">ChainRemover</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the ChainRemover class</span>
<span class="n">chain_remover</span> <span class="o">=</span> <span class="n">ChainRemover</span><span class="p">(</span><span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">)</span>

<span class="c1"># Remove chains from the poses</span>
<span class="n">removed_chains</span> <span class="o">=</span> <span class="n">chain_remover</span><span class="o">.</span><span class="n">remove_chains</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_2&quot;</span><span class="p">,</span>
    <span class="n">chains</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">],</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">removed_chains</span><span class="p">)</span>
</pre></div>
</div>
<section id="id47">
<h5>Further Details<a class="headerlink" href="#id47" title="Link to this heading"></a></h5>
<ul class="simple">
<li><p><strong>Output Checking:</strong> The method checks if the output already exists and whether it should be overwritten, ensuring no redundant processing.</p></li>
<li><p><strong>Chain Setup:</strong> Chains can be specified as a list, a column in the <cite>poses</cite> DataFrame, or as a single chain identifier for all poses.</p></li>
<li><p><strong>Batch Processing:</strong> The method supports batch processing, splitting the inputs into sublists to optimize resource usage during execution.</p></li>
<li><p><strong>Path Configuration:</strong> Ensure the paths to the scripts and executables are correctly configured as per ProtFlow setup. Using default paths is recommended unless customization is necessary.</p></li>
<li><p><strong>JobStarter Integration:</strong> The JobStarter object is used to manage job execution, ensuring processes are handled efficiently. If a JobStarter is not provided, the method will operate without it, but using one is recommended for better job management.</p></li>
</ul>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.protein_edits.ChainRemover.run">
<span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">poses</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/tools/protein_edits.html#ChainRemover.run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.protein_edits.ChainRemover.run" title="Link to this definition"></a></dt>
<dd><p>.run() method is not implemented in this class. Use .remove_chains() instead!!!</p>
</dd></dl>

</section>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="protflow.tools.protein_edits.parse_chain">
<span class="sig-prename descclassname"><span class="pre">protflow.tools.protein_edits.</span></span><span class="sig-name descname"><span class="pre">parse_chain</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chain</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Series</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">str</span></span></span><a class="reference internal" href="_modules/protflow/tools/protein_edits.html#parse_chain"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.protein_edits.parse_chain" title="Link to this definition"></a></dt>
<dd><p>Sets up chain for add_chains_batch.py</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="protflow.tools.protein_edits.setup_chain_list">
<span class="sig-prename descclassname"><span class="pre">protflow.tools.protein_edits.</span></span><span class="sig-name descname"><span class="pre">setup_chain_list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chain_arg</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/protflow/tools/protein_edits.html#setup_chain_list"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.protein_edits.setup_chain_list" title="Link to this definition"></a></dt>
<dd><p>Set up chains for add_chains_batch.py.</p>
<p>This function configures the list of chains to be used in the <cite>add_chains_batch.py</cite> script based on the provided <cite>chain_arg</cite>.
It supports specifying a single chain, a column in the <cite>poses</cite> DataFrame, or a list of chains.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chain_arg</strong> (<em>str</em><em> or </em><em>list</em><em>[</em><em>str</em><em>]</em>) – The chain specification. It can be a single chain identifier (e.g., ‘A’),
the name of a column in the <cite>poses</cite> DataFrame where the chains are listed,
or a list of chain identifiers.</p></li>
<li><p><strong>poses</strong> (<a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><em>Poses</em></a>) – The Poses object containing the protein structures.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A list of chain identifiers to be used in <cite>add_chains_batch.py</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[str]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the <cite>chain_arg</cite> value is inappropriate, such as when the specified column does not exist in the DataFrame
    or the length of the list does not match the number of poses.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>setup_chain_list</cite> function:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protein_edits</span> <span class="kn">import</span> <span class="n">setup_chain_list</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>

<span class="c1"># Set up a single chain</span>
<span class="n">chain_list</span> <span class="o">=</span> <span class="n">setup_chain_list</span><span class="p">(</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="n">poses</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">chain_list</span><span class="p">)</span>

<span class="c1"># Set up chains from a column in the poses DataFrame</span>
<span class="n">chain_list</span> <span class="o">=</span> <span class="n">setup_chain_list</span><span class="p">(</span><span class="s1">&#39;chain_col&#39;</span><span class="p">,</span> <span class="n">poses</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">chain_list</span><span class="p">)</span>

<span class="c1"># Set up chains from a list</span>
<span class="n">chain_list</span> <span class="o">=</span> <span class="n">setup_chain_list</span><span class="p">([</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">],</span> <span class="n">poses</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">chain_list</span><span class="p">)</span>
</pre></div>
</div>
<section id="id48">
<h3>Further Details<a class="headerlink" href="#id48" title="Link to this heading"></a></h3>
<ul class="simple">
<li><p><strong>Single Chain Identifier:</strong> If a single chain identifier (e.g., ‘A’) is provided, it is used for all poses.</p></li>
<li><dl class="simple">
<dt><strong>DataFrame Column:</strong> If the name of a column in the <cite>poses</cite> DataFrame is provided, the function extracts the chain identifiers</dt><dd><p>from that column for each pose.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>List of Chains:</strong> If a list of chain identifiers is provided, it must match the length of the <cite>poses</cite> DataFrame.</dt><dd><p>The function raises an error if this condition is not met.</p>
</dd>
</dl>
</li>
</ul>
</section>
</dd></dl>

</section>
<section id="module-protflow.tools.protein_generator">
<span id="protflow-tools-protein-generator-module"></span><h2>protflow.tools.protein_generator module<a class="headerlink" href="#module-protflow.tools.protein_generator" title="Link to this heading"></a></h2>
<section id="proteingenerator-module">
<h3>ProteinGenerator Module<a class="headerlink" href="#proteingenerator-module" title="Link to this heading"></a></h3>
<p>This module provides the functionality to integrate the protein generation process within the ProtFlow framework. It offers tools to run the protein_generator script, handle its inputs and outputs, and process the resulting data in a structured and automated manner.</p>
<section id="id49">
<h4>Detailed Description<a class="headerlink" href="#id49" title="Link to this heading"></a></h4>
<p>The <cite>ProteinGenerator</cite> class encapsulates the functionality necessary to run protein_generator as described in the publication. It manages the configuration of paths to essential scripts and Python executables, sets up the environment, and handles the execution of protein generation processes. It also includes methods for collecting and processing output data, ensuring that the results are organized and accessible for further analysis within the ProtFlow ecosystem.</p>
<p>The module is designed to streamline the integration of protein generation into larger computational workflows. It supports the automatic setup of job parameters, execution of protein generator commands, and parsing of output files into a structured DataFrame format. This facilitates subsequent data analysis and visualization steps.</p>
</section>
<section id="id50">
<h4>Usage<a class="headerlink" href="#id50" title="Link to this heading"></a></h4>
<p>To use this module, create an instance of the <cite>ProteinGenerator</cite> class and invoke its <cite>run</cite> method with appropriate parameters. The module will handle the configuration, execution, and result collection processes. Detailed control over the protein generation process is provided through various parameters, allowing for customized runs tailored to specific research needs.</p>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize and use the <cite>ProteinGenerator</cite> class within a ProtFlow pipeline:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">protein_generator</span> <span class="kn">import</span> <span class="n">ProteinGenerator</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the ProteinGenerator class</span>
<span class="n">protein_generator</span> <span class="o">=</span> <span class="n">ProteinGenerator</span><span class="p">()</span>

<span class="c1"># Run the protein generation process</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">protein_generator</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;generation.num_proteins=10&quot;</span><span class="p">,</span>
    <span class="n">pose_options</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;generation.input_pdb=&#39;input.pdb&#39;&quot;</span><span class="p">],</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id51">
<h4>Further Details<a class="headerlink" href="#id51" title="Link to this heading"></a></h4>
<blockquote>
<div><ul class="simple">
<li><p>Edge Cases: The module handles various edge cases, such as empty pose lists and the need to overwrite previous results. It ensures robust error handling and logging for easier debugging and verification of the protein generation process.</p></li>
<li><p>Customizability: Users can customize the protein generation process through multiple parameters, including the number of generated proteins, specific options for the protein generator script, and options for handling pose-specific parameters.</p></li>
<li><p>Integration: The module seamlessly integrates with other components of the ProtFlow framework, leveraging shared configurations and data structures to provide a cohesive user experience.</p></li>
</ul>
</div></blockquote>
<p>This module is intended for researchers and developers who need to incorporate protein generation into their protein design and analysis workflows. By automating many of the setup and execution steps, it allows users to focus on interpreting results and advancing their scientific inquiries.</p>
<p class="rubric">Notes</p>
<p>This module is part of the ProtFlow package and is designed to work in tandem with other components of the package, especially those related to job management in HPC environments.</p>
</section>
<section id="id52">
<h4>Author<a class="headerlink" href="#id52" title="Link to this heading"></a></h4>
<p>Markus Braun, Adrian Tripp</p>
</section>
<section id="id53">
<h4>Version<a class="headerlink" href="#id53" title="Link to this heading"></a></h4>
<p>0.1.0</p>
</section>
</section>
<dl class="py class">
<dt class="sig sig-object py" id="protflow.tools.protein_generator.ProteinGenerator">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">protflow.tools.protein_generator.</span></span><span class="sig-name descname"><span class="pre">ProteinGenerator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">script_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">python_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><span class="pre">JobStarter</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/tools/protein_generator.html#ProteinGenerator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.protein_generator.ProteinGenerator" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="protflow.html#protflow.runners.Runner" title="protflow.runners.Runner"><code class="xref py py-class docutils literal notranslate"><span class="pre">Runner</span></code></a></p>
<section id="proteingenerator-class">
<h3>ProteinGenerator Class<a class="headerlink" href="#proteingenerator-class" title="Link to this heading"></a></h3>
<p>The <cite>ProteinGenerator</cite> class is a specialized class designed to facilitate the execution of protein generation within the ProtFlow framework. It extends the <cite>Runner</cite> class and incorporates specific methods to handle the setup, execution, and data collection associated with protein generation processes.</p>
<section id="id54">
<h4>Detailed Description<a class="headerlink" href="#id54" title="Link to this heading"></a></h4>
<p>The <cite>ProteinGenerator</cite> class manages all aspects of running protein generation simulations. It handles the configuration of necessary scripts and executables, prepares the environment for protein generation processes, and executes the generation commands. Additionally, it collects and processes the output data, organizing it into a structured format for further analysis.</p>
<dl class="simple">
<dt>Key functionalities include:</dt><dd><ul class="simple">
<li><p>Setting up paths to protein generation scripts and Python executables.</p></li>
<li><p>Configuring job starter options, either automatically or manually.</p></li>
<li><p>Handling the execution of protein generation commands with support for multiple generation runs.</p></li>
<li><p>Collecting and processing output data into a pandas DataFrame.</p></li>
<li><p>Ensuring robust error handling and logging for easier debugging and verification of the generation process.</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">rtype<span class="colon">:</span></dt>
<dd class="field-odd"><p>An instance of the <cite>ProteinGenerator</cite> class, configured to run protein generation processes and handle outputs efficiently.</p>
</dd>
<dt class="field-even">raises - FileNotFoundError<span class="colon">:</span></dt>
<dd class="field-even"><p>If required files or directories are not found during the execution process.:</p>
</dd>
<dt class="field-odd">raises - ValueError<span class="colon">:</span></dt>
<dd class="field-odd"><p>If invalid arguments are provided to the methods.:</p>
</dd>
<dt class="field-even">raises - TypeError<span class="colon">:</span></dt>
<dd class="field-even"><p>If the types of arguments are not as expected.:</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize and use the <cite>ProteinGenerator</cite> class:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">protein_generator</span> <span class="kn">import</span> <span class="n">ProteinGenerator</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the ProteinGenerator class</span>
<span class="n">protein_generator</span> <span class="o">=</span> <span class="n">ProteinGenerator</span><span class="p">()</span>

<span class="c1"># Run the protein generation process</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">protein_generator</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;generation.num_proteins=10&quot;</span><span class="p">,</span>
    <span class="n">pose_options</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;generation.input_pdb=&#39;input.pdb&#39;&quot;</span><span class="p">],</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id55">
<h4>Further Details<a class="headerlink" href="#id55" title="Link to this heading"></a></h4>
<blockquote>
<div><ul class="simple">
<li><p>Edge Cases: The class includes handling for various edge cases, such as empty pose lists, the need to overwrite previous results, and the presence of existing score files.</p></li>
<li><p>Customization: The class provides extensive customization options through its parameters, allowing users to tailor the protein generation process to their specific needs.</p></li>
<li><p>Integration: Seamlessly integrates with other ProtFlow components, leveraging shared configurations and data structures for a unified workflow.</p></li>
</ul>
</div></blockquote>
<p>The ProteinGenerator class is intended for researchers and developers who need to perform protein generation as part of their protein design and analysis workflows. It simplifies the process, allowing users to focus on analyzing results and advancing their research.</p>
<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.protein_generator.ProteinGenerator.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">script_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">python_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><span class="pre">JobStarter</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/protflow/tools/protein_generator.html#ProteinGenerator.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.protein_generator.ProteinGenerator.__init__" title="Link to this definition"></a></dt>
<dd><p>Initialize the ProteinGenerator class with paths to the necessary scripts and Python executable.</p>
<p>This constructor sets up the <cite>ProteinGenerator</cite> class, configuring the paths to the protein generator script and the Python executable. It also sets the jobstarter and initializes essential attributes for the class.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>script_path</strong> (<em>str</em><em>, </em><em>optional</em>) – The path to the protein generator script. Defaults to the value set in protflow.config.PROTEIN_GENERATOR_SCRIPT_PATH.</p></li>
<li><p><strong>python_path</strong> (<em>str</em><em>, </em><em>optional</em>) – The path to the Python executable. Defaults to the value set in protflow.config.PROTEIN_GENERATOR_PYTHON_PATH.</p></li>
<li><p><strong>jobstarter</strong> (<a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><em>JobStarter</em></a><em>, </em><em>optional</em>) – An instance of the JobStarter class, which manages job execution. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If no script_path is provided or set in the configuration.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize the <cite>ProteinGenerator</cite> class:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">protein_generator</span> <span class="kn">import</span> <span class="n">ProteinGenerator</span>

<span class="c1"># Initialize the JobStarter class</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the ProteinGenerator class</span>
<span class="n">protein_generator</span> <span class="o">=</span> <span class="n">ProteinGenerator</span><span class="p">(</span>
    <span class="n">script_path</span><span class="o">=</span><span class="s2">&quot;/path/to/protein_generator.py&quot;</span><span class="p">,</span>
    <span class="n">python_path</span><span class="o">=</span><span class="s2">&quot;/path/to/python&quot;</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">protein_generator</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Script Path:</strong> The path to the protein generator script is a critical configuration that needs to be set either through the parameter or the configuration file.</p></li>
<li><p><strong>Python Path:</strong> The path to the Python executable is necessary for running the script and should be set accordingly.</p></li>
<li><p><strong>JobStarter:</strong> If provided, the JobStarter instance is used to manage job execution, otherwise it can be set later.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.protein_generator.ProteinGenerator.collect_scores">
<span class="sig-name descname"><span class="pre">collect_scores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scores_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/protflow/tools/protein_generator.html#ProteinGenerator.collect_scores"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.protein_generator.ProteinGenerator.collect_scores" title="Link to this definition"></a></dt>
<dd><p>Collect scores from the protein_generator output directory.</p>
<p>This method reads the output .pdb files generated by the protein generator and parses the corresponding .trb files into a pandas DataFrame. It consolidates the scores from multiple files into a single DataFrame for further analysis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>scores_dir</strong> (<em>str</em>) – The directory where the output files from the protein generator are stored.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A DataFrame containing the collected scores from the protein generator output files.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>FileNotFoundError</strong> – If no .pdb files are found in the specified output directory, indicating a possible issue with the protein generator execution or an incorrect path.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>collect_scores</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protein_generator</span> <span class="kn">import</span> <span class="n">ProteinGenerator</span>

<span class="c1"># Initialize the ProteinGenerator class</span>
<span class="n">protein_generator</span> <span class="o">=</span> <span class="n">ProteinGenerator</span><span class="p">()</span>

<span class="c1"># Collect scores from the output directory</span>
<span class="n">scores_df</span> <span class="o">=</span> <span class="n">protein_generator</span><span class="o">.</span><span class="n">collect_scores</span><span class="p">(</span><span class="n">scores_dir</span><span class="o">=</span><span class="s2">&quot;output_directory&quot;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">scores_df</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>File Reading:</strong> The method reads all .pdb files from the specified directory. If no .pdb files are found, a FileNotFoundError is raised.</p></li>
<li><p><strong>Data Parsing:</strong> The method parses the corresponding .trb files for each .pdb file, converting them into pandas DataFrames and concatenating them into a single DataFrame.</p></li>
<li><p><strong>Output Organization:</strong> The resulting DataFrame is organized and returned for further analysis, with all scores consolidated from the multiple output files.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.protein_generator.ProteinGenerator.parse_trbfile">
<span class="sig-name descname"><span class="pre">parse_trbfile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">trbfile</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/protflow/tools/protein_generator.html#ProteinGenerator.parse_trbfile"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.protein_generator.ProteinGenerator.parse_trbfile" title="Link to this definition"></a></dt>
<dd><p>Read a protein_generator output .trb file and parse the scores into a pandas DataFrame.</p>
<p>This method reads the specified .trb file, extracts relevant data, and organizes it into a pandas DataFrame. The data includes scores and various attributes related to the protein generation process.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>trbfile</strong> (<em>str</em>) – The file path to the .trb file generated by the protein generator.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A DataFrame containing parsed scores and attributes from the .trb file.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>parse_trbfile</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protein_generator</span> <span class="kn">import</span> <span class="n">ProteinGenerator</span>

<span class="c1"># Initialize the ProteinGenerator class</span>
<span class="n">protein_generator</span> <span class="o">=</span> <span class="n">ProteinGenerator</span><span class="p">()</span>

<span class="c1"># Parse the .trb file</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">protein_generator</span><span class="o">.</span><span class="n">parse_trbfile</span><span class="p">(</span><span class="n">trbfile</span><span class="o">=</span><span class="s2">&quot;output_directory/sample.trb&quot;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>File Reading:</strong> The method uses numpy to load the .trb file, which is expected to be in a specific format.</p></li>
<li><p><strong>Data Extraction:</strong> The method extracts various pieces of information from the .trb file, including description, location, lddt scores, sequence, and contigs.</p></li>
<li><p><strong>Data Formatting:</strong> The extracted data is organized into a dictionary and converted into a pandas DataFrame for ease of use in further analysis.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.protein_generator.ProteinGenerator.run">
<span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><span class="pre">JobStarter</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pose_options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="protflow.html#protflow.runners.RunnerOutput" title="protflow.runners.RunnerOutput"><span class="pre">RunnerOutput</span></a></span></span><a class="reference internal" href="_modules/protflow/tools/protein_generator.html#ProteinGenerator.run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.protein_generator.ProteinGenerator.run" title="Link to this definition"></a></dt>
<dd><p>Execute protein_generator with given poses and jobstarter configuration.
This method sets up and runs protein_generator using the provided poses and jobstarter object. It handles the configuration, execution, and collection of output data, ensuring that the results are organized and accessible for further analysis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>poses</strong> (<a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><em>Poses</em></a>) – The Poses object containing the protein structures.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – A prefix used to name and organize the output files.</p></li>
<li><p><strong>jobstarter</strong> (<a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><em>JobStarter</em></a>) – An instance of the JobStarter class, which manages job execution.</p></li>
<li><p><strong>options</strong> (<em>str</em><em>, </em><em>optional</em>) – Additional options for the protein generation script. Defaults to None.</p></li>
<li><p><strong>pose_options</strong> (<em>list</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – A list of pose-specific options for the protein generation script. Defaults to None.</p></li>
<li><p><strong>overwrite</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, overwrite existing output files. Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An instance of the RunnerOutput class, containing the processed poses and results of the protein generation process.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="protflow.html#protflow.runners.RunnerOutput" title="protflow.runners.RunnerOutput">RunnerOutput</a></p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>FileNotFoundError</strong> – If required files or directories are not found during the execution process.</p></li>
<li><p><strong>ValueError</strong> – If invalid arguments are provided to the method.</p></li>
<li><p><strong>TypeError</strong> – If pose_options are not of the expected type.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>run</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">protein_generator</span> <span class="kn">import</span> <span class="n">ProteinGenerator</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the ProteinGenerator class</span>
<span class="n">protein_generator</span> <span class="o">=</span> <span class="n">ProteinGenerator</span><span class="p">()</span>

<span class="c1"># Run the protein generation process</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">protein_generator</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;generation.num_proteins=10&quot;</span><span class="p">,</span>
    <span class="n">pose_options</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;generation.input_pdb=&#39;input.pdb&#39;&quot;</span><span class="p">],</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Setup and Execution:</strong> The method ensures that the environment is correctly set up, directories are prepared, and necessary commands are constructed and executed.</p></li>
<li><p><strong>Output Management:</strong> The method handles the collection and processing of output data, ensuring that results are organized and accessible for further analysis.</p></li>
<li><p><strong>Customization:</strong> Extensive customization options are provided through parameters, allowing users to tailor protein_generator to their specific needs.</p></li>
</ul>
</dd>
</dl>
<p>This method is designed to streamline the execution of protein generation processes within the ProtFlow framework, making it easier for researchers and developers to perform and analyze protein generation simulations.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.protein_generator.ProteinGenerator.write_cmd">
<span class="sig-name descname"><span class="pre">write_cmd</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pose_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pose_options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/tools/protein_generator.html#ProteinGenerator.write_cmd"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.protein_generator.ProteinGenerator.write_cmd" title="Link to this definition"></a></dt>
<dd><p>Write the command to run the protein_generator.py script with specified options and pose configurations.</p>
<p>This method constructs the command string necessary to execute the protein generator script, incorporating specified options and pose-specific parameters. The generated command can be used to run the protein generation process in a computational environment.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pose_path</strong> (<em>str</em>) – The file path to the input pose.</p></li>
<li><p><strong>output_dir</strong> (<em>str</em>) – The directory where output files will be stored.</p></li>
<li><p><strong>options</strong> (<em>str</em>) – Additional options for the protein generator script.</p></li>
<li><p><strong>pose_options</strong> (<em>str</em>) – Specific options for the protein generator script related to the pose.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The constructed command string to execute the protein generator script.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>write_cmd</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protein_generator</span> <span class="kn">import</span> <span class="n">ProteinGenerator</span>

<span class="c1"># Initialize the ProteinGenerator class</span>
<span class="n">protein_generator</span> <span class="o">=</span> <span class="n">ProteinGenerator</span><span class="p">(</span>
    <span class="n">script_path</span><span class="o">=</span><span class="s2">&quot;/path/to/protein_generator.py&quot;</span><span class="p">,</span>
    <span class="n">python_path</span><span class="o">=</span><span class="s2">&quot;/path/to/python&quot;</span>
<span class="p">)</span>

<span class="c1"># Construct the command string</span>
<span class="n">cmd</span> <span class="o">=</span> <span class="n">protein_generator</span><span class="o">.</span><span class="n">write_cmd</span><span class="p">(</span>
    <span class="n">pose_path</span><span class="o">=</span><span class="s2">&quot;input_poses/pose1.pdb&quot;</span><span class="p">,</span>
    <span class="n">output_dir</span><span class="o">=</span><span class="s2">&quot;output_directory&quot;</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;generation.num_proteins=10&quot;</span><span class="p">,</span>
    <span class="n">pose_options</span><span class="o">=</span><span class="s2">&quot;generation.input_pdb=&#39;input.pdb&#39;&quot;</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Command Construction:</strong> The method parses the input pose path to derive a description and combines it with provided options and flags to construct a command string.</p></li>
<li><p><strong>Options Parsing:</strong> The options and pose_options parameters are parsed into a format compatible with the protein generator script.</p></li>
<li><p><strong>Output Directory:</strong> The output directory is specified to ensure that generated files are stored in the appropriate location.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
</section>
</dd></dl>

</section>
<section id="module-protflow.tools.residue_selectors">
<span id="protflow-tools-residue-selectors-module"></span><h2>protflow.tools.residue_selectors module<a class="headerlink" href="#module-protflow.tools.residue_selectors" title="Link to this heading"></a></h2>
<p>Module to select residues in a Poses class and add the resulting motif into Poses.df.</p>
<p>This module provides the functionality to select specific residues from protein structures
represented as Poses objects. It includes various selector classes that allow for different
criteria of residue selection, such as by chain or based on existing residue selections.</p>
<dl class="simple">
<dt>Classes:</dt><dd><ul class="simple">
<li><p>ResidueSelector: Abstract base class for all residue selectors.</p></li>
<li><p>ChainSelector: Selects all residues of specified chains.</p></li>
<li><p>TrueSelector: Selects all residues of a pose.</p></li>
<li><p>NotSelector: Selects all residues except those specified by a residue selection.</p></li>
</ul>
</dd>
<dt>Dependencies:</dt><dd><ul class="simple">
<li><p>protflow.residues</p></li>
<li><p>protflow.poses</p></li>
<li><p>protflow.utils.biopython_tools</p></li>
</ul>
</dd>
</dl>
<p>Examples:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Example usage of ChainSelector</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">chain_selector</span> <span class="o">=</span> <span class="n">ChainSelector</span><span class="p">(</span><span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span> <span class="n">chain</span><span class="o">=</span><span class="s1">&#39;A&#39;</span><span class="p">)</span>
<span class="n">chain_selector</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="n">prefix</span><span class="o">=</span><span class="s1">&#39;selected_chain_A&#39;</span><span class="p">)</span>

<span class="c1"># Example usage of TrueSelector</span>
<span class="n">true_selector</span> <span class="o">=</span> <span class="n">TrueSelector</span><span class="p">(</span><span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">)</span>
<span class="n">true_selector</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="n">prefix</span><span class="o">=</span><span class="s1">&#39;all_residues&#39;</span><span class="p">)</span>

<span class="c1"># Example usage of NotSelector</span>
<span class="n">residue_selection</span> <span class="o">=</span> <span class="n">ResidueSelection</span><span class="p">([</span><span class="s1">&#39;A1&#39;</span><span class="p">,</span> <span class="s1">&#39;A2&#39;</span><span class="p">])</span>
<span class="n">not_selector</span> <span class="o">=</span> <span class="n">NotSelector</span><span class="p">(</span><span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span> <span class="n">residue_selection</span><span class="o">=</span><span class="n">residue_selection</span><span class="p">)</span>
<span class="n">not_selector</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="n">prefix</span><span class="o">=</span><span class="s1">&#39;not_selected&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p class="rubric">Notes</p>
<p>This module is part of the ProtFlow package and is designed to work in tandem with other components of the package, especially those related to job management in HPC environments.</p>
<section id="id56">
<h3>Author<a class="headerlink" href="#id56" title="Link to this heading"></a></h3>
<p>Markus Braun, Adrian Tripp</p>
</section>
<section id="id57">
<h3>Version<a class="headerlink" href="#id57" title="Link to this heading"></a></h3>
<p>0.1.0</p>
</section>
<dl class="py class">
<dt class="sig sig-object py" id="protflow.tools.residue_selectors.ChainSelector">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">protflow.tools.residue_selectors.</span></span><span class="sig-name descname"><span class="pre">ChainSelector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chain</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chains</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/tools/residue_selectors.html#ChainSelector"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.residue_selectors.ChainSelector" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#protflow.tools.residue_selectors.ResidueSelector" title="protflow.tools.residue_selectors.ResidueSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">ResidueSelector</span></code></a></p>
<p>Selects all residues of a given chain in Poses.</p>
<p>This class extends ResidueSelector to allow selection of residues based on specific chains
from the protein structures contained in a Poses object.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="protflow.tools.residue_selectors.ChainSelector.poses">
<span class="sig-name descname"><span class="pre">poses</span></span><a class="headerlink" href="#protflow.tools.residue_selectors.ChainSelector.poses" title="Link to this definition"></a></dt>
<dd><p>The Poses object containing the protein structures.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses">Poses</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="protflow.tools.residue_selectors.ChainSelector.chains">
<span class="sig-name descname"><span class="pre">chains</span></span><a class="headerlink" href="#protflow.tools.residue_selectors.ChainSelector.chains" title="Link to this definition"></a></dt>
<dd><p>A list of chain identifiers to select residues from.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="protflow.tools.residue_selectors.ChainSelector.chain">
<span class="sig-name descname"><span class="pre">chain</span></span><a class="headerlink" href="#protflow.tools.residue_selectors.ChainSelector.chain" title="Link to this definition"></a></dt>
<dd><p>A single chain identifier to select residues from.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.residue_selectors.ChainSelector.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chain</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chains</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/tools/residue_selectors.html#ChainSelector.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.residue_selectors.ChainSelector.__init__" title="Link to this definition"></a></dt>
<dd><p>Initialize the ChainSelector with optional Poses object and chain(s).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>poses</strong> (<a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><em>Poses</em></a><em>, </em><em>optional</em>) – The Poses object containing the protein structures. Defaults to None.</p></li>
<li><p><strong>chain</strong> (<em>list</em><em>, </em><em>optional</em>) – A single chain identifier. Defaults to None.</p></li>
<li><p><strong>chains</strong> (<em>list</em><em>, </em><em>optional</em>) – A list of chain identifiers. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If both chain and chains are provided.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">selector</span> <span class="o">=</span> <span class="n">ChainSelector</span><span class="p">(</span><span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span> <span class="n">chain</span><span class="o">=</span><span class="s1">&#39;A&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.residue_selectors.ChainSelector.prep_chain_input">
<span class="sig-name descname"><span class="pre">prep_chain_input</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chain</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chains</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/protflow/tools/residue_selectors.html#ChainSelector.prep_chain_input"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.residue_selectors.ChainSelector.prep_chain_input" title="Link to this definition"></a></dt>
<dd><p>Prepares chain input for chain selection.</p>
<p>This method ensures that method parameters take precedence over class attributes.
This means that ChainSelector.select(chain=”A”) has higher priority than ChainSelector(chain=”C”).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chain</strong> (<em>str</em><em>, </em><em>optional</em>) – A single chain identifier. Defaults to None.</p></li>
<li><p><strong>chains</strong> (<em>list</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – A list of chain identifiers. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The list of chain identifiers to use for selection.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[str]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError</strong> – If both chain and chains are provided.</p></li>
<li><p><strong>ValueError</strong> – If both self.chain and self.chains are set.</p></li>
<li><p><strong>ValueError</strong> – If no chain identifiers are set.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">chains</span> <span class="o">=</span> <span class="n">selector</span><span class="o">.</span><span class="n">prep_chain_input</span><span class="p">(</span><span class="n">chain</span><span class="o">=</span><span class="s1">&#39;A&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.residue_selectors.ChainSelector.select">
<span class="sig-name descname"><span class="pre">select</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chain</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chains</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/protflow/tools/residue_selectors.html#ChainSelector.select"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.residue_selectors.ChainSelector.select" title="Link to this definition"></a></dt>
<dd><p>Selects all residues of a given chain for all poses in a Poses object.</p>
<p>Selected residues are added as ResidueSelection objects under the column <cite>prefix</cite> in Poses.df.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>prefix</strong> (<em>str</em>) – The name of the column that will be added to Poses.df.</p></li>
<li><p><strong>poses</strong> (<a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><em>Poses</em></a><em>, </em><em>optional</em>) – The Poses object containing the protein structures. Defaults to None.</p></li>
<li><p><strong>chain</strong> (<em>str</em><em>, </em><em>optional</em>) – A single chain identifier. Defaults to None.</p></li>
<li><p><strong>chains</strong> (<em>list</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – A list of chain identifiers. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError</strong> – If no poses are provided or set in the instance.</p></li>
<li><p><strong>ValueError</strong> – If both chain and chains are provided.</p></li>
<li><p><strong>KeyError</strong> – If specified chains are not present in the pose.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">selector</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="n">prefix</span><span class="o">=</span><span class="s1">&#39;selected_chain_A&#39;</span><span class="p">,</span> <span class="n">chain</span><span class="o">=</span><span class="s1">&#39;A&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.residue_selectors.ChainSelector.select_single">
<span class="sig-name descname"><span class="pre">select_single</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pose_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chains</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="protflow.html#protflow.residues.ResidueSelection" title="protflow.residues.ResidueSelection"><span class="pre">ResidueSelection</span></a></span></span><a class="reference internal" href="_modules/protflow/tools/residue_selectors.html#ChainSelector.select_single"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.residue_selectors.ChainSelector.select_single" title="Link to this definition"></a></dt>
<dd><p>Selects residues of a given chain of poses and returns them as a ResidueSelection object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pose_path</strong> (<em>str</em>) – The file path to the pose structure.</p></li>
<li><p><strong>chains</strong> (<em>list</em><em>[</em><em>str</em><em>]</em>) – A list of chain identifiers.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The selected residues of the pose.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="protflow.html#protflow.residues.ResidueSelection" title="protflow.residues.ResidueSelection">ResidueSelection</a></p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>KeyError</strong> – If specified chains are not present in the pose.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">selection</span> <span class="o">=</span> <span class="n">selector</span><span class="o">.</span><span class="n">select_single</span><span class="p">(</span><span class="n">pose_path</span><span class="o">=</span><span class="s1">&#39;path/to/pose.pdb&#39;</span><span class="p">,</span> <span class="n">chains</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.residue_selectors.ChainSelector.set_chain">
<span class="sig-name descname"><span class="pre">set_chain</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chain</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/protflow/tools/residue_selectors.html#ChainSelector.set_chain"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.residue_selectors.ChainSelector.set_chain" title="Link to this definition"></a></dt>
<dd><p>Sets a single chain for the select() method.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>chain</strong> (<em>str</em><em>, </em><em>optional</em>) – A single chain identifier. Defaults to None.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If chain is not a single-character string.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">selector</span><span class="o">.</span><span class="n">set_chain</span><span class="p">(</span><span class="n">chain</span><span class="o">=</span><span class="s1">&#39;A&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.residue_selectors.ChainSelector.set_chains">
<span class="sig-name descname"><span class="pre">set_chains</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chains</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/protflow/tools/residue_selectors.html#ChainSelector.set_chains"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.residue_selectors.ChainSelector.set_chains" title="Link to this definition"></a></dt>
<dd><p>Sets chains for the select() method.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>chains</strong> (<em>list</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – A list of chain identifiers to select residues from. Defaults to None.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If chains is not a list of single-character strings.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">selector</span><span class="o">.</span><span class="n">set_chains</span><span class="p">(</span><span class="n">chains</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">])</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="protflow.tools.residue_selectors.NotSelector">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">protflow.tools.residue_selectors.</span></span><span class="sig-name descname"><span class="pre">NotSelector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">residue_selection</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.residues.ResidueSelection" title="protflow.residues.ResidueSelection"><span class="pre">ResidueSelection</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">contig</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/tools/residue_selectors.html#NotSelector"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.residue_selectors.NotSelector" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#protflow.tools.residue_selectors.ResidueSelector" title="protflow.tools.residue_selectors.ResidueSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">ResidueSelector</span></code></a></p>
<p>ResidueSelector that selects all residues except the ones specified by a residue selection.</p>
<p>This class extends ResidueSelector to exclude specified residues from selection.
The excluded residues can be provided either as a ResidueSelection object or
as a contig string.</p>
<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.residue_selectors.NotSelector.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">residue_selection</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.residues.ResidueSelection" title="protflow.residues.ResidueSelection"><span class="pre">ResidueSelection</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">contig</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/tools/residue_selectors.html#NotSelector.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.residue_selectors.NotSelector.__init__" title="Link to this definition"></a></dt>
<dd><p>Initialize the NotSelector with optional Poses object and exclusion criteria.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>poses</strong> (<a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><em>Poses</em></a><em>, </em><em>optional</em>) – The Poses object containing the protein structures. Defaults to None.</p></li>
<li><p><strong>residue_selection</strong> (<a class="reference internal" href="protflow.html#protflow.residues.ResidueSelection" title="protflow.residues.ResidueSelection"><em>ResidueSelection</em></a><em>|</em><em>str</em><em>, </em><em>optional</em>) – The residues to be excluded from selection. Can be a ResidueSelection object or a string. Defaults to None.</p></li>
<li><p><strong>contig</strong> (<em>str</em><em>, </em><em>optional</em>) – A string specifying the residues to be excluded in a contig format. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If both residue_selection and contig are provided.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">residue_selection</span> <span class="o">=</span> <span class="n">ResidueSelection</span><span class="p">([</span><span class="s1">&#39;A1&#39;</span><span class="p">,</span> <span class="s1">&#39;A2&#39;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">selector</span> <span class="o">=</span> <span class="n">NotSelector</span><span class="p">(</span><span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span> <span class="n">residue_selection</span><span class="o">=</span><span class="n">residue_selection</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.residue_selectors.NotSelector.prep_residue_selection">
<span class="sig-name descname"><span class="pre">prep_residue_selection</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">residue_selection</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.residues.ResidueSelection" title="protflow.residues.ResidueSelection"><span class="pre">ResidueSelection</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="protflow.html#protflow.residues.ResidueSelection" title="protflow.residues.ResidueSelection"><span class="pre">ResidueSelection</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/protflow/tools/residue_selectors.html#NotSelector.prep_residue_selection"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.residue_selectors.NotSelector.prep_residue_selection" title="Link to this definition"></a></dt>
<dd><p>Prepares the residue_selection parameter for the select() function.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>residue_selection</strong> (<a class="reference internal" href="protflow.html#protflow.residues.ResidueSelection" title="protflow.residues.ResidueSelection"><em>ResidueSelection</em></a><em>|</em><em>str</em>) – The residues to be excluded from selection. Can be a ResidueSelection object or a string.</p></li>
<li><p><strong>poses</strong> (<a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><em>Poses</em></a>) – The Poses object containing the protein structures.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A list of ResidueSelection objects for each pose.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[<a class="reference internal" href="protflow.html#protflow.residues.ResidueSelection" title="protflow.residues.ResidueSelection">ResidueSelection</a>]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>TypeError</strong> – If the residue_selection parameter is not of a supported type.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">residue_selection_list</span> <span class="o">=</span> <span class="n">selector</span><span class="o">.</span><span class="n">prep_residue_selection</span><span class="p">(</span><span class="n">residue_selection</span><span class="o">=</span><span class="s1">&#39;selected_residues&#39;</span><span class="p">,</span> <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.residue_selectors.NotSelector.select">
<span class="sig-name descname"><span class="pre">select</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">residue_selection</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.residues.ResidueSelection" title="protflow.residues.ResidueSelection"><span class="pre">ResidueSelection</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">contig</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/protflow/tools/residue_selectors.html#NotSelector.select"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.residue_selectors.NotSelector.select" title="Link to this definition"></a></dt>
<dd><p>Selects all residues except the ones specified in residue_selection or by contig.</p>
<p>The parameter residue_selection can be either a ResidueSelection object or a string pointing to a column in the poses.df that contains ResidueSelection objects.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>prefix</strong> (<em>str</em>) – The name of the column that will be added to Poses.df.</p></li>
<li><p><strong>poses</strong> (<a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><em>Poses</em></a><em>, </em><em>optional</em>) – The Poses object containing the protein structures. Defaults to None.</p></li>
<li><p><strong>residue_selection</strong> (<a class="reference internal" href="protflow.html#protflow.residues.ResidueSelection" title="protflow.residues.ResidueSelection"><em>ResidueSelection</em></a><em>|</em><em>str</em><em>, </em><em>optional</em>) – The residues to be excluded from selection. Defaults to None.</p></li>
<li><p><strong>contig</strong> (<em>str</em><em>, </em><em>optional</em>) – A string specifying the residues to be excluded in a contig format. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError</strong> – If no poses are provided or set in the instance.</p></li>
<li><p><strong>ValueError</strong> – If both residue_selection and contig are provided.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">selector</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="n">prefix</span><span class="o">=</span><span class="s1">&#39;not_selected&#39;</span><span class="p">,</span> <span class="n">residue_selection</span><span class="o">=</span><span class="s1">&#39;selected_residues&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.residue_selectors.NotSelector.select_single">
<span class="sig-name descname"><span class="pre">select_single</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pose_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">residue_selection</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.residues.ResidueSelection" title="protflow.residues.ResidueSelection"><span class="pre">ResidueSelection</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="protflow.html#protflow.residues.ResidueSelection" title="protflow.residues.ResidueSelection"><span class="pre">ResidueSelection</span></a></span></span><a class="reference internal" href="_modules/protflow/tools/residue_selectors.html#NotSelector.select_single"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.residue_selectors.NotSelector.select_single" title="Link to this definition"></a></dt>
<dd><p>Selects all residues except the ones specified in residue_selection or by contig.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pose_path</strong> (<em>str</em>) – The file path to the pose structure.</p></li>
<li><p><strong>residue_selection</strong> (<a class="reference internal" href="protflow.html#protflow.residues.ResidueSelection" title="protflow.residues.ResidueSelection"><em>ResidueSelection</em></a>) – The residues to be excluded from selection.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The selected residues of the pose.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="protflow.html#protflow.residues.ResidueSelection" title="protflow.residues.ResidueSelection">ResidueSelection</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">selection</span> <span class="o">=</span> <span class="n">selector</span><span class="o">.</span><span class="n">select_single</span><span class="p">(</span><span class="n">pose_path</span><span class="o">=</span><span class="s1">&#39;path/to/pose.pdb&#39;</span><span class="p">,</span> <span class="n">residue_selection</span><span class="o">=</span><span class="n">residue_selection</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.residue_selectors.NotSelector.set_contig">
<span class="sig-name descname"><span class="pre">set_contig</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">contig</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/protflow/tools/residue_selectors.html#NotSelector.set_contig"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.residue_selectors.NotSelector.set_contig" title="Link to this definition"></a></dt>
<dd><p>Sets the contig attribute for the NotSelector class.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>contig</strong> (<em>str</em>) – A string specifying the residues to be excluded in a contig format.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If contig is not a string.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">selector</span><span class="o">.</span><span class="n">set_contig</span><span class="p">(</span><span class="n">contig</span><span class="o">=</span><span class="s1">&#39;A1-7,A25-109,B45-50,C1,C3,C5&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.residue_selectors.NotSelector.set_residue_selection">
<span class="sig-name descname"><span class="pre">set_residue_selection</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">residue_selection</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.residues.ResidueSelection" title="protflow.residues.ResidueSelection"><span class="pre">ResidueSelection</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/protflow/tools/residue_selectors.html#NotSelector.set_residue_selection"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.residue_selectors.NotSelector.set_residue_selection" title="Link to this definition"></a></dt>
<dd><p>Sets the residue_selection attribute for the NotSelector class.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>residue_selection</strong> (<a class="reference internal" href="protflow.html#protflow.residues.ResidueSelection" title="protflow.residues.ResidueSelection"><em>ResidueSelection</em></a><em>, </em><em>optional</em>) – The residues to be excluded from selection. Defaults to None.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">residue_selection</span> <span class="o">=</span> <span class="n">ResidueSelection</span><span class="p">([</span><span class="s1">&#39;A1&#39;</span><span class="p">,</span> <span class="s1">&#39;A2&#39;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">selector</span><span class="o">.</span><span class="n">set_residue_selection</span><span class="p">(</span><span class="n">residue_selection</span><span class="o">=</span><span class="n">residue_selection</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="protflow.tools.residue_selectors.ResidueSelector">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">protflow.tools.residue_selectors.</span></span><span class="sig-name descname"><span class="pre">ResidueSelector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/tools/residue_selectors.html#ResidueSelector"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.residue_selectors.ResidueSelector" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Abstract base class for ResidueSelectors.</p>
<p>All ResidueSelector classes must implement a select() method that selects residues from Poses.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="protflow.tools.residue_selectors.ResidueSelector.poses">
<span class="sig-name descname"><span class="pre">poses</span></span><a class="headerlink" href="#protflow.tools.residue_selectors.ResidueSelector.poses" title="Link to this definition"></a></dt>
<dd><p>The Poses object containing the protein structures.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses">Poses</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.residue_selectors.ResidueSelector.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/tools/residue_selectors.html#ResidueSelector.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.residue_selectors.ResidueSelector.__init__" title="Link to this definition"></a></dt>
<dd><p>Initializes the ResidueSelector with an optional Poses object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>poses</strong> (<a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><em>Poses</em></a><em>, </em><em>optional</em>) – The Poses object containing the protein structures. Defaults to None.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>poses = Poses()
selector = ResidueSelector(poses=poses)</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.residue_selectors.ResidueSelector.select">
<span class="sig-name descname"><span class="pre">select</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/protflow/tools/residue_selectors.html#ResidueSelector.select"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.residue_selectors.ResidueSelector.select" title="Link to this definition"></a></dt>
<dd><p>Abstract method to select residues in poses.</p>
<p>This method must be implemented by subclasses. The selected residues will be added as
ResidueSelection objects under the column <cite>prefix</cite> in <cite>Poses.df</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>prefix</strong> (<em>str</em>) – The name of the column that will be added to Poses.df. Poses.df[prefix] holds the selected Residues as a ResidueSelection object.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>NotImplementedError</strong> – If the method is not implemented by a subclass.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">class</span> <span class="nc">MySelector</span><span class="p">(</span><span class="n">ResidueSelector</span><span class="p">):</span>
<span class="gp">... </span>    <span class="k">def</span> <span class="nf">select</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">prefix</span><span class="p">):</span>
<span class="gp">... </span>        <span class="c1"># implementation here</span>
<span class="gp">... </span>        <span class="k">pass</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">selector</span> <span class="o">=</span> <span class="n">MySelector</span><span class="p">(</span><span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">selector</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="n">prefix</span><span class="o">=</span><span class="s1">&#39;selected_residues&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.residue_selectors.ResidueSelector.select_single">
<span class="sig-name descname"><span class="pre">select_single</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="protflow.html#protflow.residues.ResidueSelection" title="protflow.residues.ResidueSelection"><span class="pre">ResidueSelection</span></a></span></span><a class="reference internal" href="_modules/protflow/tools/residue_selectors.html#ResidueSelector.select_single"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.residue_selectors.ResidueSelector.select_single" title="Link to this definition"></a></dt>
<dd><p>Abstract method to select residues for a single pose.</p>
<p>This method must be implemented by subclasses. It returns a ResidueSelection that contains the selected residues of the pose.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The selected residues of the pose.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="protflow.html#protflow.residues.ResidueSelection" title="protflow.residues.ResidueSelection">ResidueSelection</a></p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>NotImplementedError</strong> – If the method is not implemented by a subclass.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">class</span> <span class="nc">MySelector</span><span class="p">(</span><span class="n">ResidueSelector</span><span class="p">):</span>
<span class="gp">... </span>    <span class="k">def</span> <span class="nf">select_single</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">):</span>
<span class="gp">... </span>        <span class="c1"># implementation here</span>
<span class="gp">... </span>        <span class="k">return</span> <span class="n">ResidueSelection</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">selector</span> <span class="o">=</span> <span class="n">MySelector</span><span class="p">(</span><span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">selection</span> <span class="o">=</span> <span class="n">selector</span><span class="o">.</span><span class="n">select_single</span><span class="p">()</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.residue_selectors.ResidueSelector.set_poses">
<span class="sig-name descname"><span class="pre">set_poses</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/protflow/tools/residue_selectors.html#ResidueSelector.set_poses"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.residue_selectors.ResidueSelector.set_poses" title="Link to this definition"></a></dt>
<dd><p>Sets the poses for the ResidueSelector class.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>poses</strong> (<a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><em>Poses</em></a><em>, </em><em>optional</em>) – The Poses object containing the protein structures. Defaults to None.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>TypeError</strong> – If the poses parameter is not of type Poses.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>poses = Poses()
selector.set_poses(poses=poses)</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="protflow.tools.residue_selectors.TrueSelector">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">protflow.tools.residue_selectors.</span></span><span class="sig-name descname"><span class="pre">TrueSelector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/tools/residue_selectors.html#TrueSelector"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.residue_selectors.TrueSelector" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#protflow.tools.residue_selectors.ResidueSelector" title="protflow.tools.residue_selectors.ResidueSelector"><code class="xref py py-class docutils literal notranslate"><span class="pre">ResidueSelector</span></code></a></p>
<p>ResidueSelector that selects all residues of a pose.</p>
<p>This class extends ResidueSelector to select all residues from each pose in a Poses object.
It adds the selected residues as ResidueSelection objects under a specified column in Poses.df.</p>
<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.residue_selectors.TrueSelector.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/tools/residue_selectors.html#TrueSelector.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.residue_selectors.TrueSelector.__init__" title="Link to this definition"></a></dt>
<dd><p>Initialize the TrueSelector with an optional Poses object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>poses</strong> (<a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><em>Poses</em></a><em>, </em><em>optional</em>) – The Poses object containing the protein structures. Defaults to None.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">selector</span> <span class="o">=</span> <span class="n">TrueSelector</span><span class="p">(</span><span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.residue_selectors.TrueSelector.select">
<span class="sig-name descname"><span class="pre">select</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/tools/residue_selectors.html#TrueSelector.select"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.residue_selectors.TrueSelector.select" title="Link to this definition"></a></dt>
<dd><p>Selects all residues of a given pose for all poses in a Poses object.</p>
<p>Selected residues are added as ResidueSelection objects under the column <cite>prefix</cite> in Poses.df.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>prefix</strong> (<em>str</em>) – The name of the column that will be added to Poses.df.</p></li>
<li><p><strong>poses</strong> (<a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><em>Poses</em></a><em>, </em><em>optional</em>) – The Poses object containing the protein structures. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If no poses are provided or set in the instance.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">selector</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="n">prefix</span><span class="o">=</span><span class="s1">&#39;all_residues&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.residue_selectors.TrueSelector.select_single">
<span class="sig-name descname"><span class="pre">select_single</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pose_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="protflow.html#protflow.residues.ResidueSelection" title="protflow.residues.ResidueSelection"><span class="pre">ResidueSelection</span></a></span></span><a class="reference internal" href="_modules/protflow/tools/residue_selectors.html#TrueSelector.select_single"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.residue_selectors.TrueSelector.select_single" title="Link to this definition"></a></dt>
<dd><p>Selects all residues in a single pose and returns them as a ResidueSelection object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>pose_path</strong> (<em>str</em>) – The file path to the pose structure.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The selected residues of the pose.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="protflow.html#protflow.residues.ResidueSelection" title="protflow.residues.ResidueSelection">ResidueSelection</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">selection</span> <span class="o">=</span> <span class="n">selector</span><span class="o">.</span><span class="n">select_single</span><span class="p">(</span><span class="n">pose_path</span><span class="o">=</span><span class="s1">&#39;path/to/pose.pdb&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

</section>
<section id="module-protflow.tools.rfdiffusion">
<span id="protflow-tools-rfdiffusion-module"></span><h2>protflow.tools.rfdiffusion module<a class="headerlink" href="#module-protflow.tools.rfdiffusion" title="Link to this heading"></a></h2>
<section id="rfdiffusion-module">
<h3>RFdiffusion Module<a class="headerlink" href="#rfdiffusion-module" title="Link to this heading"></a></h3>
<p>This module provides the functionality to integrate RFdiffusion within the ProtFlow framework. It offers tools to run RFdiffusion, handle its inputs and outputs, and process the resulting data in a structured and automated manner.</p>
<section id="id58">
<h4>Detailed Description<a class="headerlink" href="#id58" title="Link to this heading"></a></h4>
<p>The <cite>RFdiffusion</cite> class encapsulates the functionality necessary to execute RFdiffusion runs. It manages the configuration of paths to essential scripts and Python executables, sets up the environment, and handles the execution of diffusion processes. It also includes methods for collecting and processing output data, ensuring that the results are organized and accessible for further analysis within the ProtFlow ecosystem.
The module is designed to streamline the integration of RFdiffusion into larger computational workflows. It supports the automatic setup of job parameters, execution of RFdiffusion commands, and parsing of output files into a structured DataFrame format. This facilitates subsequent data analysis and visualization steps.</p>
</section>
<section id="id59">
<h4>Usage<a class="headerlink" href="#id59" title="Link to this heading"></a></h4>
<p>To use this module, create an instance of the <cite>RFdiffusion</cite> class and invoke its <cite>run</cite> method with appropriate parameters. The module will handle the configuration, execution, and result collection processes. Detailed control over the diffusion process is provided through various parameters, allowing for customized runs tailored to specific research needs.</p>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize and use the <cite>RFdiffusion</cite> class within a ProtFlow pipeline:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">rfdiffusion</span> <span class="kn">import</span> <span class="n">RFdiffusion</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the RFdiffusion class</span>
<span class="n">rfdiffusion</span> <span class="o">=</span> <span class="n">RFdiffusion</span><span class="p">()</span>

<span class="c1"># Run the diffusion process</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">rfdiffusion</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">num_diffusions</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;inference.num_designs=10&quot;</span><span class="p">,</span>
    <span class="n">pose_options</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;inference.input_pdb=&#39;input.pdb&#39;&quot;</span><span class="p">],</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id60">
<h4>Further Details<a class="headerlink" href="#id60" title="Link to this heading"></a></h4>
<blockquote>
<div><ul class="simple">
<li><p>Edge Cases: The module handles various edge cases, such as empty pose lists and the need to overwrite previous results. It ensures robust error handling and logging for easier debugging and verification of the diffusion process.</p></li>
<li><p>Customizability: Users can customize the diffusion process through multiple parameters, including the number of diffusions, specific options for the RFdiffusion script, and options for handling pose-specific parameters.</p></li>
<li><p>Integration: The module seamlessly integrates with other components of the ProtFlow framework, leveraging shared configurations and data structures to provide a cohesive user experience.</p></li>
</ul>
</div></blockquote>
<p>This module is intended for researchers and developers who need to incorporate RFdiffusion into their protein design and analysis workflows. By automating many of the setup and execution steps, it allows users to focus on interpreting results and advancing their scientific inquiries.</p>
<p class="rubric">Notes</p>
<p>This module is part of the ProtFlow package and is designed to work in tandem with other components of the package, especially those related to job management in HPC environments.</p>
</section>
<section id="id61">
<h4>Authors<a class="headerlink" href="#id61" title="Link to this heading"></a></h4>
<p>Markus Braun, Adrian Tripp</p>
</section>
<section id="id62">
<h4>Version<a class="headerlink" href="#id62" title="Link to this heading"></a></h4>
<p>0.1.0</p>
</section>
</section>
<dl class="py class">
<dt class="sig sig-object py" id="protflow.tools.rfdiffusion.RFdiffusion">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">protflow.tools.rfdiffusion.</span></span><span class="sig-name descname"><span class="pre">RFdiffusion</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">script_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">python_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><span class="pre">JobStarter</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/tools/rfdiffusion.html#RFdiffusion"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.rfdiffusion.RFdiffusion" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="protflow.html#protflow.runners.Runner" title="protflow.runners.Runner"><code class="xref py py-class docutils literal notranslate"><span class="pre">Runner</span></code></a></p>
<section id="rfdiffusion-class">
<h3>RFdiffusion Class<a class="headerlink" href="#rfdiffusion-class" title="Link to this heading"></a></h3>
<p>The <cite>RFdiffusion</cite> class is a specialized class designed to facilitate the execution of RFdiffusion within the ProtFlow framework. It extends the <cite>Runner</cite> class and incorporates specific methods to handle the setup, execution, and data collection associated with RFdiffusion processes.</p>
<section id="id63">
<h4>Detailed Description<a class="headerlink" href="#id63" title="Link to this heading"></a></h4>
<p>The <cite>RFdiffusion</cite> class manages all aspects of running RFdiffusion simulations. It handles the configuration of necessary scripts and executables, prepares the environment for diffusion processes, and executes the diffusion commands. Additionally, it collects and processes the output data, organizing it into a structured format for further analysis.</p>
<dl class="simple">
<dt>Key functionalities include:</dt><dd><ul class="simple">
<li><p>Setting up paths to RFdiffusion scripts and Python executables.</p></li>
<li><p>Configuring job starter options, either automatically or manually.</p></li>
<li><p>Handling the execution of RFdiffusion commands with support for multiple diffusions.</p></li>
<li><p>Collecting and processing output data into a pandas DataFrame.</p></li>
<li><p>Updating motifs based on RFdiffusion outputs and remapping residue selections.</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">rtype<span class="colon">:</span></dt>
<dd class="field-odd"><p>An instance of the <cite>RFdiffusion</cite> class, configured to run RFdiffusion processes and handle outputs efficiently.</p>
</dd>
<dt class="field-even">raises - FileNotFoundError<span class="colon">:</span></dt>
<dd class="field-even"><p>If required files or directories are not found during the execution process.:</p>
</dd>
<dt class="field-odd">raises - ValueError<span class="colon">:</span></dt>
<dd class="field-odd"><p>If invalid arguments are provided to the methods.:</p>
</dd>
<dt class="field-even">raises - TypeError<span class="colon">:</span></dt>
<dd class="field-even"><p>If motifs are not of the expected type.:</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize and use the <cite>RFdiffusion</cite> class:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">rfdiffusion</span> <span class="kn">import</span> <span class="n">RFdiffusion</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the RFdiffusion class</span>
<span class="n">rfdiffusion</span> <span class="o">=</span> <span class="n">RFdiffusion</span><span class="p">()</span>

<span class="c1"># Run the diffusion process</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">rfdiffusion</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">num_diffusions</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;inference.num_designs=10&quot;</span><span class="p">,</span>
    <span class="n">pose_options</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;inference.input_pdb=&#39;input.pdb&#39;&quot;</span><span class="p">],</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id64">
<h4>Further Details<a class="headerlink" href="#id64" title="Link to this heading"></a></h4>
<blockquote>
<div><ul class="simple">
<li><p>Edge Cases: The class includes handling for various edge cases, such as empty pose lists, the need to overwrite previous results, and the presence of existing score files.</p></li>
<li><p>Customization: The class provides extensive customization options through its parameters, allowing users to tailor the diffusion process to their specific needs.</p></li>
<li><p>Integration: Seamlessly integrates with other ProtFlow components, leveraging shared configurations and data structures for a unified workflow.</p></li>
</ul>
</div></blockquote>
<p>The RFdiffusion class is intended for researchers and developers who need to perform RFdiffusion simulations as part of their protein design and analysis workflows. It simplifies the process, allowing users to focus on analyzing results and advancing their research.</p>
<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.rfdiffusion.RFdiffusion.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">script_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">python_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><span class="pre">JobStarter</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/protflow/tools/rfdiffusion.html#RFdiffusion.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.rfdiffusion.RFdiffusion.__init__" title="Link to this definition"></a></dt>
<dd><p>Initialize the RFdiffusion class.</p>
<p>This constructor sets up the necessary paths to the RFdiffusion script and Python executable, and initializes the job starter. The paths are configured using default values from the ProtFlow configuration, but they can be manually set if required. However, manual setting is generally not recommended due to the potential for misconfiguration.</p>
<section id="id65">
<h5>Detailed Description<a class="headerlink" href="#id65" title="Link to this heading"></a></h5>
<p>The <cite>__init__</cite> method initializes the RFdiffusion class by setting up essential paths and configurations. It ensures that the paths to the RFdiffusion script and Python executable are correctly set, and it initializes the job starter object. This setup is crucial for the proper execution of RFdiffusion processes within the ProtFlow framework.</p>
<dl class="field-list simple">
<dt class="field-odd">param script_path (str<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type script_path (str<span class="colon">:</span></dt>
<dd class="field-even"><p>The path to the RFdiffusion script. Defaults to the value specified in the ProtFlow configuration (<cite>protflow.config.RFDIFFUSION_SCRIPT_PATH</cite>).</p>
</dd>
<dt class="field-odd">param optional)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type optional)<span class="colon">:</span></dt>
<dd class="field-even"><p>The path to the RFdiffusion script. Defaults to the value specified in the ProtFlow configuration (<cite>protflow.config.RFDIFFUSION_SCRIPT_PATH</cite>).</p>
</dd>
<dt class="field-odd">param python_path (str<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type python_path (str<span class="colon">:</span></dt>
<dd class="field-even"><p>The path to the Python executable used to run the RFdiffusion script. Defaults to the value specified in the ProtFlow configuration (<cite>protflow.config.RFDIFFUSION_PYTHON_PATH</cite>).</p>
</dd>
<dt class="field-odd">param optional)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type optional)<span class="colon">:</span></dt>
<dd class="field-even"><p>The path to the Python executable used to run the RFdiffusion script. Defaults to the value specified in the ProtFlow configuration (<cite>protflow.config.RFDIFFUSION_PYTHON_PATH</cite>).</p>
</dd>
<dt class="field-odd">param jobstarter (JobStarter<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type jobstarter (JobStarter<span class="colon">:</span></dt>
<dd class="field-even"><p>An instance of the JobStarter class, which manages job execution. If not provided, the default is <cite>None</cite>.</p>
</dd>
<dt class="field-odd">param optional)<span class="colon">:</span></dt>
<dd class="field-odd"><p></p></dd>
<dt class="field-even">type optional)<span class="colon">:</span></dt>
<dd class="field-even"><p>An instance of the JobStarter class, which manages job execution. If not provided, the default is <cite>None</cite>.</p>
</dd>
<dt class="field-odd">rtype<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
<dt class="field-even">raises ValueError<span class="colon">:</span></dt>
<dd class="field-even"><p>If the provided paths are invalid or if there are issues with the configuration.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize the RFdiffusion class:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">rfdiffusion</span> <span class="kn">import</span> <span class="n">RFdiffusion</span>

<span class="c1"># Initialize the RFdiffusion class</span>
<span class="n">rfdiffusion</span> <span class="o">=</span> <span class="n">RFdiffusion</span><span class="p">()</span>

<span class="c1"># Initialize with custom paths</span>
<span class="n">custom_rfdiffusion</span> <span class="o">=</span> <span class="n">RFdiffusion</span><span class="p">(</span>
    <span class="n">script_path</span><span class="o">=</span><span class="s2">&quot;/path/to/custom/rfdiffusion_script.py&quot;</span><span class="p">,</span>
    <span class="n">python_path</span><span class="o">=</span><span class="s2">&quot;/path/to/custom/python&quot;</span>
<span class="p">)</span>

<span class="c1"># Initialize with a job starter</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>
<span class="n">rfdiffusion_with_jobstarter</span> <span class="o">=</span> <span class="n">RFdiffusion</span><span class="p">(</span><span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id66">
<h5>Further Details<a class="headerlink" href="#id66" title="Link to this heading"></a></h5>
<ul class="simple">
<li><p><strong>Path Configuration:</strong> The paths to the RFdiffusion script and Python executable are critical for the correct functioning of the class. It is recommended to use the default paths provided by the ProtFlow configuration unless there is a specific need to customize them.</p></li>
<li><p><strong>JobStarter Integration:</strong> The JobStarter object is used to manage job execution, ensuring that RFdiffusion processes are handled efficiently. If a JobStarter is not provided, the class will operate without it, but it is recommended to use one for better job management.</p></li>
</ul>
<p>This method is designed for initializing the RFdiffusion class with the necessary configurations, making it ready for executing RFdiffusion processes within the ProtFlow framework.</p>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.rfdiffusion.RFdiffusion.collect_scores">
<span class="sig-name descname"><span class="pre">collect_scores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">work_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rename_pdbs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/protflow/tools/rfdiffusion.html#RFdiffusion.collect_scores"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.rfdiffusion.RFdiffusion.collect_scores" title="Link to this definition"></a></dt>
<dd><p>Collect scores from RFdiffusion output files.</p>
<p>This method collects scores from .trb files generated by RFdiffusion into a single pandas DataFrame. It also optionally renames the output .pdb files based on the diffusion process.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>work_dir</strong> (<em>str</em>) – The working directory where RFdiffusion output files are stored.</p></li>
<li><p><strong>rename_pdbs</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, rename the .pdb files based on the new descriptions. Defaults to True.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A DataFrame containing the collected scores from the RFdiffusion output.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>FileNotFoundError</strong> – If no .pdb files are found in the specified directory.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>collect_scores</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">work_dir</span> <span class="o">=</span> <span class="s2">&quot;/path/to/output&quot;</span>
<span class="n">scores_df</span> <span class="o">=</span> <span class="n">rfdiffusion</span><span class="o">.</span><span class="n">collect_scores</span><span class="p">(</span><span class="n">work_dir</span><span class="p">,</span> <span class="n">rename_pdbs</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="c1"># scores_df will contain the combined scores from the RFdiffusion output files</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Score Collection:</strong> The method iterates over .pdb files in the specified directory, collecting corresponding .trb files and concatenating their scores into a DataFrame.</p></li>
<li><p><strong>File Renaming:</strong> If <cite>rename_pdbs</cite> is set to True, the method renames the .pdb files based on new descriptions to ensure unique identification.</p></li>
<li><p><strong>DataFrame Structure:</strong> The resulting DataFrame includes relevant score information, and columns are renamed appropriately if files are renamed.</p></li>
</ul>
</dd>
</dl>
<p>This method is designed to streamline the collection and organization of RFdiffusion output scores, facilitating further analysis and processing.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.rfdiffusion.RFdiffusion.parse_rfdiffusion_opts">
<span class="sig-name descname"><span class="pre">parse_rfdiffusion_opts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pose_options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span></span></span><a class="reference internal" href="_modules/protflow/tools/rfdiffusion.html#RFdiffusion.parse_rfdiffusion_opts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.rfdiffusion.RFdiffusion.parse_rfdiffusion_opts" title="Link to this definition"></a></dt>
<dd><p>Parse and combine general and pose-specific RFdiffusion options into a dictionary.</p>
<p>This method splits and processes both general options and pose-specific options, combining them into a single dictionary. Pose-specific options will overwrite general options if there are conflicts.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>options</strong> (<em>str</em>) – General options for the RFdiffusion script.</p></li>
<li><p><strong>pose_options</strong> (<em>str</em>) – Pose-specific options for the RFdiffusion script.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dictionary containing the combined options, with pose-specific options taking precedence over general options.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>parse_rfdiffusion_opts</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">options</span> <span class="o">=</span> <span class="s2">&quot;inference.num_designs=10 inference.use_gpu=True&quot;</span>
<span class="n">pose_options</span> <span class="o">=</span> <span class="s2">&quot;inference.input_pdb=&#39;input.pdb&#39; inference.num_designs=5&quot;</span>
<span class="n">parsed_opts</span> <span class="o">=</span> <span class="n">rfdiffusion</span><span class="o">.</span><span class="n">parse_rfdiffusion_opts</span><span class="p">(</span><span class="n">options</span><span class="p">,</span> <span class="n">pose_options</span><span class="p">)</span>
<span class="c1"># parsed_opts will be:</span>
<span class="c1"># {&#39;inference.num_designs&#39;: &#39;5&#39;, &#39;inference.use_gpu&#39;: &#39;True&#39;, &#39;inference.input_pdb&#39;: &quot;&#39;input.pdb&#39;&quot;}</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Option Splitting:</strong> The method uses regular expressions to split the options string into individual option entries, ensuring that options within quotes are not split incorrectly.</p></li>
<li><p><strong>Option Overwriting:</strong> By adding pose-specific options after general options, the method ensures that pose-specific options can overwrite general options if necessary.</p></li>
</ul>
</dd>
</dl>
<p>This method is designed to create a consolidated dictionary of options for the RFdiffusion script, facilitating the construction of command strings with the appropriate parameters.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.rfdiffusion.RFdiffusion.remap_motifs">
<span class="sig-name descname"><span class="pre">remap_motifs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">motifs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/protflow/tools/rfdiffusion.html#RFdiffusion.remap_motifs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.rfdiffusion.RFdiffusion.remap_motifs" title="Link to this definition"></a></dt>
<dd><p>Updates ResidueSelection type motifs in poses.df when given prefix of RFdiffusion run.</p>
<p>This method updates the residue mappings of specified motifs in the poses DataFrame based on the RFdiffusion outputs. It ensures that the motifs are correctly mapped to the new residue indices as generated by the diffusion process.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>poses</strong> (<a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><em>Poses</em></a>) – The Poses object containing the protein structures and associated data.</p></li>
<li><p><strong>motifs</strong> (<em>list</em>) – A list of motif columns to update in the poses DataFrame.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – The prefix used to identify the relevant columns in the poses DataFrame from the RFdiffusion outputs.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>TypeError</strong> – If the motifs are not of the expected type ResidueSelection.</p>
</dd>
</dl>
<p>Further Details:
- <strong>Motif Update:</strong> The method ensures that the specified motifs in the poses DataFrame are updated with the new residue mappings from the RFdiffusion outputs.
- <strong>Residue Mapping:</strong> The method uses the reference and halogenated residue indices generated by RFdiffusion to update the motifs.
- <strong>Integration:</strong> This method integrates seamlessly with the RFdiffusion workflow, ensuring that motifs are correctly remapped after diffusion processes.</p>
<p>This method is designed to update the residue mappings of motifs in the poses DataFrame, facilitating accurate representation of the protein structures after RFdiffusion processes.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.rfdiffusion.RFdiffusion.run">
<span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><span class="pre">JobStarter</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_diffusions</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pose_options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">multiplex_poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">update_motifs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="protflow.html#protflow.runners.RunnerOutput" title="protflow.runners.RunnerOutput"><span class="pre">RunnerOutput</span></a></span></span><a class="reference internal" href="_modules/protflow/tools/rfdiffusion.html#RFdiffusion.run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.rfdiffusion.RFdiffusion.run" title="Link to this definition"></a></dt>
<dd><p>Execute the RFdiffusion process with given poses and jobstarter configuration.</p>
<p>This method sets up and runs the RFdiffusion process using the provided poses and jobstarter object. It handles the configuration, execution, and collection of output data, ensuring that the results are organized and accessible for further analysis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>poses</strong> (<a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><em>Poses</em></a><em>, </em><em>optional</em>) – The Poses object containing the protein structures. Defaults to None.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – A prefix used to name and organize the output files.</p></li>
<li><p><strong>jobstarter</strong> (<a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><em>JobStarter</em></a><em>, </em><em>optional</em>) – An instance of the JobStarter class, which manages job execution. Defaults to None.</p></li>
<li><p><strong>num_diffusions</strong> (<em>int</em><em>, </em><em>optional</em>) – The number of diffusions to run for each input pose. Defaults to 1.</p></li>
<li><p><strong>options</strong> (<em>str</em><em>, </em><em>optional</em>) – Additional options for the RFdiffusion script. Defaults to None.</p></li>
<li><p><strong>pose_options</strong> (<em>list</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – A list of pose-specific options for the RFdiffusion script. Defaults to None.</p></li>
<li><p><strong>overwrite</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, overwrite existing output files. Defaults to False.</p></li>
<li><p><strong>multiplex_poses</strong> (<em>int</em><em>, </em><em>optional</em>) – If specified, create multiple copies of poses to fully utilize parallel computing. Defaults to None.</p></li>
<li><p><strong>update_motifs</strong> (<em>list</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em>) – A list of motifs to update based on the RFdiffusion outputs. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An instance of the RunnerOutput class, containing the processed poses and results of the RFdiffusion process.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="protflow.html#protflow.runners.RunnerOutput" title="protflow.runners.RunnerOutput">RunnerOutput</a></p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>FileNotFoundError</strong> – If required files or directories are not found during the execution process.</p></li>
<li><p><strong>ValueError</strong> – If invalid arguments are provided to the method.</p></li>
<li><p><strong>TypeError</strong> – If motifs are not of the expected type.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>run</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">rfdiffusion</span> <span class="kn">import</span> <span class="n">RFdiffusion</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the RFdiffusion class</span>
<span class="n">rfdiffusion</span> <span class="o">=</span> <span class="n">RFdiffusion</span><span class="p">()</span>

<span class="c1"># Run the diffusion process</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">rfdiffusion</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">num_diffusions</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;inference.num_designs=10&quot;</span><span class="p">,</span>
    <span class="n">pose_options</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;inference.input_pdb=&#39;input.pdb&#39;&quot;</span><span class="p">],</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Setup and Execution:</strong> The method ensures that the environment is correctly set up, directories are prepared, and necessary commands are constructed and executed.</p></li>
<li><p><strong>Output Management:</strong> The method handles the collection and processing of output data, ensuring that results are organized and accessible for further analysis.</p></li>
<li><p><strong>Customization:</strong> Extensive customization options are provided through parameters, allowing users to tailor the diffusion process to their specific needs.</p></li>
</ul>
</dd>
</dl>
<p>This method is designed to streamline the execution of RFdiffusion processes within the ProtFlow framework, making it easier for researchers and developers to perform and analyze diffusion simulations.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.rfdiffusion.RFdiffusion.write_cmd">
<span class="sig-name descname"><span class="pre">write_cmd</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pose_opts</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_diffusions</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">str</span></span></span><a class="reference internal" href="_modules/protflow/tools/rfdiffusion.html#RFdiffusion.write_cmd"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.rfdiffusion.RFdiffusion.write_cmd" title="Link to this definition"></a></dt>
<dd><p>Construct the command to run the RFdiffusion process.</p>
<p>This method constructs the command string required to execute the RFdiffusion process. It combines the specified options and pose-specific options, ensuring that all necessary parameters are included.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pose</strong> (<em>str</em>) – The path to the input pose file.</p></li>
<li><p><strong>options</strong> (<em>str</em>) – General options for the RFdiffusion script.</p></li>
<li><p><strong>pose_opts</strong> (<em>str</em>) – Pose-specific options for the RFdiffusion script.</p></li>
<li><p><strong>output_dir</strong> (<em>str</em>) – The directory where output files will be saved.</p></li>
<li><p><strong>num_diffusions</strong> (<em>int</em><em>, </em><em>optional</em>) – The number of diffusions to run for each input pose. Defaults to 1.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The constructed command string to execute the RFdiffusion process.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the provided options or pose_opts are invalid.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Construct a command for RFdiffusion:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">cmd</span> <span class="o">=</span> <span class="n">rfdiffusion</span><span class="o">.</span><span class="n">write_cmd</span><span class="p">(</span><span class="s2">&quot;input.pdb&quot;</span><span class="p">,</span> <span class="s2">&quot;inference.num_designs=10&quot;</span><span class="p">,</span> <span class="s2">&quot;inference.input_pdb=&#39;input.pdb&#39;&quot;</span><span class="p">,</span> <span class="s2">&quot;/output&quot;</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Option Parsing:</strong> The method parses both general and pose-specific options, ensuring that they are correctly formatted and included in the command string.</p></li>
<li><p><strong>Command Construction:</strong> The constructed command string includes the path to the RFdiffusion script, the specified options, and the output directory.</p></li>
<li><p><strong>Default Values:</strong> Default values for unspecified options, such as <cite>inference.num_designs</cite>, are included to ensure the command string is complete.</p></li>
</ul>
</dd>
</dl>
<p>This method is designed to create a fully-formed command string for running RFdiffusion, making it easier to execute diffusion processes with the desired parameters.</p>
</dd></dl>

</section>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="protflow.tools.rfdiffusion.get_residue_mapping">
<span class="sig-prename descclassname"><span class="pre">protflow.tools.rfdiffusion.</span></span><span class="sig-name descname"><span class="pre">get_residue_mapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">con_ref_idx</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">con_hal_idx</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span></span></span><a class="reference internal" href="_modules/protflow/tools/rfdiffusion.html#get_residue_mapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.rfdiffusion.get_residue_mapping" title="Link to this definition"></a></dt>
<dd><p>Create a residue mapping dictionary from RFdiffusion outputs.</p>
<p>This method creates a mapping dictionary that maps old residue indices (from con_ref_idx) to new residue indices (from con_hal_idx).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>con_ref_idx</strong> (<em>list</em>) – A list of reference residue indices from the RFdiffusion outputs, where each element is a tuple of (chain, residue_id).</p></li>
<li><p><strong>con_hal_idx</strong> (<em>list</em>) – A list of halogenated residue indices from the RFdiffusion outputs, where each element is a tuple of (chain, residue_id).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dictionary where keys are tuples of (chain, residue_id) from con_ref_idx and values are tuples of (chain, residue_id) from con_hal_idx.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>get_residue_mapping</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">con_ref_idx</span> <span class="o">=</span> <span class="p">[(</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="mi">10</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="mi">20</span><span class="p">)]</span>
<span class="n">con_hal_idx</span> <span class="o">=</span> <span class="p">[(</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="mi">11</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="mi">21</span><span class="p">)]</span>
<span class="n">residue_mapping</span> <span class="o">=</span> <span class="n">rfdiffusion</span><span class="o">.</span><span class="n">get_residue_mapping</span><span class="p">(</span><span class="n">con_ref_idx</span><span class="p">,</span> <span class="n">con_hal_idx</span><span class="p">)</span>
<span class="c1"># residue_mapping will be: {(&quot;A&quot;, 10): (&quot;A&quot;, 11), (&quot;A&quot;, 20): (&quot;A&quot;, 21)}</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Mapping Creation:</strong> The method pairs each element in con_ref_idx with the corresponding element in con_hal_idx to create the mapping.</p></li>
<li><p><strong>Usage:</strong> This mapping is useful for updating residue selections based on RFdiffusion outputs.</p></li>
</ul>
</dd>
</dl>
<p>This method is designed to facilitate the creation of residue mappings for updating motifs or other residue-based selections.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="protflow.tools.rfdiffusion.parse_diffusion_trbfile">
<span class="sig-prename descclassname"><span class="pre">protflow.tools.rfdiffusion.</span></span><span class="sig-name descname"><span class="pre">parse_diffusion_trbfile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/protflow/tools/rfdiffusion.html#parse_diffusion_trbfile"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.rfdiffusion.parse_diffusion_trbfile" title="Link to this definition"></a></dt>
<dd><p>Parse a .trb file from RFdiffusion and extract relevant scores into a pandas DataFrame.</p>
<p>This method reads a .trb file generated by RFdiffusion, extracts relevant scoring information, and organizes it into a DataFrame. The extracted information includes pLDDT scores, residue indices, and metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>path</strong> (<em>str</em>) – The path to the .trb file.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A DataFrame containing the extracted scores and metadata from the .trb file.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the provided file path does not end with .trb.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>parse_diffusion_trbfile</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">path</span> <span class="o">=</span> <span class="s2">&quot;/path/to/output.trb&quot;</span>
<span class="n">scores_df</span> <span class="o">=</span> <span class="n">rfdiffusion</span><span class="o">.</span><span class="n">parse_diffusion_trbfile</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>
<span class="c1"># scores_df will contain the extracted scores and metadata</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>File Reading:</strong> The method uses numpy to load the .trb file and allows for pickled objects.</p></li>
<li><p><strong>Score Extraction:</strong> Extracted scores include mean pLDDT, per-residue pLDDT, and other relevant metrics.</p></li>
<li><p><strong>Metadata Collection:</strong> Metadata such as file location, description, and input PDB are included in the DataFrame.</p></li>
</ul>
</dd>
</dl>
<p>This method is designed to parse and organize the data from RFdiffusion .trb files, making it easier to analyze the results.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="protflow.tools.rfdiffusion.prep_motif_input">
<span class="sig-prename descclassname"><span class="pre">protflow.tools.rfdiffusion.</span></span><span class="sig-name descname"><span class="pre">prep_motif_input</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Any</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="_modules/protflow/tools/rfdiffusion.html#prep_motif_input"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.rfdiffusion.prep_motif_input" title="Link to this definition"></a></dt>
<dd><p>Ensure motif input is a list and validate that motifs are present in the DataFrame.</p>
<p>This method checks if the given motif is a string or a list, and ensures it is returned as a list. It also validates that the specified motifs are present as columns in the provided DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>motif</strong> (<em>Any</em>) – The motif or list of motifs to validate and process.</p></li>
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – The DataFrame in which to check for the presence of the motifs.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A list of motif column names.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[str]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If any of the specified motifs are not present in the DataFrame.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>prep_motif_input</cite> function:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;motif1&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span> <span class="s2">&quot;motif2&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">4</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">6</span><span class="p">]})</span>
<span class="n">motif</span> <span class="o">=</span> <span class="s2">&quot;motif1&quot;</span>
<span class="n">motifs_list</span> <span class="o">=</span> <span class="n">rfdiffusion</span><span class="o">.</span><span class="n">prep_motif_input</span><span class="p">(</span><span class="n">motif</span><span class="p">,</span> <span class="n">df</span><span class="p">)</span>
<span class="c1"># motifs_list will be: [&quot;motif1&quot;]</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Motif Handling:</strong> The method ensures that even a single motif string is converted into a list to standardize processing.</p></li>
<li><p><strong>Validation:</strong> It checks that each motif in the list is a column in the provided DataFrame, raising an error if any are missing.</p></li>
</ul>
</dd>
</dl>
<p>This function is designed to prepare and validate motif inputs for further processing in the RFdiffusion workflow.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="protflow.tools.rfdiffusion.update_motif_res_mapping">
<span class="sig-prename descclassname"><span class="pre">protflow.tools.rfdiffusion.</span></span><span class="sig-name descname"><span class="pre">update_motif_res_mapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif_l</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="protflow.html#protflow.residues.ResidueSelection" title="protflow.residues.ResidueSelection"><span class="pre">ResidueSelection</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">con_ref_idx</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">con_hal_idx</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span></span></span><a class="reference internal" href="_modules/protflow/tools/rfdiffusion.html#update_motif_res_mapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.rfdiffusion.update_motif_res_mapping" title="Link to this definition"></a></dt>
<dd><p>Update motifs in motif_l based on con_ref_idx and con_hal_idx.</p>
<p>This method updates the residue mappings of motifs in the provided list based on the reference and halogenated residue indices from RFdiffusion outputs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>motif_l</strong> (<em>list</em><em>[</em><a class="reference internal" href="protflow.html#protflow.residues.ResidueSelection" title="protflow.residues.ResidueSelection"><em>ResidueSelection</em></a><em>]</em>) – A list of ResidueSelection objects representing the motifs to be updated.</p></li>
<li><p><strong>con_ref_idx</strong> (<em>list</em>) – A list of reference residue indices from the RFdiffusion outputs.</p></li>
<li><p><strong>con_hal_idx</strong> (<em>list</em>) – A list of halogenated residue indices from the RFdiffusion outputs.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A list of updated ResidueSelection objects with new residue mappings.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>TypeError</strong> – If any element in motif_l is not of type ResidueSelection.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>update_motif_res_mapping</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">motif_l</span> <span class="o">=</span> <span class="p">[</span><span class="n">ResidueSelection</span><span class="p">([</span><span class="s2">&quot;A:10&quot;</span><span class="p">,</span> <span class="s2">&quot;A:20&quot;</span><span class="p">]),</span> <span class="n">ResidueSelection</span><span class="p">([</span><span class="s2">&quot;B:30&quot;</span><span class="p">,</span> <span class="s2">&quot;B:40&quot;</span><span class="p">])]</span>
<span class="n">con_ref_idx</span> <span class="o">=</span> <span class="p">[(</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="mi">10</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="mi">20</span><span class="p">)]</span>
<span class="n">con_hal_idx</span> <span class="o">=</span> <span class="p">[(</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="mi">11</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="mi">21</span><span class="p">)]</span>
<span class="n">updated_motifs</span> <span class="o">=</span> <span class="n">rfdiffusion</span><span class="o">.</span><span class="n">update_motif_res_mapping</span><span class="p">(</span><span class="n">motif_l</span><span class="p">,</span> <span class="n">con_ref_idx</span><span class="p">,</span> <span class="n">con_hal_idx</span><span class="p">)</span>
<span class="c1"># updated_motifs will contain the updated ResidueSelection objects</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Residue Mapping:</strong> The method sets up a mapping dictionary from reference to halogenated residue indices.</p></li>
<li><p><strong>Motif Update:</strong> Each motif in the input list is updated according to the new residue mappings and returned as a new ResidueSelection object.</p></li>
</ul>
</dd>
</dl>
<p>This method is designed to update residue selections in motifs based on the outputs from RFdiffusion, facilitating accurate downstream analysis and interpretation.</p>
</dd></dl>

</section>
<section id="module-protflow.tools.rosetta">
<span id="protflow-tools-rosetta-module"></span><h2>protflow.tools.rosetta module<a class="headerlink" href="#module-protflow.tools.rosetta" title="Link to this heading"></a></h2>
<section id="rosetta-module">
<h3>Rosetta Module<a class="headerlink" href="#rosetta-module" title="Link to this heading"></a></h3>
<p>This module provides the functionality to integrate Rosetta within the ProtFlow framework. It offers tools to run various Rosetta applications, handle their inputs and outputs, and process the resulting data in a structured and automated manner.</p>
<section id="id67">
<h4>Detailed Description<a class="headerlink" href="#id67" title="Link to this heading"></a></h4>
<p>The <cite>Rosetta</cite> class encapsulates the functionality necessary to execute Rosetta runs. It manages the configuration of paths to essential scripts and executables, sets up the environment, and handles the execution of Rosetta processes. It also includes methods for collecting and processing output data, ensuring that the results are organized and accessible for further analysis within the ProtFlow ecosystem.
The module is designed to streamline the integration of Rosetta into larger computational workflows. It supports the automatic setup of job parameters, execution of Rosetta commands, and parsing of output files into a structured DataFrame format. This facilitates subsequent data analysis and visualization steps.</p>
</section>
<section id="id68">
<h4>Usage<a class="headerlink" href="#id68" title="Link to this heading"></a></h4>
<p>To use this module, create an instance of the <cite>Rosetta</cite> class and invoke its <cite>run</cite> method with appropriate parameters. The module will handle the configuration, execution, and result collection processes. Detailed control over the Rosetta process is provided through various parameters, allowing for customized runs tailored to specific research needs.</p>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize and use the <cite>Rosetta</cite> class within a ProtFlow pipeline:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">rosetta</span> <span class="kn">import</span> <span class="n">Rosetta</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the Rosetta class</span>
<span class="n">rosetta</span> <span class="o">=</span> <span class="n">Rosetta</span><span class="p">()</span>

<span class="c1"># Run the Rosetta process</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">rosetta</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">rosetta_application</span><span class="o">=</span><span class="s2">&quot;RosettaScripts&quot;</span><span class="p">,</span>
    <span class="n">nstruct</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
    <span class="n">pose_options</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;-parser:protocol my_protocol.xml&quot;</span><span class="p">],</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id69">
<h4>Further Details<a class="headerlink" href="#id69" title="Link to this heading"></a></h4>
<blockquote>
<div><ul class="simple">
<li><p>Edge Cases: The module handles various edge cases, such as invalid paths for executables and the need to overwrite previous results. It ensures robust error handling and logging for easier debugging and verification of the Rosetta process.</p></li>
<li><p>Customizability: Users can customize the Rosetta process through multiple parameters, including the number of structures (nstruct), specific options for the Rosetta application, and options for handling pose-specific parameters.</p></li>
<li><p>Integration: The module seamlessly integrates with other components of the ProtFlow framework, leveraging shared configurations and data structures to provide a cohesive user experience.</p></li>
</ul>
</div></blockquote>
<p>This module is intended for researchers and developers who need to incorporate Rosetta into their protein design and analysis workflows. By automating many of the setup and execution steps, it allows users to focus on interpreting results and advancing their scientific inquiries.</p>
<p class="rubric">Notes</p>
<p>This module is part of the ProtFlow package and is designed to work in tandem with other components of the package, especially those related to job management in HPC environments.</p>
</section>
<section id="id70">
<h4>Author<a class="headerlink" href="#id70" title="Link to this heading"></a></h4>
<p>Markus Braun, Adrian Tripp</p>
</section>
<section id="id71">
<h4>Version<a class="headerlink" href="#id71" title="Link to this heading"></a></h4>
<p>0.1.0</p>
</section>
</section>
<dl class="py class">
<dt class="sig sig-object py" id="protflow.tools.rosetta.Rosetta">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">protflow.tools.rosetta.</span></span><span class="sig-name descname"><span class="pre">Rosetta</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">script_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/tools/rosetta.html#Rosetta"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.rosetta.Rosetta" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="protflow.html#protflow.runners.Runner" title="protflow.runners.Runner"><code class="xref py py-class docutils literal notranslate"><span class="pre">Runner</span></code></a></p>
<section id="rosetta-class">
<h3>Rosetta Class<a class="headerlink" href="#rosetta-class" title="Link to this heading"></a></h3>
<p>The <cite>Rosetta</cite> class is a specialized class designed to facilitate the execution of Rosetta applications within the ProtFlow framework. It extends the <cite>Runner</cite> class and incorporates specific methods to handle the setup, execution, and data collection associated with Rosetta processes.</p>
<section id="id72">
<h4>Detailed Description<a class="headerlink" href="#id72" title="Link to this heading"></a></h4>
<p>The <cite>Rosetta</cite> class manages all aspects of running Rosetta simulations. It handles the configuration of necessary scripts and executables, prepares the environment for Rosetta processes, and executes the Rosetta commands. Additionally, it collects and processes the output data, organizing it into a structured format for further analysis.</p>
<dl class="simple">
<dt>Key functionalities include:</dt><dd><ul class="simple">
<li><p>Setting up paths to Rosetta scripts and executables.</p></li>
<li><p>Configuring job starter options, either automatically or manually.</p></li>
<li><p>Handling the execution of Rosetta commands with support for multiple structures (nstruct).</p></li>
<li><p>Collecting and processing output data into a pandas DataFrame.</p></li>
<li><p>Cleaning and renaming PDB files based on Rosetta outputs.</p></li>
<li><p>Handling score files and converting them into a readable format.</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">rtype<span class="colon">:</span></dt>
<dd class="field-odd"><p>An instance of the <cite>Rosetta</cite> class, configured to run Rosetta processes and handle outputs efficiently.</p>
</dd>
<dt class="field-even">raises - FileNotFoundError<span class="colon">:</span></dt>
<dd class="field-even"><p>If required files or directories are not found during the execution process.:</p>
</dd>
<dt class="field-odd">raises - ValueError<span class="colon">:</span></dt>
<dd class="field-odd"><p>If invalid arguments are provided to the methods.:</p>
</dd>
<dt class="field-even">raises - KeyError<span class="colon">:</span></dt>
<dd class="field-even"><p>If forbidden options are provided to the methods.:</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize and use the <cite>Rosetta</cite> class:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">rosetta</span> <span class="kn">import</span> <span class="n">Rosetta</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the Rosetta class</span>
<span class="n">rosetta</span> <span class="o">=</span> <span class="n">Rosetta</span><span class="p">()</span>

<span class="c1"># Run the Rosetta process</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">rosetta</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">rosetta_application</span><span class="o">=</span><span class="s2">&quot;RosettaScripts&quot;</span><span class="p">,</span>
    <span class="n">nstruct</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
    <span class="n">pose_options</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;-parser:protocol my_protocol.xml&quot;</span><span class="p">],</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id73">
<h4>Further Details<a class="headerlink" href="#id73" title="Link to this heading"></a></h4>
<blockquote>
<div><ul class="simple">
<li><p>Edge Cases: The class includes handling for various edge cases, such as invalid paths for executables, the need to overwrite previous results, and the presence of existing score files.</p></li>
<li><p>Customization: The class provides extensive customization options through its parameters, allowing users to tailor the Rosetta process to their specific needs.</p></li>
<li><p>Integration: Seamlessly integrates with other ProtFlow components, leveraging shared configurations and data structures for a unified workflow.</p></li>
</ul>
</div></blockquote>
<p>The Rosetta class is intended for researchers and developers who need to perform Rosetta simulations as part of their protein design and analysis workflows. It simplifies the process, allowing users to focus on analyzing results and advancing their research.</p>
<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.rosetta.Rosetta.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">script_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="reference internal" href="_modules/protflow/tools/rosetta.html#Rosetta.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.rosetta.Rosetta.__init__" title="Link to this definition"></a></dt>
<dd><p>Initialize the Rosetta class with the necessary configuration.</p>
<p>This method sets up the Rosetta class with the provided script path and job starter configuration. It initializes the necessary parameters and prepares the environment for executing Rosetta processes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>script_path</strong> (<em>str</em><em>, </em><em>optional</em>) – The path to the Rosetta executable scripts. Defaults to the path specified in <cite>protflow.config.ROSETTA_BIN_PATH</cite>.</p></li>
<li><p><strong>jobstarter</strong> (<a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><em>JobStarter</em></a><em>, </em><em>optional</em>) – An instance of the JobStarter class, which manages job execution. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If no valid script path is provided.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to initialize the <cite>Rosetta</cite> class:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">rosetta</span> <span class="kn">import</span> <span class="n">Rosetta</span>

<span class="c1"># Initialize the Rosetta class with a specific script path</span>
<span class="n">rosetta</span> <span class="o">=</span> <span class="n">Rosetta</span><span class="p">(</span><span class="n">script_path</span><span class="o">=</span><span class="s2">&quot;/path/to/rosetta&quot;</span><span class="p">,</span> <span class="n">jobstarter</span><span class="o">=</span><span class="n">JobStarter</span><span class="p">())</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Configuration:</strong> The method checks if the provided script path is valid and sets it up for further use. If no script path is provided, it defaults to the path specified in the ProtFlow configuration.</p></li>
<li><p><strong>Job Starter:</strong> The job starter parameter can be provided to manage the execution of Rosetta jobs. If not provided, it defaults to None, and the default job starter configuration from ProtFlow will be used.</p></li>
</ul>
</dd>
</dl>
<p>This method ensures that the Rosetta class is correctly initialized with the necessary configurations to run Rosetta applications within the ProtFlow framework.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.rosetta.Rosetta.run">
<span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">poses</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">jobstarter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><span class="pre">JobStarter</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rosetta_application</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nstruct</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pose_options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><span class="pre">Poses</span></a></span></span><a class="reference internal" href="_modules/protflow/tools/rosetta.html#Rosetta.run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.rosetta.Rosetta.run" title="Link to this definition"></a></dt>
<dd><p>Execute the Rosetta process with given poses and jobstarter configuration.</p>
<p>This method sets up and runs the Rosetta process using the provided poses and jobstarter object. It handles the configuration, execution, and collection of output data, ensuring that the results are organized and accessible for further analysis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>poses</strong> (<a class="reference internal" href="protflow.html#protflow.poses.Poses" title="protflow.poses.Poses"><em>Poses</em></a><em>, </em><em>optional</em>) – The Poses object containing the protein structures. Defaults to None.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – A prefix used to name and organize the output files.</p></li>
<li><p><strong>jobstarter</strong> (<a class="reference internal" href="protflow.html#protflow.jobstarters.JobStarter" title="protflow.jobstarters.JobStarter"><em>JobStarter</em></a><em>, </em><em>optional</em>) – An instance of the JobStarter class, which manages job execution. Defaults to None.</p></li>
<li><p><strong>rosetta_application</strong> (<em>str</em><em>, </em><em>optional</em>) – The specific Rosetta application to be executed. Defaults to None.</p></li>
<li><p><strong>nstruct</strong> (<em>int</em><em>, </em><em>optional</em>) – The number of structures to generate for each input pose. Defaults to 1.</p></li>
<li><p><strong>options</strong> (<em>str</em><em>, </em><em>optional</em>) – Additional options for the Rosetta application. Defaults to None.</p></li>
<li><p><strong>pose_options</strong> (<em>list</em><em>[</em><em>str</em><em>] </em><em>| </em><em>str</em><em>, </em><em>optional</em>) – A list of pose-specific options for the Rosetta application. Defaults to None.</p></li>
<li><p><strong>overwrite</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, overwrite existing output files. Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An instance of the RunnerOutput class, containing the processed poses and results of the Rosetta process.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="protflow.html#protflow.runners.RunnerOutput" title="protflow.runners.RunnerOutput">RunnerOutput</a></p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>FileNotFoundError</strong> – If required files or directories are not found during the execution process.</p></li>
<li><p><strong>ValueError</strong> – If invalid arguments are provided to the method.</p></li>
<li><p><strong>KeyError</strong> – If forbidden options are provided to the method.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>run</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">protflow.poses</span> <span class="kn">import</span> <span class="n">Poses</span>
<span class="kn">from</span> <span class="nn">protflow.jobstarters</span> <span class="kn">import</span> <span class="n">JobStarter</span>
<span class="kn">from</span> <span class="nn">rosetta</span> <span class="kn">import</span> <span class="n">Rosetta</span>

<span class="c1"># Create instances of necessary classes</span>
<span class="n">poses</span> <span class="o">=</span> <span class="n">Poses</span><span class="p">()</span>
<span class="n">jobstarter</span> <span class="o">=</span> <span class="n">JobStarter</span><span class="p">()</span>

<span class="c1"># Initialize the Rosetta class</span>
<span class="n">rosetta</span> <span class="o">=</span> <span class="n">Rosetta</span><span class="p">()</span>

<span class="c1"># Run the Rosetta process</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">rosetta</span><span class="o">.</span><span class="n">run</span><span class="p">(</span>
    <span class="n">poses</span><span class="o">=</span><span class="n">poses</span><span class="p">,</span>
    <span class="n">prefix</span><span class="o">=</span><span class="s2">&quot;experiment_1&quot;</span><span class="p">,</span>
    <span class="n">jobstarter</span><span class="o">=</span><span class="n">jobstarter</span><span class="p">,</span>
    <span class="n">rosetta_application</span><span class="o">=</span><span class="s2">&quot;RosettaScripts&quot;</span><span class="p">,</span>
    <span class="n">nstruct</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
    <span class="n">pose_options</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;-parser:protocol my_protocol.xml&quot;</span><span class="p">],</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Access and process the results</span>
<span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Setup and Execution:</strong> The method ensures that the environment is correctly set up, directories are prepared, and necessary commands are constructed and executed. It verifies that the Rosetta application is executable and configures the execution environment accordingly.</p></li>
<li><p><strong>Output Management:</strong> The method handles the collection and processing of output data, ensuring that results are organized and accessible for further analysis. It manages score files, renames PDB files, and compiles output data into a structured format.</p></li>
<li><p><strong>Customization:</strong> Extensive customization options are provided through parameters, allowing users to tailor the Rosetta process to their specific needs. This includes setting the number of structures to generate, providing additional options for the Rosetta application, and specifying pose-specific options.</p></li>
</ul>
</dd>
</dl>
<p>This method is designed to streamline the execution of Rosetta processes within the ProtFlow framework, making it easier for researchers and developers to perform and analyze Rosetta simulations.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="protflow.tools.rosetta.Rosetta.write_cmd">
<span class="sig-name descname"><span class="pre">write_cmd</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rosetta_application</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pose_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">i</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pose_options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/protflow/tools/rosetta.html#Rosetta.write_cmd"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.rosetta.Rosetta.write_cmd" title="Link to this definition"></a></dt>
<dd><p>Writes the command to run a Rosetta application.</p>
<p>This method constructs the command string needed to execute a Rosetta application with the specified options and parameters. It ensures that the command includes all necessary arguments and handles the setup for running the application.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>rosetta_application</strong> (<em>str</em>) – The path to the Rosetta executable.</p></li>
<li><p><strong>pose_path</strong> (<em>str</em>) – The path to the input pose file.</p></li>
<li><p><strong>output_dir</strong> (<em>str</em>) – The directory where output files will be stored.</p></li>
<li><p><strong>i</strong> (<em>int</em>) – The index of the current structure being processed.</p></li>
<li><p><strong>overwrite</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, overwrite existing output files. Defaults to False.</p></li>
<li><p><strong>options</strong> (<em>str</em><em>, </em><em>optional</em>) – Additional options for the Rosetta application. Defaults to None.</p></li>
<li><p><strong>pose_options</strong> (<em>str</em><em>, </em><em>optional</em>) – Pose-specific options for the Rosetta application. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The command string to execute the Rosetta application.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>KeyError</strong> – If forbidden options are included in the provided options or pose_options.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>write_cmd</cite> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">rosetta</span> <span class="kn">import</span> <span class="n">Rosetta</span>

<span class="c1"># Initialize the Rosetta class</span>
<span class="n">rosetta</span> <span class="o">=</span> <span class="n">Rosetta</span><span class="p">(</span><span class="n">script_path</span><span class="o">=</span><span class="s2">&quot;/path/to/rosetta&quot;</span><span class="p">)</span>

<span class="c1"># Write the command to run the Rosetta application</span>
<span class="n">cmd</span> <span class="o">=</span> <span class="n">rosetta</span><span class="o">.</span><span class="n">write_cmd</span><span class="p">(</span>
    <span class="n">rosetta_application</span><span class="o">=</span><span class="s2">&quot;/path/to/rosetta/RosettaScripts&quot;</span><span class="p">,</span>
    <span class="n">pose_path</span><span class="o">=</span><span class="s2">&quot;input.pdb&quot;</span><span class="p">,</span>
    <span class="n">output_dir</span><span class="o">=</span><span class="s2">&quot;/path/to/output&quot;</span><span class="p">,</span>
    <span class="n">i</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">options</span><span class="o">=</span><span class="s2">&quot;-parser:protocol my_protocol.xml&quot;</span><span class="p">,</span>
    <span class="n">pose_options</span><span class="o">=</span><span class="s2">&quot;-in:file:s input.pdb&quot;</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Command Construction:</strong> The method constructs the command string by combining the executable path, input pose path, output directory, and additional options. It ensures that all necessary arguments are included in the command.</p></li>
<li><p><strong>Option Parsing:</strong> The method parses the provided options and pose_options, ensuring they are correctly formatted and do not include any forbidden options. Forbidden options include arguments that could interfere with the correct execution of the Rosetta application, such as output path settings.</p></li>
<li><p><strong>Overwrite Handling:</strong> If the overwrite parameter is set to True, the command includes the necessary argument to overwrite existing output files. This ensures that the process can be re-run without conflicts.</p></li>
</ul>
</dd>
</dl>
<p>This method is designed to facilitate the construction of command strings for running Rosetta applications, making it easier for researchers and developers to execute and manage Rosetta simulations within the ProtFlow framework.</p>
</dd></dl>

</section>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="protflow.tools.rosetta.clean_rosetta_scorefile">
<span class="sig-prename descclassname"><span class="pre">protflow.tools.rosetta.</span></span><span class="sig-name descname"><span class="pre">clean_rosetta_scorefile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path_to_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">str</span></span></span><a class="reference internal" href="_modules/protflow/tools/rosetta.html#clean_rosetta_scorefile"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.rosetta.clean_rosetta_scorefile" title="Link to this definition"></a></dt>
<dd><p>Cleans a faulty Rosetta scorefile.</p>
<p>This function reads a Rosetta scorefile and removes any lines that do not match the expected format (i.e., lines with a different number of columns than the header). It writes the cleaned scores to a new file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path_to_file</strong> (<em>str</em>) – The path to the original Rosetta scorefile.</p></li>
<li><p><strong>out_path</strong> (<em>str</em>) – The path where the cleaned scorefile will be saved.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The path to the cleaned scorefile.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>clean_rosetta_scorefile</cite> function:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">rosetta</span> <span class="kn">import</span> <span class="n">clean_rosetta_scorefile</span>

<span class="c1"># Clean the Rosetta scorefile</span>
<span class="n">cleaned_file_path</span> <span class="o">=</span> <span class="n">clean_rosetta_scorefile</span><span class="p">(</span>
    <span class="n">path_to_file</span><span class="o">=</span><span class="s2">&quot;path/to/original_scorefile.sc&quot;</span><span class="p">,</span>
    <span class="n">out_path</span><span class="o">=</span><span class="s2">&quot;path/to/cleaned_scorefile.sc&quot;</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Cleaned scorefile saved at: </span><span class="si">{</span><span class="n">cleaned_file_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>File Reading:</strong> The function reads the scorefile line by line and splits each line into columns.</p></li>
<li><p><strong>Format Verification:</strong> It verifies that each line has the same number of columns as the header. Lines with a different number of columns are removed.</p></li>
<li><p><strong>File Writing:</strong> The cleaned scores are written to the specified output file. A warning is logged indicating the number of lines removed during the cleaning process.</p></li>
</ul>
</dd>
</dl>
<p>This function is useful for ensuring that Rosetta scorefiles are properly formatted and free of inconsistencies, facilitating accurate data analysis.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="protflow.tools.rosetta.collect_scores">
<span class="sig-prename descclassname"><span class="pre">protflow.tools.rosetta.</span></span><span class="sig-name descname"><span class="pre">collect_scores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">work_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/protflow/tools/rosetta.html#collect_scores"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#protflow.tools.rosetta.collect_scores" title="Link to this definition"></a></dt>
<dd><p>Collects scores from Rosetta output files and reindexes PDB files.</p>
<p>This function collects scores from Rosetta output files, reindexes PDB files based on the scores, and stores the scores in a pandas DataFrame. It also renames PDB files in the working directory to match the reindexed names.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>work_dir</strong> (<em>str</em>) – The directory where Rosetta output files are stored.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A DataFrame containing the collected scores with reindexed PDB file names.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Here is an example of how to use the <cite>collect_scores</cite> function:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">rosetta</span> <span class="kn">import</span> <span class="n">collect_scores</span>

<span class="c1"># Collect scores from the Rosetta output directory</span>
<span class="n">scores_df</span> <span class="o">=</span> <span class="n">collect_scores</span><span class="p">(</span><span class="n">work_dir</span><span class="o">=</span><span class="s2">&quot;/path/to/output_directory&quot;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">scores_df</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>Further Details:</dt><dd><ul class="simple">
<li><p><strong>Score Collection:</strong> The function collects score files from the specified directory and reads them into a pandas DataFrame.</p></li>
<li><p><strong>Reindexing:</strong> The function reindexes the PDB files based on the scores and renames the files in the working directory accordingly.</p></li>
<li><p><strong>File Renaming:</strong> The function ensures that all Rosetta output PDB files are renamed to match the reindexed names, and the paths to these files are stored in the DataFrame.</p></li>
<li><p><strong>Consistency Check:</strong> The function waits for all Rosetta output files to appear in the output directory, ensuring that the renaming process is consistent and complete.</p></li>
</ul>
</dd>
</dl>
<p>This function is designed to streamline the process of collecting and organizing Rosetta output data, making it easier for researchers and developers to analyze the results of Rosetta simulations within the ProtFlow framework.</p>
</dd></dl>

</section>
<section id="protflow-tools-rosettascripts-module">
<h2>protflow.tools.rosettascripts module<a class="headerlink" href="#protflow-tools-rosettascripts-module" title="Link to this heading"></a></h2>
</section>
<section id="module-protflow.tools">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-protflow.tools" title="Link to this heading"></a></h2>
<p>runners submodule init</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="protflow.metrics.html" class="btn btn-neutral float-left" title="protflow.metrics package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Your Name.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>